// @freshworks/crayons/dist/collection/components/input/input.js
import { Host, h } from "@stencil/core";
import { renderHiddenField } from '../../utils/utils';
export class Input {
    constructor() {
        this.hasFocus = false;
        /**
         * Label displayed on the interface, for the component.
         */
        this.label = '';
        /**
         * Default value displayed in the input box.
         */
        this.value = '';
        /**
         * Type of value accepted as the input value. If a user enters a value other than the specified type, the input box is not populated.
         */
        this.type = 'text';
        /**
         * Specifies whether the browser can display suggestions to autocomplete the text value.
         */
        this.autocomplete = 'off';
        /**
         * Specifies whether the browser can auto focus the input field
         */
        this.autofocus = false;
        /**
         * Displays a right-justified clear icon in the text box. Clicking the icon clears the input text. If the attribute’s value is undefined, the value is set to false. For a read-only input box, the clear icon is not displayed unless a default value is specified for the input box.
         */
        this.clearInput = false;
        /**
         * Name of the component, saved as part of form data.
         */
        this.name = '';
        /**
         * Theme based on which the text box is styled.
         */
        this.state = 'normal';
        /**
         * Descriptive or instructional text displayed below the text box.
         */
        this.stateText = '';
        /**
         * If true, the user cannot enter a value in the input box. If the attribute’s value is undefined, the value is set to false.
         */
        this.readonly = false;
        /**
         * Specifies the input box as a mandatory field and displays an asterisk next to the label. If the attribute’s value is undefined, the value is set to false.
         */
        this.required = false;
        /**
         * Disables the component on the interface. If the attribute’s value is undefined, the value is set to false.
         */
        this.disabled = false;
        /**
         * Identifier of the icon that is displayed in the left side of the text box. The attribute’s value must be a valid svg file in the repo of icons (assets/icons).
         */
        this.iconLeft = undefined;
        /**
         * Identifier of the icon that is displayed in the right side of the text box. The attribute’s value must be a valid svg file in the repo of icons (assets/icons).
         */
        this.iconRight = undefined;
        this.onInput = (ev) => {
            const input = ev.target;
            if (input) {
                this.value = input.value || '';
            }
            this.fwInput.emit(ev);
        };
        this.onFocus = () => {
            this.hasFocus = true;
            this.fwFocus.emit();
        };
        this.onBlur = () => {
            this.hasFocus = false;
            this.fwBlur.emit();
        };
        this.clearTextInput = (ev) => {
            if (!this.readonly && !this.disabled && ev) {
                this.value = '';
                if (this.nativeInput) {
                    this.nativeInput.value = '';
                }
                this.fwInputClear.emit();
            }
        };
    }
    watchHandler(newValue) {
        this.fwChange.emit({ value: newValue });
    }
    showClearButton() {
        return this.clearInput && !this.readonly && !this.disabled && this.value.length > 0;
    }
    getValue() {
        return this.value || '';
    }
    hasValue() {
        return this.getValue().length > 0;
    }
    /**
     * Sets focus on a specific `fw-input`. Use this method instead of the global `input.focus()`.
     */
    async setFocus() {
        if (this.nativeInput) {
            this.nativeInput.focus();
        }
    }
    render() {
        const { host, name, value } = this;
        renderHiddenField(host, name, value);
        return (h(Host, { "aria-disabled": this.disabled, class: {
                'has-value': this.hasValue(),
                'has-focus': this.hasFocus,
            } },
            h("div", { class: "input-container" },
                this.label !== '' ? h("label", { class: {
                        'required': this.required,
                    } }, this.label) : '',
                h("div", { class: {
                        'input-container-inner': true,
                        [this.state]: true,
                        'left-icon': this.iconLeft !== undefined,
                        'right-icon': this.iconRight !== undefined,
                    } },
                    h("input", { ref: input => this.nativeInput = input, autoComplete: this.autocomplete, autoFocus: this.autofocus, disabled: this.disabled, name: this.name, placeholder: this.placeholder || '', minLength: this.minlength, maxLength: this.maxlength, readOnly: this.readonly, required: this.required, type: this.type, value: this.value, onInput: e => this.onInput(e), onBlur: this.onBlur, onFocus: this.onFocus }),
                    this.iconLeft !== undefined ? h("fw-icon", { class: "icon left", name: this.iconLeft }) : '',
                    this.iconRight !== undefined ? h("fw-icon", { class: "icon right", name: this.iconRight }) : '',
                    this.showClearButton() ?
                        h("div", { class: "clear-button", onClick: e => this.clearTextInput(e) },
                            h("svg", { xmlns: "http://www.w3.org/2000/svg", width: "32", height: "32", viewBox: "0 0 32 32", class: "clear-button-img" },
                                h("path", { d: "M17.992 16l8.796-8.796a1.409 1.409 0 0 0-1.992-1.992L16 14.008 7.204 5.212a1.409 1.409 0 0 0-1.992 1.992L14.008 16l-8.796 8.796a1.409 1.409 0 0 0 1.992 1.992L16 17.992l8.796 8.796a1.409 1.409 0 0 0 1.992-1.992L17.992 16z" }))) : ''),
                this.stateText !== '' ?
                    h("span", { class: "help-block" }, this.stateText) : '')));
    }
    static get is() { return "fw-input"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() { return {
        "$": ["input.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["input.css"]
    }; }
    static get properties() { return {
        "label": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Label displayed on the interface, for the component."
            },
            "attribute": "label",
            "reflect": false,
            "defaultValue": "''"
        },
        "value": {
            "type": "string",
            "mutable": true,
            "complexType": {
                "original": "string | null",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "Default value displayed in the input box."
            },
            "attribute": "value",
            "reflect": false,
            "defaultValue": "''"
        },
        "type": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "'text' | 'number'",
                "resolved": "\"number\" | \"text\"",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Type of value accepted as the input value. If a user enters a value other than the specified type, the input box is not populated."
            },
            "attribute": "type",
            "reflect": false,
            "defaultValue": "'text'"
        },
        "autocomplete": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "'on' | 'off'",
                "resolved": "\"off\" | \"on\"",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Specifies whether the browser can display suggestions to autocomplete the text value."
            },
            "attribute": "autocomplete",
            "reflect": false,
            "defaultValue": "'off'"
        },
        "autofocus": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Specifies whether the browser can auto focus the input field"
            },
            "attribute": "autofocus",
            "reflect": false,
            "defaultValue": "false"
        },
        "clearInput": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Displays a right-justified clear icon in the text box. Clicking the icon clears the input text. If the attribute\u2019s value is undefined, the value is set to false. For a read-only input box, the clear icon is not displayed unless a default value is specified for the input box."
            },
            "attribute": "clear-input",
            "reflect": false,
            "defaultValue": "false"
        },
        "maxlength": {
            "type": "number",
            "mutable": false,
            "complexType": {
                "original": "number",
                "resolved": "number",
                "references": {}
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "Maximum number of characters a user can enter in the text box."
            },
            "attribute": "maxlength",
            "reflect": false
        },
        "minlength": {
            "type": "number",
            "mutable": false,
            "complexType": {
                "original": "number",
                "resolved": "number",
                "references": {}
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "Minimum number of characters a user must enter in the text box for the value to be valid."
            },
            "attribute": "minlength",
            "reflect": false
        },
        "name": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Name of the component, saved as part of form data."
            },
            "attribute": "name",
            "reflect": false,
            "defaultValue": "''"
        },
        "placeholder": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string | null",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "Text displayed in the text box before a user enters a value."
            },
            "attribute": "placeholder",
            "reflect": false
        },
        "state": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "'normal' | 'warning' | 'error'",
                "resolved": "\"error\" | \"normal\" | \"warning\"",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Theme based on which the text box is styled."
            },
            "attribute": "state",
            "reflect": false,
            "defaultValue": "'normal'"
        },
        "stateText": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Descriptive or instructional text displayed below the text box."
            },
            "attribute": "state-text",
            "reflect": false,
            "defaultValue": "''"
        },
        "readonly": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "If true, the user cannot enter a value in the input box. If the attribute\u2019s value is undefined, the value is set to false."
            },
            "attribute": "readonly",
            "reflect": false,
            "defaultValue": "false"
        },
        "required": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Specifies the input box as a mandatory field and displays an asterisk next to the label. If the attribute\u2019s value is undefined, the value is set to false."
            },
            "attribute": "required",
            "reflect": false,
            "defaultValue": "false"
        },
        "disabled": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Disables the component on the interface. If the attribute\u2019s value is undefined, the value is set to false."
            },
            "attribute": "disabled",
            "reflect": false,
            "defaultValue": "false"
        },
        "iconLeft": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Identifier of the icon that is displayed in the left side of the text box. The attribute\u2019s value must be a valid svg file in the repo of icons (assets/icons)."
            },
            "attribute": "icon-left",
            "reflect": false,
            "defaultValue": "undefined"
        },
        "iconRight": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Identifier of the icon that is displayed in the right side of the text box. The attribute\u2019s value must be a valid svg file in the repo of icons (assets/icons)."
            },
            "attribute": "icon-right",
            "reflect": false,
            "defaultValue": "undefined"
        }
    }; }
    static get states() { return {
        "hasFocus": {}
    }; }
    static get events() { return [{
            "method": "fwChange",
            "name": "fwChange",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": "Triggered when the value in the input box is modified."
            },
            "complexType": {
                "original": "any",
                "resolved": "any",
                "references": {}
            }
        }, {
            "method": "fwFocus",
            "name": "fwFocus",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": "Triggered when the input box comes into focus."
            },
            "complexType": {
                "original": "void",
                "resolved": "void",
                "references": {}
            }
        }, {
            "method": "fwBlur",
            "name": "fwBlur",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": "Triggered when the input box loses focus."
            },
            "complexType": {
                "original": "void",
                "resolved": "void",
                "references": {}
            }
        }, {
            "method": "fwInput",
            "name": "fwInput",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": "Triggered when a value is entered in the input box."
            },
            "complexType": {
                "original": "KeyboardEvent",
                "resolved": "KeyboardEvent",
                "references": {
                    "KeyboardEvent": {
                        "location": "global"
                    }
                }
            }
        }, {
            "method": "fwInputClear",
            "name": "fwInputClear",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": "Triggered when clear icon is clicked."
            },
            "complexType": {
                "original": "any",
                "resolved": "any",
                "references": {}
            }
        }]; }
    static get methods() { return {
        "setFocus": {
            "complexType": {
                "signature": "() => Promise<void>",
                "parameters": [],
                "references": {
                    "Promise": {
                        "location": "global"
                    }
                },
                "return": "Promise<void>"
            },
            "docs": {
                "text": "Sets focus on a specific `fw-input`. Use this method instead of the global `input.focus()`.",
                "tags": []
            }
        }
    }; }
    static get elementRef() { return "host"; }
    static get watchers() { return [{
            "propName": "value",
            "methodName": "watchHandler"
        }]; }
}
