// @freshworks/crayons/dist/collection/components/tabs/tabs.js
import { h } from "@stencil/core";
export class Tabs {
    constructor() {
        /**
         * The index of the activated Tab(Starts from 0)
         */
        this.activeTabIndex = 0;
        /**
         * Active class for tab container
         */
        this.activeChildClass = '';
    }
    changeIndex(newValue, oldValue) {
        var _a;
        if (((_a = this.tabs[newValue]) === null || _a === void 0 ? void 0 : _a.disabled) || newValue >= this.tabs.length) {
            this.activeTabIndex = 0;
        }
        else if (newValue !== oldValue) {
            this.displayTab(newValue);
        }
    }
    init() {
        this.tabs = Array.from(this.el.querySelectorAll('fw-tab')).filter(node => node.parentNode === this.el);
        this.displayTab(this.activeTabIndex);
    }
    displayTab(index) {
        var _a;
        this.fwChange.emit({ tabIndex: this.activeTabIndex });
        this.activeTabIndex = index;
        this.tabs = (_a = this.tabs) === null || _a === void 0 ? void 0 : _a.map((tab, i) => {
            tab.style.display = index === i ? 'block' : 'none';
            return tab;
        });
    }
    componentWillLoad() {
        this.init();
    }
    connectedCallback() {
        this.mutationO = new MutationObserver(() => {
            this.init();
        });
        this.mutationO.observe(this.el, {
            childList: true,
        });
    }
    disconnectedCallback() {
        if (this.mutationO) {
            this.mutationO.disconnect();
            this.mutationO = undefined;
        }
    }
    render() {
        return (h("div", { class: "tabs" },
            h("ul", { role: "tablist", class: "tabs__items" }, this.tabs.map((tab, index) => h("li", { onClick: () => tab.disabled ? '' : this.activeTabIndex = index, class: "tabs__item" },
                h("div", { id: '#tab-' + index, class: 'tabs__item__nav ' + (index === this.activeTabIndex ? 'active' : '') + (tab.disabled ? 'disabled' : '') },
                    h("span", { class: "tab-title--tab-icon" }, tab.tabHeaderHtml
                        ? h("span", { innerHTML: tab.tabHeaderHtml })
                        : h("span", { class: "tab-title" }, tab.tabHeader)))))),
            h("div", { class: "tabs__content" },
                h("slot", null))));
    }
    static get is() { return "fw-tabs"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() { return {
        "$": ["tabs.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["tabs.css"]
    }; }
    static get properties() { return {
        "activeTabIndex": {
            "type": "number",
            "mutable": true,
            "complexType": {
                "original": "number",
                "resolved": "number",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "The index of the activated Tab(Starts from 0)"
            },
            "attribute": "active-tab-index",
            "reflect": true,
            "defaultValue": "0"
        }
    }; }
    static get states() { return {
        "tabs": {},
        "activeChildClass": {}
    }; }
    static get events() { return [{
            "method": "fwChange",
            "name": "fwChange",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": "Triggered when a the view switches to a new tab."
            },
            "complexType": {
                "original": "any",
                "resolved": "any",
                "references": {}
            }
        }]; }
    static get elementRef() { return "el"; }
    static get watchers() { return [{
            "propName": "activeTabIndex",
            "methodName": "changeIndex"
        }]; }
    static get listeners() { return [{
            "name": "propChanged",
            "method": "init",
            "target": undefined,
            "capture": false,
            "passive": false
        }]; }
}
