// @freshworks/crayons/dist/collection/components/toggle/toggle.js
import { h } from "@stencil/core";
export class Toggle {
    constructor() {
        /**
         * Sets the selected state as the default state. If the attribute’s value is undefined, the value is set to false.
         */
        this.checked = false;
        /**
         * Size of the input control.
         */
        this.size = 'medium';
        /**
         * Name of the component, saved as part of the form data.
         */
        this.name = '';
        /**
         * Specifies whether to disable the control on the interface. If the attribute’s value is undefined, the value is set to false.
         */
        this.disabled = false;
    }
    watchHandler(newValue) {
        this.fwChange.emit({ checked: newValue });
    }
    toggle() {
        if (!this.disabled) {
            this.checked = !this.checked;
        }
    }
    render() {
        return (h("div", { class: {
                'toggle-switch': true,
                [this.size]: true,
            }, onClick: () => this.toggle() },
            h("input", { name: this.name, type: "checkbox", disabled: this.disabled, checked: this.checked, class: "checkboxClass" }),
            h("span", { class: {
                    'slider': true,
                    [this.size]: true,
                } })));
    }
    static get is() { return "fw-toggle"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() { return {
        "$": ["toggle.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["toggle.css"]
    }; }
    static get properties() { return {
        "checked": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Sets the selected state as the default state. If the attribute\u2019s value is undefined, the value is set to false."
            },
            "attribute": "checked",
            "reflect": false,
            "defaultValue": "false"
        },
        "size": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "'small' | 'medium' | 'large'",
                "resolved": "\"large\" | \"medium\" | \"small\"",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Size of the input control."
            },
            "attribute": "size",
            "reflect": false,
            "defaultValue": "'medium'"
        },
        "name": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Name of the component, saved as part of the form data."
            },
            "attribute": "name",
            "reflect": false,
            "defaultValue": "''"
        },
        "disabled": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Specifies whether to disable the control on the interface. If the attribute\u2019s value is undefined, the value is set to false."
            },
            "attribute": "disabled",
            "reflect": false,
            "defaultValue": "false"
        }
    }; }
    static get events() { return [{
            "method": "fwChange",
            "name": "fwChange",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": "Triggered when the input control is selected or deselected."
            },
            "complexType": {
                "original": "any",
                "resolved": "any",
                "references": {}
            }
        }]; }
    static get watchers() { return [{
            "propName": "checked",
            "methodName": "watchHandler"
        }]; }
}
