// @freshworks/crayons/dist/collection/components/dropdown-button/dropdown-button.js
import { h } from "@stencil/core";
export class DropdownButton {
    constructor() {
        /**
         * Dropdown Button color
         */
        this.color = 'primary';
        /**
         * Displays a split dropdown button
         */
        this.split = false;
        /**
         * Disables the dropdown button if its true
         */
        this.disabled = false;
        /**
         * Displays a searchable dropdown button
         */
        this.searchable = false;
        /**
         * Value of the dropdown button
         */
        this.value = undefined;
        /**
         * Placeholder text for search input. Validated only if dropdown and searchable is true
         */
        this.placeholder = '';
        /**
         * Options to show in the dropdown button
         */
        this.options = [];
        /**
         * Option input for searching through options
         */
        this.optionInput = '';
        /**
         * Filtered options if the dropdown is searchable
         */
        this.filteredOptions = [];
        /**
         * State representing dropdown is open or not
         */
        this.isDropdownOpen = false;
    }
    setInitialState(newVal) {
        // Do not open dropdown if options is absent
        if (newVal) {
            if (this.options.length === 0) {
                this.isDropdownOpen = false;
            }
        }
        // Reset searchable dropdown to initial state on close
        if (!newVal) {
            if (this.searchable) {
                this.filteredOptions = this.options;
                this.dropdownInput.value = '';
            }
        }
    }
    /**
     * Toggles dropdown b/w open and close
     */
    handleDropdownToggle() {
        if (!this.disabled) {
            this.isDropdownOpen = !this.isDropdownOpen;
        }
    }
    /**
     * Sets dropdown options from slot
     */
    setDropdownOptions() {
        // Set options if the prop is not supplied by user
        if (this.options.length === 0) {
            this.options = Array.from(this.host.querySelectorAll('[slot=dropdown-options] option')).map(option => ({
                id: option.id,
                value: option.getAttribute('value'),
                label: option.textContent,
            }));
        }
        // Set filtered options to inital options for search dropdown
        if (this.searchable) {
            this.filteredOptions = this.options;
        }
    }
    /**
     * Sets search dropdown input
     */
    setSearchInput() {
        this.optionInput = this.dropdownInput.value;
        this.filteredOptions = this.optionInput !== ''
            ? this.options.filter(option => option.label.toLowerCase()
                .includes(this.optionInput.toLowerCase()))
            : this.options;
    }
    /**
     * Listens to checkbox change for search dropdown
     */
    handleCheckboxChange(ev) {
        if (ev.target.checked) {
            this.value = [
                ...(this.value || []),
                this.options.find(({ value }) => value === ev.target.id).value,
            ];
        }
        else {
            this.value = this.value.filter(val => val !== ev.target.id);
        }
    }
    /**
     * Handles `Add` button click in search dropdown
     */
    handleAddClick() {
        this.fwOptionsAdd.emit({ value: this.value });
        this.handleDropdownToggle();
    }
    /**
     * Handles option select from the dropdown
     */
    handleOptionClick(option) {
        if (!this.searchable) {
            this.value = option.value;
            this.fwOptionClick.emit({ value: this.value, label: option.label });
            this.handleDropdownToggle();
        }
    }
    render() {
        const ChevronArrow = () => {
            const iconColor = ['secondary', 'link', 'text'].includes(this.color) ? '#12344d' : '#fff';
            const iconSize = 8;
            const direction = this.isDropdownOpen ? 'up' : 'down';
            return h("fw-icon", { name: `chevron-${direction}`, color: iconColor, size: iconSize }, " ");
        };
        const DropdownMenu = () => {
            const SearchInput = () => {
                return (h("fw-input", { placeholder: this.placeholder, "icon-left": "search", ref: dropdownInput => this.dropdownInput = dropdownInput, onInput: () => this.setSearchInput() }));
            };
            const renderBtnGroup = () => {
                return (h("div", { class: "search-btn-grp" },
                    h("fw-button", { id: "addBtn", size: "small", color: "primary", onClick: () => this.handleAddClick() }, " Add "),
                    h("fw-button", { id: "cancelBtn", size: "small", color: "secondary", onClick: () => this.handleDropdownToggle() }, " Cancel ")));
            };
            const validOptions = this.searchable ? this.filteredOptions : this.options;
            return (h("ul", { class: `dropdown-menu ${this.isDropdownOpen ? 'dropdown-menu--open' : ''}` },
                this.searchable ? h(SearchInput, null) : '',
                h("div", { class: this.searchable ? `search-list` : '' }, validOptions.map(option => {
                    const liEl = h("li", { key: option.id || option.value, onClick: () => this.handleOptionClick(option), class: "dropdown-item" },
                        " ",
                        option.label,
                        " ");
                    const checkboxEl = h("fw-checkbox", { id: option.value, checked: (this.value || []).includes(option.value), onFwChange: e => this.handleCheckboxChange(e) });
                    return this.searchable
                        ? h("div", { class: "searchable-item" },
                            " ",
                            checkboxEl,
                            " ",
                            liEl,
                            " ")
                        : liEl;
                })),
                this.searchable ? renderBtnGroup() : ''));
        };
        const DropdownState = () => (!this.split ? h("span", { class: "down-arrow" },
            h(ChevronArrow, null)) : '');
        const SplitDropdownState = () => {
            const dropdownStateClasses = `
        dropdown-state
        ${'dropdown-state--' + this.color.toLowerCase()}
        ${this.disabled ? 'disabled' : ''}
      `;
            return (this.split ?
                h("div", { onClick: () => this.handleDropdownToggle(), class: dropdownStateClasses },
                    h("div", { class: "state-icon" },
                        h(ChevronArrow, null))) : '');
        };
        return (h("div", { class: "dropdown-container" },
            h("div", { class: "btn-container" },
                h("fw-button", { color: this.color, disabled: this.disabled, onClick: () => this.handleDropdownToggle() },
                    this.label,
                    h(DropdownState, null)),
                h(SplitDropdownState, null)),
            h(DropdownMenu, null)));
    }
    componentWillLoad() {
        document.addEventListener('click', e => {
            const { target } = e;
            const canClose = this.isDropdownOpen
                && this.host !== target
                && !this.host.contains(target);
            if (canClose) {
                this.isDropdownOpen = false;
            }
        });
        this.setDropdownOptions();
    }
    static get is() { return "fw-dropdown-button"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() { return {
        "$": ["dropdown-button.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["dropdown-button.css"]
    }; }
    static get properties() { return {
        "label": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Label for the dropdown button"
            },
            "attribute": "label",
            "reflect": false
        },
        "color": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "'primary' | 'secondary' | 'danger' | 'link' | 'text'",
                "resolved": "\"danger\" | \"link\" | \"primary\" | \"secondary\" | \"text\"",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Dropdown Button color"
            },
            "attribute": "color",
            "reflect": false,
            "defaultValue": "'primary'"
        },
        "split": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Displays a split dropdown button"
            },
            "attribute": "split",
            "reflect": false,
            "defaultValue": "false"
        },
        "disabled": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Disables the dropdown button if its true"
            },
            "attribute": "disabled",
            "reflect": false,
            "defaultValue": "false"
        },
        "searchable": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Displays a searchable dropdown button"
            },
            "attribute": "searchable",
            "reflect": false,
            "defaultValue": "false"
        },
        "value": {
            "type": "any",
            "mutable": false,
            "complexType": {
                "original": "any",
                "resolved": "any",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Value of the dropdown button"
            },
            "attribute": "value",
            "reflect": false,
            "defaultValue": "undefined"
        },
        "placeholder": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Placeholder text for search input. Validated only if dropdown and searchable is true"
            },
            "attribute": "placeholder",
            "reflect": false,
            "defaultValue": "''"
        },
        "options": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "any[]",
                "resolved": "any[]",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Options to show in the dropdown button"
            },
            "defaultValue": "[]"
        }
    }; }
    static get states() { return {
        "optionInput": {},
        "filteredOptions": {},
        "isDropdownOpen": {}
    }; }
    static get events() { return [{
            "method": "fwOptionClick",
            "name": "fwOptionClick",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": "Triggered when an option is clicked"
            },
            "complexType": {
                "original": "any",
                "resolved": "any",
                "references": {}
            }
        }, {
            "method": "fwOptionsAdd",
            "name": "fwOptionsAdd",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": "Triggered when Add button for searchable dropdown is clicked"
            },
            "complexType": {
                "original": "any",
                "resolved": "any",
                "references": {}
            }
        }]; }
    static get elementRef() { return "host"; }
    static get watchers() { return [{
            "propName": "isDropdownOpen",
            "methodName": "setInitialState"
        }]; }
}
