// @freshworks/crayons/dist/cjs/fw-checkbox.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-19272a91.js');
const utils = require('./utils-a88d9e0c.js');

const Checkbox = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        /**
         * Sets the state of the check box to selected. If the attribute’s value is undefined, the value is set to false.
         */
        this.checked = false;
        /**
         * Disables the check box on the interface. If the attribute’s value is undefined, the value is set to false.
         */
        this.disabled = false;
        /**
         * Label displayed on the interface, for the check box.
         */
        this.label = '';
        /**
         * Name of the component, saved as part of form data.
         */
        this.name = '';
        /**
         * Identifier corresponding to the component, that is saved when the form data is saved.
         */
        this.value = '';
        this.fwChange = index.createEvent(this, "fwChange", 7);
        this.fwFocus = index.createEvent(this, "fwFocus", 7);
        this.fwBlur = index.createEvent(this, "fwBlur", 7);
    }
    componentDidLoad() {
        this.checkbox.checked = this.checked;
        this.checkbox.disabled = this.disabled;
    }
    checkChanged(isChecked) {
        if (!this.disabled) {
            this.fwChange.emit({
                value: this.value,
                checked: isChecked,
            });
        }
        this.checkbox.checked = isChecked;
    }
    disabledChanged(isDisabled) {
        this.checkbox.disabled = isDisabled;
    }
    onFocus() {
        this.fwFocus.emit();
    }
    onBlur() {
        this.fwBlur.emit();
    }
    toggle() {
        if (!this.disabled) {
            this.checked = !this.checked;
        }
    }
    render() {
        const { host, name, value } = this;
        if (this.checked) {
            utils.renderHiddenField(host, name, value);
        }
        return (index.h(index.Host, { class: "checkbox-container", onClick: () => this.toggle(), role: "checkbox", tabIndex: "0", "aria-disabled": this.disabled ? 'true' : 'false', "aria-checked": `${this.checked}`, onFocus: () => this.onFocus(), onBlur: () => this.onBlur() }, index.h("input", { type: "checkbox", ref: el => this.checkbox = el }), index.h("label", null, index.h("span", { class: "text" }, index.h("slot", null)), index.h("br", null), this.label !== ''
            ? index.h("span", { class: "label-field" }, this.label)
            : '')));
    }
    get host() { return index.getElement(this); }
    static get watchers() { return {
        "checked": ["checkChanged"],
        "disabled": ["disabledChanged"]
    }; }
    static get style() { return "/* Need to check with designer */\n/* Need to check with designer */\n:host {\n  font-family: -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, Oxygen, Ubuntu, Cantarell, \"Open Sans\", \"Helvetica Neue\", sans-serif;\n}\n\n:host(.checkbox-container) {\n  display: inline;\n  position: relative;\n}\n\n/* Focus event occurs on the root element */\n:host(:focus) {\n  outline: none;\n}\n:host(:focus) input[type=checkbox] + label::before {\n  border: 1px solid transparent;\n  -webkit-box-shadow: 0 0 0 2px #2c5cc5;\n  box-shadow: 0 0 0 2px #2c5cc5;\n}\n:host(:focus) input[type=checkbox][disabled] + label::before {\n  -webkit-box-shadow: none;\n  box-shadow: none;\n  border: 1px solid #dadfe3;\n}\n\n.label-field {\n  font-size: 12px;\n  color: #475867;\n  letter-spacing: 0;\n  line-height: 20px;\n  position: relative;\n  padding-left: 28px;\n  font-weight: 400;\n}\n\n/* Hover event occurs on the root element */\n:host(:hover) input[type=checkbox] + label::before {\n  border-color: #12344d;\n  -webkit-box-shadow: 0 0 0 5px #ebeff3;\n  box-shadow: 0 0 0 5px #ebeff3;\n}\n:host(:hover) input[type=checkbox][disabled] + label {\n  cursor: not-allowed;\n}\n:host(:hover) input[type=checkbox][disabled] + label::before {\n  -webkit-box-shadow: none;\n  box-shadow: none;\n  border: 1px solid #dadfe3;\n}\n\ninput[type=checkbox] {\n  display: none;\n}\ninput[type=checkbox] + label {\n  -webkit-user-select: none;\n     -moz-user-select: none;\n      -ms-user-select: none;\n          user-select: none;\n  cursor: pointer;\n  margin-bottom: 0;\n  vertical-align: middle;\n  font-size: 12px;\n  color: #12344d;\n  font-weight: 500;\n}\ninput[type=checkbox] + label .text {\n  padding-left: 27px;\n}\ninput[type=checkbox] + label::before {\n  position: absolute;\n  left: 0;\n  top: 3px;\n  display: block;\n  content: \"\";\n  border: 1px solid #cfd7df;\n  height: 14px;\n  width: 14px;\n  background-color: #fff;\n  -webkit-transition: all 0.2s ease;\n  transition: all 0.2s ease;\n  -webkit-box-sizing: border-box;\n  box-sizing: border-box;\n  border-radius: 2px;\n}\ninput[type=checkbox] + label::after {\n  position: absolute;\n  display: block;\n  content: \"\";\n  left: 4px;\n  top: 5px;\n  width: 5px;\n  height: 7px;\n  -webkit-transform: rotate(45deg);\n  transform: rotate(45deg);\n  opacity: 0;\n  -webkit-transition: opacity 0.2s ease-in-out;\n  transition: opacity 0.2s ease-in-out;\n  -webkit-box-sizing: border-box;\n  box-sizing: border-box;\n}\ninput[type=checkbox]:checked + label::before {\n  background: #2c5cc5;\n  border-color: #2c5cc5;\n  -webkit-box-shadow: none;\n  box-shadow: none;\n}\ninput[type=checkbox]:checked + label::after {\n  border-right: 2px solid #fff;\n  border-bottom: 2px solid #fff;\n  opacity: 1;\n  top: 5px;\n}\ninput[type=checkbox]:checked:hover + label::before {\n  -webkit-box-shadow: 0 0 0 5px #ebeff3;\n  box-shadow: 0 0 0 5px #ebeff3;\n}\ninput[type=checkbox]:checked:hover + label::after {\n  border-right: 2px solid #fff;\n  border-bottom: 2px solid #fff;\n  opacity: 1;\n  top: 5px;\n}\ninput[type=checkbox]:checked:focus + label::before {\n  background: #2c5cc5;\n  border-color: #fff;\n  -webkit-box-shadow: 0 0 0 1px #2c5cc5;\n  box-shadow: 0 0 0 1px #2c5cc5;\n}\ninput[type=checkbox]:checked:focus + label::after {\n  border-right: 2px solid #fff;\n  border-bottom: 2px solid #fff;\n  opacity: 1;\n  top: 5px;\n}\ninput[type=checkbox][disabled] + label {\n  color: #92a2b1;\n}\ninput[type=checkbox][disabled] + label .label-field {\n  color: #92a2b1;\n}\ninput[type=checkbox][disabled] + label::before {\n  border-color: #dadfe3;\n  background-color: #ebeff3;\n}\ninput[type=checkbox][disabled]:checked + label {\n  color: #92a2b1;\n}\ninput[type=checkbox][disabled]:checked + label::before {\n  background: #dadfe3;\n}"; }
};

exports.fw_checkbox = Checkbox;
