// @freshworks/crayons/dist/cjs/fw-datepicker.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-19272a91.js');
const moment_min = require('./moment.min-0023acf8.js');
const utils = require('./utils-a88d9e0c.js');

const weekDay = ['S', 'M', 'T', 'W', 'T', 'F', 'S'];
const monthArr = [
    'Jan',
    'Feb',
    'Mar',
    'Apr',
    'May',
    'Jun',
    'Jul',
    'Aug',
    'Sep',
    'Oct',
    'Nov',
    'Dec',
];
const Datepicker = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        /**
         *   Type of date selection enabled for the calendar. If the value is range, a user can select a date range in the calendar.
         */
        this.mode = 'single date';
        /**
         *   Format in which the date values selected in the calendar are populated in the input box and saved when the form data is saved.
         */
        this.dateFormat = 'DD-MM-YYYY';
        /**
         * Name of the component, saved as part of form data.
         */
        this.name = '';
        this.getSupportedYears = () => {
            const yearsArr = [];
            let year = 1970;
            while (year < 2050) {
                yearsArr.push(year.toString());
                year++;
            }
            return yearsArr;
        };
        this.getDayDetails = args => {
            const date = args.index - args.firstDay;
            const day = args.index % 7;
            let prevMonth = args.month - 1;
            let prevYear = Number(args.year);
            if (prevMonth < 0) {
                prevMonth = 11;
                prevYear--;
            }
            const prevMonthNumberOfDays = moment_min.moment([prevYear, prevMonth]).daysInMonth() || 0;
            const _date = (date < 0
                ? prevMonthNumberOfDays + date
                : date % args.numberOfDays) + 1;
            const month = this._getValidDateInMonth(date, args);
            const timestamp = moment_min.moment([args.year, args.month, _date]).valueOf();
            return { date: _date, day, month, timestamp };
        };
        this.getMonthDetails = (year, month) => {
            const firstDay = (new Date(year, month)).getDay();
            const numberOfDays = moment_min.moment([year, month]).daysInMonth() || 0;
            const monthArray = [];
            const rows = 6;
            let currentDay;
            let index = 0;
            const cols = 7;
            for (let row = 0; row < rows; row++) {
                for (let col = 0; col < cols; col++) {
                    currentDay = this.getDayDetails({ index, numberOfDays, firstDay, year, month });
                    monthArray.push(currentDay);
                    index++;
                }
            }
            return monthArray;
        };
        this.setMonth = offset => {
            let year = Number(this.year);
            let month = this.month + offset;
            if (month === -1) {
                month = 11;
                year--;
            }
            else if (month === 12) {
                month = 0;
                year++;
            }
            this.year = year.toString();
            this.month = month;
            this.toMonth = this.month === 11 ? 0 : this.month + 1;
            this.toYear = this.toMonth === 0 ? this.yearCalculation(this.year, 1) : this.year;
            this.monthDetails = this.getMonthDetails(year, month);
            this.nextMonthDetails = this.month === 11
                ? this.getMonthDetails(this.yearCalculation(this.year, 1), 0)
                : this.getMonthDetails(this.year, this.month + 1);
        };
        this.isCurrentDay = day => {
            return day.timestamp === this.todayTimestamp;
        };
        this.isSelectedDay = ({ timestamp }) => {
            return timestamp === this.selectedDay || timestamp === this.startDate || timestamp === this.endDate;
        };
        this.isInRange = ({ timestamp }) => {
            const { endDate } = this;
            const { startDate } = this;
            return startDate && endDate && ((timestamp >= startDate && timestamp <= endDate));
        };
        this.onDateClick = ({ timestamp }) => {
            if (this.showSingleDatePicker()) {
                this.selectedDay = timestamp;
            }
            else if (this.showDateRangePicker()) {
                this.handleRangeSelection(timestamp);
                this.dateHovered = '';
            }
        };
        this.fwChange = index.createEvent(this, "fwChange", 7);
    }
    displayDatePicker() {
        this.showDatePicker = true;
    }
    handleButtonClick(e) {
        const isUpdateRange = e.composedPath()[0].classList.value.includes('update-range-value');
        const isUpdateDate = e.composedPath()[0].classList.value.includes('update-date-value');
        if (isUpdateRange) {
            this.startDateFormatted = moment_min.moment(this.startDate).format(this.dateFormat);
            this.endDateFormatted = moment_min.moment(this.endDate).format(this.dateFormat);
            if (this.startDate && this.endDate) {
                this.value = this.startDateFormatted + ' To ' + this.endDateFormatted;
            }
            this.fromDate = this.startDateFormatted;
            this.toDate = this.endDateFormatted;
            this.fwChange.emit({ fromDate: this.startDateFormatted, toDate: this.endDateFormatted });
        }
        else if (isUpdateDate) {
            this.value = this.selectedDay ? moment_min.moment(this.selectedDay).format(this.dateFormat) : '';
            this.fwChange.emit(this.value);
        }
        this.showDatePicker = false;
    }
    /* Listener to handle Month Year dropdown changes
    */
    handleMonthYearDropDownSelection(e) {
        if (e.path[0].tagName !== 'FW-DATEPICKER') {
            e.stopPropagation();
        }
        const newValue = e.detail && e.detail.value;
        if (!newValue) {
            return;
        }
        if (this.mode === 'range') {
            this.handleDateRangeDropDownUpdate(e, newValue);
            this.nextMonthDetails = this.getMonthDetails(this.toYear, this.toMonth);
        }
        else {
            this.handleSingleDateDropDownUpdate(e, newValue);
        }
        this.monthDetails = this.getMonthDetails(this.year, this.month);
    }
    handleSingleDateDropDownUpdate(e, newValue) {
        const isMonthUpdate = e.composedPath()[0].classList.value.includes('single-month-selector');
        const isYearUpdate = e.composedPath()[0].classList.value.includes('single-year-selector');
        if (isMonthUpdate) {
            this.month = monthArr.indexOf(newValue);
        }
        else if (isYearUpdate) {
            this.year = newValue;
        }
    }
    handleDateRangeDropDownUpdate(e, newValue) {
        const isFromMonthUpdate = e.composedPath()[0].classList.value.includes('from-month-selector');
        const isFromYearUpdate = e.composedPath()[0].classList.value.includes('from-year-selector');
        const isToMonthUpdate = e.composedPath()[0].classList.value.includes('to-month-selector');
        const isToYearUpdate = e.composedPath()[0].classList.value.includes('to-year-selector');
        if (isFromMonthUpdate) {
            this.month = monthArr.indexOf(newValue);
            if (this.month === 11) {
                this.toMonth = 0;
                this.toYear = this.yearCalculation(this.year, 1);
            }
            else {
                this.toMonth = this.month + 1;
            }
        }
        else if (isFromYearUpdate) {
            this.year = newValue;
            this.toYear = this.month === 11 ? this.yearCalculation(this.year, 1) : this.year;
        }
        else if (isToMonthUpdate) {
            this.toMonth = monthArr.indexOf(newValue);
            if (this.toMonth === 0) {
                this.month = 11;
                this.year = this.yearCalculation(this.toYear, -1);
            }
            else {
                this.month = this.toMonth - 1;
            }
        }
        else if (isToYearUpdate) {
            this.toYear = newValue;
            this.year = this.toMonth === 0 ? this.yearCalculation(this.toYear, -1) : this.toYear;
        }
    }
    yearCalculation(year, offset) {
        const resultYear = Number(year) + offset;
        return resultYear.toString();
    }
    componentWillLoad() {
        this.year = moment_min.moment().year().toString();
        this.month = moment_min.moment().month();
        this.toMonth = this.month === 11 ? 0 : this.month + 1;
        this.toYear = this.toMonth === 0 ? this.yearCalculation(this.year, 1) : this.year;
        this.monthDetails = this.getMonthDetails(this.year, this.month);
        this.todayTimestamp = moment_min.moment().startOf('date').valueOf();
        this.setInitalValues();
    }
    setInitalValues() {
        this.nextMonthDetails = this.month === 11
            ? this.getMonthDetails(this.yearCalculation(this.year, 1), 0)
            : this.getMonthDetails(this.year, this.month + 1);
        this.placeholder = this.placeholder || (this.mode === 'range'
            ? 'Select Date Range'
            : 'Select Date');
        this.supportedYears = this.getSupportedYears();
        this.selectedDay = this.value !== undefined ? moment_min.moment(this.value, this.dateFormat).valueOf() : undefined;
        this.startDate = this.fromDate !== undefined ? moment_min.moment(this.fromDate, this.dateFormat).valueOf() : undefined;
        this.endDate = this.toDate !== undefined ? moment_min.moment(this.toDate, this.dateFormat).valueOf() : undefined;
        if (this.mode === 'range' && this.startDate && this.endDate) {
            const formattedFromDate = moment_min.moment(this.startDate).format(this.dateFormat);
            const formattedToDate = moment_min.moment(this.endDate).format(this.dateFormat);
            this.value = `${formattedFromDate} To ${formattedToDate}`;
        }
    }
    _getValidDateInMonth(date, args) {
        if (this.minDate !== undefined && this.maxDate !== undefined) {
            if (date < 0) {
                return -1;
            }
            const dateFormat = this.dateFormat || 'DD-MM-YYYY';
            const minDate = moment_min.moment(this.minDate, dateFormat);
            const maxDate = moment_min.moment(this.maxDate, dateFormat);
            const argDate = moment_min.moment([args.year, args.month, date + 1]);
            const isValid = minDate.valueOf() <= argDate.valueOf() && argDate.valueOf() <= maxDate.valueOf();
            return !isValid
                ? -1
                : date >= args.numberOfDays
                    ? 1
                    : 0;
        }
        return date < 0
            ? -1
            : date >= args.numberOfDays
                ? 1
                : 0;
    }
    handleDateHover(day) {
        if (this.startDate && !this.endDate) {
            if (this.startDate > day.timestamp) {
                this.endDate = this.startDate;
                this.startDate = undefined;
            }
            this.dateHovered = day.timestamp;
        }
        else if (!this.startDate && this.endDate) {
            if (this.endDate < day.timestamp) {
                this.startDate = this.endDate;
                this.endDate = undefined;
            }
            this.dateHovered = day.timestamp;
        }
    }
    isHoverInRange({ timestamp }) {
        const { startDate, endDate, dateHovered } = this;
        const startDateCondtion = startDate && dateHovered && (timestamp <= dateHovered && timestamp >= startDate);
        const endDateCondition = endDate && dateHovered && (timestamp >= dateHovered && timestamp <= endDate);
        return startDateCondtion || endDateCondition;
    }
    handleRangeSelection(timestamp) {
        if (this.startDate && this.endDate) {
            this.endDate = undefined;
            this.startDate = timestamp;
        }
        else if (this.startDate && !this.endDate) {
            if (timestamp > this.startDate) {
                this.endDate = timestamp;
            }
            else if (timestamp < this.startDate) {
                this.endDate = this.startDate;
                this.startDate = timestamp;
            }
        }
        else if (!this.startDate && this.endDate) {
            this.startDate = timestamp;
        }
        else if (!this.startDate && !this.endDate) {
            this.startDate = timestamp;
        }
    }
    getCellStyle(day) {
        let cellStyle = 'c-day-container';
        if (day.month !== 0) {
            cellStyle += ' disabled';
        }
        if (this.isCurrentDay(day)) {
            cellStyle += ' highlight';
        }
        if (this.isSelectedDay(day) || day.timestamp === this.dateHovered) {
            cellStyle += ' highlight-blue';
        }
        if (this.isInRange(day) || this.isHoverInRange(day)) {
            cellStyle += ' highlight-range';
        }
        if (day.timestamp === this.startDate) {
            cellStyle += ' start-day';
        }
        if (day.timestamp === this.endDate) {
            cellStyle += ' end-day';
        }
        if (this.startDate && this.dateHovered < this.startDate && day.timestamp === this.dateHovered) {
            cellStyle += ' start-day';
        }
        else if (this.endDate && this.dateHovered < this.endDate && day.timestamp === this.dateHovered) {
            cellStyle += ' start-day';
        }
        else if (this.startDate && this.dateHovered > this.startDate && day.timestamp === this.dateHovered) {
            cellStyle += ' end-day';
        }
        return cellStyle;
    }
    renderCalendar(monthDetails) {
        const days = monthDetails.map((day, index$1) => {
            return (index.h("div", { class: this.getCellStyle(day), key: index$1 }, index.h("div", { class: "cdc-day" }, index.h("span", { onClick: () => this.onDateClick(day), onMouseOver: () => this.handleDateHover(day) }, day.date))));
        });
        return (index.h("div", { class: "c-container" }, index.h("div", { class: "cc-head" }, weekDay.map((day, index$1) => index.h("div", { key: index$1, class: "cch-name" }, day))), index.h("div", { class: "cc-body" }, days)));
    }
    showSingleDatePicker() {
        return this.showDatePicker && this.mode !== 'range';
    }
    showDateRangePicker() {
        return this.showDatePicker && this.mode === 'range';
    }
    render() {
        const { host, name, value } = this;
        utils.renderHiddenField(host, name, value);
        return (index.h("div", null, index.h("div", { onClick: () => this.showDatePicker = false, class: this.showDatePicker ? 'overlay-show' : 'overlay-hide' }), index.h("fw-input", { value: this.value, class: "date-input", placeholder: this.placeholder, readonly: true }), this.showSingleDatePicker()
            ? (index.h("div", { class: "datepicker" }, index.h("div", { class: "mdp-container" }, index.h("div", { class: "mdpc-head" }, index.h("div", { class: "mdpch-button" }, index.h("div", { class: "mdpchb-inner", onClick: () => this.setMonth(-1) }, index.h("span", { class: "mdpchbi-left-arrow" }))), index.h("div", { class: "mdpch-container" }, index.h("span", { class: "mdpchc-month" }, index.h("fw-select", { class: "single-month-selector", readonly: true, value: monthArr[this.month] }, monthArr.map((month, i) => index.h("fw-select-option", { value: month, key: i, selected: month === monthArr[this.month] }, month)))), index.h("span", { class: "mdpchc-year" }, index.h("fw-select", { class: "single-year-selector", readonly: true, value: this.year }, this.supportedYears.map((year, i) => index.h("fw-select-option", { value: year, key: i, selected: year === this.year }, year))))), index.h("div", { class: "mdpch-button-right" }, index.h("div", { class: "mdpchb-inner", onClick: () => this.setMonth(1) }, index.h("span", { class: "mdpchbi-right-arrow" })))), index.h("div", { class: "mdpc-body" }, this.renderCalendar(this.monthDetails))), index.h("div", { class: "mdpc-footer" }, index.h("fw-button", { color: "primary", class: "update-date-value" }, "Update"), index.h("fw-button", { color: "secondary", class: "close-date-picker" }, "Cancel"))))
            : '', this.showDateRangePicker()
            ? (index.h("div", { class: "daterangepicker" }, index.h("div", { class: "mdp-range-container" }, index.h("div", { class: "mdpc-head" }, index.h("div", { class: "mdpch-button" }, index.h("div", { class: "mdpchb-inner", onClick: () => this.setMonth(-1) }, index.h("span", { class: "mdpchbi-left-arrow" }))), index.h("div", { class: "mdpch-container" }, index.h("span", { class: "mdpchc-month" }, index.h("fw-select", { class: "from-month-selector", readonly: true, value: monthArr[this.month] }, monthArr.map((month, i) => index.h("fw-select-option", { value: month, key: i, selected: month === monthArr[this.month] }, month)))), index.h("span", { class: "mdpchc-year" }, index.h("fw-select", { class: "from-year-selector", readonly: true, value: this.year }, this.supportedYears.map((year, i) => index.h("fw-select-option", { value: year, key: i, selected: year === this.year }, year))))), index.h("div", { class: "mdpch-button-right" }, index.h("div", { class: "mdpchb-inner", onClick: () => this.setMonth(1) }, index.h("span", { class: "mdpchbi-right-arrow" }))), index.h("div", { class: "mdpch-container-right" }, index.h("span", { class: "mdpchc-year" }, index.h("fw-select", { class: "to-year-selector", readonly: true, value: this.toYear }, this.supportedYears.map((year, i) => index.h("fw-select-option", { value: year, key: i, selected: year === this.toYear }, year)))), index.h("span", { class: "mdpchc-month" }, index.h("fw-select", { class: "to-month-selector", readonly: true, value: monthArr[this.toMonth] }, monthArr.map((month, i) => index.h("fw-select-option", { value: month, key: i, selected: month === monthArr[this.toMonth] }, month)))))), index.h("div", { class: "body-container" }, index.h("div", { class: "mdpc-body" }, this.renderCalendar(this.monthDetails)), index.h("div", { class: "mdpc-body mdpc-body-right" }, this.renderCalendar(this.nextMonthDetails)))), index.h("div", { class: "mdpc-range-footer" }, index.h("fw-button", { color: "primary", class: "update-range-value" }, "Update"), index.h("fw-button", { color: "secondary", class: "close-date-picker" }, "Cancel"))))
            : ''));
    }
    get host() { return index.getElement(this); }
    static get style() { return ":host {\n  --color-milk: #fff;\n  --color-elephant-900: #12344d;\n  --color-elephant-800: #264966;\n  --color-elephant-700: #345c7c;\n  --color-elephant-600: #447093;\n  --color-smoke-700: #475867;\n  --color-smoke-300: #92a2b1;\n  --color-smoke-100: #cfd7df;\n  --color-smoke-50: #ebeff3;\n  --color-smoke-25: #f3f5f7;\n  --color-jungle-800: #00795b;\n  --color-jungle-500: #00a886;\n  --color-jungle-100: #b4e5da;\n  --color-jungle-50: #e0f5f1;\n  --color-azure-800: #2c5cc5;\n  --color-azure-100: #bbdcfe;\n  --color-azure-50: #e5f2fd;\n  --color-persimmon-900: #c82124;\n  --color-persimmon-800: #d72d30;\n  --color-persimmon-100: #ffd0d6;\n  --color-persimmon-50: #ffecf0;\n  --color-casablanca-700: #e86f25;\n  --color-casablanca-300: #f8ab59;\n  --color-casablanca-100: #fedcb3;\n  --color-casablanca-50: #fef1e1;\n  --border-color: var(--color-smoke-100);\n  --border-success-color: var(--color-jungle-100);\n  --border-info-color: var(--color-azure-100);\n  --border-danger-color: var(--color-persimmon-100);\n  --border-warning-color: var(--color-casablanca-100);\n  --bg-dark: var(--color-elephant-900);\n  --bg-success: var(--color-jungle-50);\n  --bg-info: var(--color-azure-50);\n  --bg-danger: var(--color-persimmon-50);\n  --bg-warning: var(--color-casablanca-50);\n  --radius: 4px;\n  --radius-small: 2px;\n}\n\n:host {\n  --font-stack: -apple-system, blinkmacsystemfont, \"Segoe UI\", roboto, oxygen, ubuntu, cantarell, \"Open Sans\", \"Helvetica Neue\", sans-serif;\n  --font-weight-400: 400;\n  --font-weight-300: 300;\n  --font-weight-500: 500;\n  --font-weight-600: 600;\n  --font-weight-700: 700;\n  --font-size-10: 10px;\n  --font-size-12: 12px;\n  --font-size-14: 14px;\n  --font-size-16: 16px;\n  --font-size-18: 18px;\n  --font-size-20: 20px;\n  --font-size-24: 24px;\n  --text-default: var(--color-elephant-900);\n  --text-secondary: var(--color-smoke-700);\n  --text-disabled: var(--color-smoke-300);\n  --text-success: var(--color-jungle-800);\n  --text-info: var(--color-azure-800);\n  --text-danger: var(--color-persimmon-800);\n  --text-warning: var(--color-casablanca-700);\n  --text-link: var(--color-azure-800);\n  --icon-primary: var(--color-smoke-700);\n  --icon-primary-hover: var(--color-smoke-100);\n}\n\n:host {\n  font-family: var(--font-stack);\n  -webkit-font-smoothing: antialiased;\n  -moz-osx-font-smoothing: grayscale;\n  -webkit-box-sizing: border-box;\n  box-sizing: border-box;\n}\n\n:host {\n  position: relative;\n}\n:host * {\n  -webkit-user-select: none;\n     -moz-user-select: none;\n      -ms-user-select: none;\n          user-select: none;\n}\n\n.mdp-container,\n.mdp-range-container {\n  float: left;\n  position: absolute;\n  left: 0;\n  top: 40px;\n  width: 300px;\n  min-height: 336px;\n  background: var(--color-milk);\n  -webkit-box-shadow: 10px 10px 40px rgba(0, 0, 0, 0.2);\n  box-shadow: 10px 10px 40px rgba(0, 0, 0, 0.2);\n  border-radius: 5px;\n  overflow: hidden;\n  padding: 14px;\n  z-index: 1000;\n}\n\n.mdpc-footer,\n.mdpc-range-footer {\n  float: left;\n  position: absolute;\n  left: 0;\n  top: 344px;\n  width: 312px;\n  padding: 8px;\n  background: var(--color-smoke-25);\n  border-top: 1px solid var(--color-smoke-50);\n  margin-top: 15.5px;\n  -webkit-box-shadow: 4px 9px 20px rgba(0, 0, 0, 0.2);\n  box-shadow: 4px 9px 20px rgba(0, 0, 0, 0.2);\n  border-radius: 0 0 5px 5px;\n  z-index: 1000;\n}\n.mdpc-footer fw-button,\n.mdpc-range-footer fw-button {\n  float: right;\n  margin-left: 7px;\n}\n\n.mdpc-range-footer {\n  width: 612px;\n}\n\n.mdp-range-container {\n  width: 600px;\n}\n\n.mdpc-head {\n  height: 53px;\n}\n\n.mdpc-body {\n  float: left;\n  width: 100%;\n}\n\n.body-container {\n  margin: 0 10px;\n}\n.body-container .mdpc-body {\n  width: 48%;\n}\n.body-container .mdpc-body-right {\n  float: right;\n  border-left: 1px solid var(--color-smoke-50);\n  padding-left: 10px;\n}\n\n/**\n * Controls\n */\n.mdpch-button,\n.mdpch-button-right {\n  float: left;\n  width: 30px;\n  height: 100%;\n  -webkit-box-sizing: border-box;\n  box-sizing: border-box;\n  position: relative;\n}\n\n.mdpch-button-right {\n  float: right;\n}\n\n.mdpchb-inner:hover {\n  cursor: pointer;\n  background: var(--color-smoke-50);\n}\n\n.mdpchb-inner {\n  float: left;\n  height: 20px;\n  width: 20px;\n  border-radius: 100%;\n  line-height: 35px;\n  text-align: center;\n  position: absolute;\n  top: 45%;\n  left: 50%;\n  margin-left: -17px;\n  margin-top: -11px;\n}\n\n.mdpch-button-right .mdpchb-inner {\n  left: 75%;\n}\n\n.mdpchbi-left-arrow,\n.mdpchbi-left-arrows,\n.mdpchbi-left-arrows::after,\n.mdpchbi-right-arrow,\n.mdpchbi-right-arrows,\n.mdpchbi-right-arrows::after {\n  display: block;\n  float: left;\n  width: 6px;\n  height: 6px;\n  border-left: 2px solid var(--color-smoke-300);\n  border-bottom: 2px solid var(--color-smoke-300);\n  position: absolute;\n}\n\n.mdpchbi-left-arrow,\n.mdpchbi-left-arrows,\n.mdpchbi-right-arrow,\n.mdpchbi-right-arrows {\n  -webkit-transform: rotate(45deg);\n  transform: rotate(45deg);\n  left: 50%;\n  top: 50%;\n  margin-left: -2px;\n  margin-top: -4px;\n}\n\n.mdpchbi-right-arrow,\n.mdpchbi-right-arrows {\n  -webkit-transform: rotate(225deg);\n  transform: rotate(225deg);\n  margin-left: -4px;\n}\n\n.mdpchbi-left-arrows::after,\n.mdpchbi-right-arrows::after {\n  content: \"\";\n  left: 3px;\n  top: -5px;\n}\n\n.mdpchbi-left-arrows {\n  margin-left: -5px;\n}\n\n.mdpchbi-right-arrows {\n  margin-left: -2px;\n}\n\n.mdpch-container,\n.mdpch-container-right {\n  float: left;\n  width: 200px;\n  height: 100%;\n}\n\n.mdpch-container-right {\n  float: right;\n}\n.mdpch-container-right .mdpchc-month,\n.mdpch-container-right .mdpchc-year {\n  float: right;\n}\n\n.mdpchc-year {\n  float: left;\n  height: 30px;\n  font-size: 16px;\n  color: var(--color-smoke-700);\n  text-align: center;\n  width: 90px;\n}\n\n.mdpchc-month {\n  float: left;\n  font-size: 16px;\n  color: var(--color-smoke-700);\n  text-align: center;\n  margin-right: 10px;\n  width: 90px;\n}\n.mdpchc-month fw-select .input-container-inner input {\n  width: 98%;\n}\n\n/**\n *  Calendar\n */\n.c-container,\n.c-day-container,\n.cc-body,\n.cc-head,\n.cch-name,\n.cdc-day,\n.cdc-day span {\n  position: relative;\n  display: block;\n  float: left;\n  -webkit-box-sizing: border-box;\n  box-sizing: border-box;\n}\n\n.c-container {\n  width: 100%;\n  height: 100%;\n}\n\n.cc-head {\n  height: 30px;\n  width: 100%;\n}\n\n.cch-name {\n  width: 14.285%;\n  height: 30px;\n  line-height: 30px;\n  font-weight: 700;\n  color: var(--color-smoke-700);\n  font-size: 13px;\n  text-align: center;\n}\n\n.cc-body {\n  height: 210px;\n  width: 100%;\n}\n\n.c-day-container {\n  width: 14.285%;\n  height: 11.185%;\n  margin-top: 10px;\n}\n\n.cdc-day {\n  width: 100%;\n  height: 100%;\n  font-size: 12px;\n  font-weight: 300;\n  color: var(--color-smoke-700);\n  text-align: center;\n}\n\n.cdc-day span:hover {\n  cursor: pointer;\n  background: var(--color-smoke-50);\n}\n\n.cdc-day span {\n  width: 24px;\n  height: 22px;\n  font-size: 13px;\n  margin-top: -11px;\n  margin-left: -13px;\n  left: 50%;\n  top: 50%;\n  font-weight: 400;\n  border-radius: 10%;\n  line-height: 20px;\n  color: #12344d;\n}\n\n.c-day-container.disabled {\n  pointer-events: none;\n}\n\n.date-input {\n  width: 210px;\n}\n\n.c-day-container.disabled .cdc-day span {\n  color: var(--color-smoke-100);\n  background: var(--color-milk) !important;\n}\n\n.c-day-container.highlight:not(.disabled) .cdc-day span {\n  border: 1px solid var(--color-smoke-100);\n}\n\n.c-day-container.highlight-blue:not(.disabled) .cdc-day span {\n  background: var(--color-azure-800);\n  color: var(--color-smoke-25);\n}\n\n.c-day-container.start-day .cdc-day {\n  margin-left: 10px;\n}\n.c-day-container.start-day .cdc-day span {\n  -webkit-transform: translateX(-10px);\n  transform: translateX(-10px);\n}\n\n.c-day-container.end-day .cdc-day {\n  margin-left: -10px;\n}\n.c-day-container.end-day .cdc-day span {\n  -webkit-transform: translateX(10px);\n  transform: translateX(10px);\n}\n\n.c-day-container.highlight-range:not(.disabled) .cdc-day {\n  background: #e4f2fd;\n}\n\n.overlay-show {\n  background: transparent;\n  width: 100vw;\n  height: 100vh;\n  position: fixed;\n  top: 0;\n  left: 0;\n}\n\n.overlay-hide {\n  display: none;\n}"; }
};

exports.fw_datepicker = Datepicker;
