// @freshworks/crayons/dist/cjs/fw-dropdown-button.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-19272a91.js');

const DropdownButton = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        /**
         * Dropdown Button color
         */
        this.color = 'primary';
        /**
         * Displays a split dropdown button
         */
        this.split = false;
        /**
         * Disables the dropdown button if its true
         */
        this.disabled = false;
        /**
         * Displays a searchable dropdown button
         */
        this.searchable = false;
        /**
         * Value of the dropdown button
         */
        this.value = undefined;
        /**
         * Placeholder text for search input. Validated only if dropdown and searchable is true
         */
        this.placeholder = '';
        /**
         * Options to show in the dropdown button
         */
        this.options = [];
        /**
         * Option input for searching through options
         */
        this.optionInput = '';
        /**
         * Filtered options if the dropdown is searchable
         */
        this.filteredOptions = [];
        /**
         * State representing dropdown is open or not
         */
        this.isDropdownOpen = false;
        this.fwOptionClick = index.createEvent(this, "fwOptionClick", 7);
        this.fwOptionsAdd = index.createEvent(this, "fwOptionsAdd", 7);
    }
    setInitialState(newVal) {
        // Do not open dropdown if options is absent
        if (newVal) {
            if (this.options.length === 0) {
                this.isDropdownOpen = false;
            }
        }
        // Reset searchable dropdown to initial state on close
        if (!newVal) {
            if (this.searchable) {
                this.filteredOptions = this.options;
                this.dropdownInput.value = '';
            }
        }
    }
    /**
     * Toggles dropdown b/w open and close
     */
    handleDropdownToggle() {
        if (!this.disabled) {
            this.isDropdownOpen = !this.isDropdownOpen;
        }
    }
    /**
     * Sets dropdown options from slot
     */
    setDropdownOptions() {
        // Set options if the prop is not supplied by user
        if (this.options.length === 0) {
            this.options = Array.from(this.host.querySelectorAll('[slot=dropdown-options] option')).map(option => ({
                id: option.id,
                value: option.getAttribute('value'),
                label: option.textContent,
            }));
        }
        // Set filtered options to inital options for search dropdown
        if (this.searchable) {
            this.filteredOptions = this.options;
        }
    }
    /**
     * Sets search dropdown input
     */
    setSearchInput() {
        this.optionInput = this.dropdownInput.value;
        this.filteredOptions = this.optionInput !== ''
            ? this.options.filter(option => option.label.toLowerCase()
                .includes(this.optionInput.toLowerCase()))
            : this.options;
    }
    /**
     * Listens to checkbox change for search dropdown
     */
    handleCheckboxChange(ev) {
        if (ev.target.checked) {
            this.value = [
                ...(this.value || []),
                this.options.find(({ value }) => value === ev.target.id).value,
            ];
        }
        else {
            this.value = this.value.filter(val => val !== ev.target.id);
        }
    }
    /**
     * Handles `Add` button click in search dropdown
     */
    handleAddClick() {
        this.fwOptionsAdd.emit({ value: this.value });
        this.handleDropdownToggle();
    }
    /**
     * Handles option select from the dropdown
     */
    handleOptionClick(option) {
        if (!this.searchable) {
            this.value = option.value;
            this.fwOptionClick.emit({ value: this.value, label: option.label });
            this.handleDropdownToggle();
        }
    }
    render() {
        const ChevronArrow = () => {
            const iconColor = ['secondary', 'link', 'text'].includes(this.color) ? '#12344d' : '#fff';
            const iconSize = 8;
            const direction = this.isDropdownOpen ? 'up' : 'down';
            return index.h("fw-icon", { name: `chevron-${direction}`, color: iconColor, size: iconSize }, " ");
        };
        const DropdownMenu = () => {
            const SearchInput = () => {
                return (index.h("fw-input", { placeholder: this.placeholder, "icon-left": "search", ref: dropdownInput => this.dropdownInput = dropdownInput, onInput: () => this.setSearchInput() }));
            };
            const renderBtnGroup = () => {
                return (index.h("div", { class: "search-btn-grp" }, index.h("fw-button", { id: "addBtn", size: "small", color: "primary", onClick: () => this.handleAddClick() }, " Add "), index.h("fw-button", { id: "cancelBtn", size: "small", color: "secondary", onClick: () => this.handleDropdownToggle() }, " Cancel ")));
            };
            const validOptions = this.searchable ? this.filteredOptions : this.options;
            return (index.h("ul", { class: `dropdown-menu ${this.isDropdownOpen ? 'dropdown-menu--open' : ''}` }, this.searchable ? index.h(SearchInput, null) : '', index.h("div", { class: this.searchable ? `search-list` : '' }, validOptions.map(option => {
                const liEl = index.h("li", { key: option.id || option.value, onClick: () => this.handleOptionClick(option), class: "dropdown-item" }, " ", option.label, " ");
                const checkboxEl = index.h("fw-checkbox", { id: option.value, checked: (this.value || []).includes(option.value), onFwChange: e => this.handleCheckboxChange(e) });
                return this.searchable
                    ? index.h("div", { class: "searchable-item" }, " ", checkboxEl, " ", liEl, " ")
                    : liEl;
            })), this.searchable ? renderBtnGroup() : ''));
        };
        const DropdownState = () => (!this.split ? index.h("span", { class: "down-arrow" }, index.h(ChevronArrow, null)) : '');
        const SplitDropdownState = () => {
            const dropdownStateClasses = `
        dropdown-state
        ${'dropdown-state--' + this.color.toLowerCase()}
        ${this.disabled ? 'disabled' : ''}
      `;
            return (this.split ?
                index.h("div", { onClick: () => this.handleDropdownToggle(), class: dropdownStateClasses }, index.h("div", { class: "state-icon" }, index.h(ChevronArrow, null))) : '');
        };
        return (index.h("div", { class: "dropdown-container" }, index.h("div", { class: "btn-container" }, index.h("fw-button", { color: this.color, disabled: this.disabled, onClick: () => this.handleDropdownToggle() }, this.label, index.h(DropdownState, null)), index.h(SplitDropdownState, null)), index.h(DropdownMenu, null)));
    }
    componentWillLoad() {
        document.addEventListener('click', e => {
            const { target } = e;
            const canClose = this.isDropdownOpen
                && this.host !== target
                && !this.host.contains(target);
            if (canClose) {
                this.isDropdownOpen = false;
            }
        });
        this.setDropdownOptions();
    }
    get host() { return index.getElement(this); }
    static get watchers() { return {
        "isDropdownOpen": ["setInitialState"]
    }; }
    static get style() { return ":host {\n  --color-milk: #fff;\n  --color-elephant-900: #12344d;\n  --color-elephant-800: #264966;\n  --color-elephant-700: #345c7c;\n  --color-elephant-600: #447093;\n  --color-smoke-700: #475867;\n  --color-smoke-300: #92a2b1;\n  --color-smoke-100: #cfd7df;\n  --color-smoke-50: #ebeff3;\n  --color-smoke-25: #f3f5f7;\n  --color-jungle-800: #00795b;\n  --color-jungle-500: #00a886;\n  --color-jungle-100: #b4e5da;\n  --color-jungle-50: #e0f5f1;\n  --color-azure-800: #2c5cc5;\n  --color-azure-100: #bbdcfe;\n  --color-azure-50: #e5f2fd;\n  --color-persimmon-900: #c82124;\n  --color-persimmon-800: #d72d30;\n  --color-persimmon-100: #ffd0d6;\n  --color-persimmon-50: #ffecf0;\n  --color-casablanca-700: #e86f25;\n  --color-casablanca-300: #f8ab59;\n  --color-casablanca-100: #fedcb3;\n  --color-casablanca-50: #fef1e1;\n  --border-color: var(--color-smoke-100);\n  --border-success-color: var(--color-jungle-100);\n  --border-info-color: var(--color-azure-100);\n  --border-danger-color: var(--color-persimmon-100);\n  --border-warning-color: var(--color-casablanca-100);\n  --bg-dark: var(--color-elephant-900);\n  --bg-success: var(--color-jungle-50);\n  --bg-info: var(--color-azure-50);\n  --bg-danger: var(--color-persimmon-50);\n  --bg-warning: var(--color-casablanca-50);\n  --radius: 4px;\n  --radius-small: 2px;\n}\n\n:host {\n  --font-stack: -apple-system, blinkmacsystemfont, \"Segoe UI\", roboto, oxygen, ubuntu, cantarell, \"Open Sans\", \"Helvetica Neue\", sans-serif;\n  --font-weight-400: 400;\n  --font-weight-300: 300;\n  --font-weight-500: 500;\n  --font-weight-600: 600;\n  --font-weight-700: 700;\n  --font-size-10: 10px;\n  --font-size-12: 12px;\n  --font-size-14: 14px;\n  --font-size-16: 16px;\n  --font-size-18: 18px;\n  --font-size-20: 20px;\n  --font-size-24: 24px;\n  --text-default: var(--color-elephant-900);\n  --text-secondary: var(--color-smoke-700);\n  --text-disabled: var(--color-smoke-300);\n  --text-success: var(--color-jungle-800);\n  --text-info: var(--color-azure-800);\n  --text-danger: var(--color-persimmon-800);\n  --text-warning: var(--color-casablanca-700);\n  --text-link: var(--color-azure-800);\n  --icon-primary: var(--color-smoke-700);\n  --icon-primary-hover: var(--color-smoke-100);\n}\n\n:host {\n  font-family: var(--font-stack);\n  -webkit-font-smoothing: antialiased;\n  -moz-osx-font-smoothing: grayscale;\n  -webkit-box-sizing: border-box;\n  box-sizing: border-box;\n}\n\n/**\n  \@prop --dropdown-min-height: Dropdown minimum height in pixels\n  \@prop --dropdown-min-width: Dropdown minimum width in pixels\n  \@prop --dropdown-max-height: Dropdown maximum height in pixels\n  \@prop --dropdown-max-width: Dropdown maximum width in pixels\n  \@prop --dropdown-font-size: Dropdown item font size in pixels\n*/\n:host {\n  --btn-primary-color: var(--color-milk);\n  --btn-primary-bg: var(--color-elephant-800);\n  --btn-primary-bg-dark: var(--color-elephant-900);\n  --btn-primary-border: var(--color-elephant-900);\n  --btn-primary-border-active: var(--color-elephant-800);\n  --btn-secondary-color: var(--text-default);\n  --btn-secondary-bg: var(--color-milk);\n  --btn-secondary-bg-dark: var(--color-smoke-25);\n  --btn-secondary-border: var(--border-color);\n  --btn-secondary-border-active: var(--color-smoke-50);\n  --btn-danger-color: var(--color-milk);\n  --btn-danger-bg: var(--color-persimmon-800);\n  --btn-danger-bg-dark: var(--color-persimmon-900);\n  --btn-danger-border: var(--color-persimmon-900);\n  --btn-link-color: var(--color-azure-800);\n  --btn-link-bg: var(--color-smoke-50);\n  --bg-color-hover: var(--color-smoke-50);\n  --active-box-shadow: inset 0 0 4px 0 rgba(0, 0, 0, 0.25);\n  --dropdown-min-height: 100px;\n  --dropdown-min-width: 200px;\n  --dropdown-max-height: 200px;\n  --dropdown-max-width: 480px;\n  --dropdown-font-size: var(--font-size-14);\n}\n\n\@-webkit-keyframes dropdown-fade {\n  0% {\n    opacity: 0;\n    -webkit-transform: translateY(-10%);\n    transform: translateY(-10%);\n  }\n  50% {\n    opacity: 0.7;\n    -webkit-transform: translateY(-5%);\n    transform: translateY(-5%);\n  }\n  100% {\n    opacity: 1;\n    -webkit-transform: translateY(0%);\n    transform: translateY(0%);\n  }\n}\n\n\@keyframes dropdown-fade {\n  0% {\n    opacity: 0;\n    -webkit-transform: translateY(-10%);\n    transform: translateY(-10%);\n  }\n  50% {\n    opacity: 0.7;\n    -webkit-transform: translateY(-5%);\n    transform: translateY(-5%);\n  }\n  100% {\n    opacity: 1;\n    -webkit-transform: translateY(0%);\n    transform: translateY(0%);\n  }\n}\n.dropdown-container {\n  position: relative;\n  display: inline-block;\n}\n.dropdown-container .btn-container {\n  display: -ms-flexbox;\n  display: flex;\n}\n.dropdown-container .btn-container .dropdown-state {\n  cursor: pointer;\n  border-top-left-radius: 0;\n  border-bottom-left-radius: 0;\n  border-top-right-radius: 4px;\n  border-bottom-right-radius: 4px;\n  width: 30px;\n  height: 30px;\n  margin-left: -10px;\n}\n.dropdown-container .btn-container .dropdown-state .state-icon {\n  padding: 5px 10px;\n}\n.dropdown-container .btn-container .dropdown-state--primary {\n  background-color: var(--btn-primary-bg);\n  color: var(--btn-primary-color);\n  border: 1px solid var(--btn-primary-border);\n  background-image: -webkit-gradient(linear, left top, left bottom, from(var(--btn-primary-bg)), to(var(--btn-primary-bg-dark)));\n  background-image: linear-gradient(to bottom, var(--btn-primary-bg), var(--btn-primary-bg-dark));\n}\n.dropdown-container .btn-container .dropdown-state--primary:active {\n  border: 1px solid var(--btn-primary-border-active);\n}\n.dropdown-container .btn-container .dropdown-state--primary:hover:not([disabled]) {\n  background-color: var(--btn-primary-bg-dark);\n  background-image: none;\n}\n.dropdown-container .btn-container .dropdown-state--secondary {\n  background-color: var(--btn-secondary-bg-dark);\n  color: var(--btn-secondary-color);\n  border: 1px solid var(--btn-secondary-border);\n  background-image: -webkit-gradient(linear, left top, left bottom, from(var(--btn-secondary-bg)), to(var(--btn-secondary-bg-dark)));\n  background-image: linear-gradient(to bottom, var(--btn-secondary-bg), var(--btn-secondary-bg-dark));\n}\n.dropdown-container .btn-container .dropdown-state--secondary:active {\n  border: 1px solid var(--btn-secondary-border-active);\n}\n.dropdown-container .btn-container .dropdown-state--secondary:hover:not([disabled]) {\n  background-color: var(--btn-secondary-bg-dark);\n  background-image: none;\n}\n.dropdown-container .btn-container .dropdown-state--danger {\n  color: var(--btn-danger-color);\n  background-color: var(--btn-danger-bg);\n  border: 1px solid var(--btn-danger-border);\n  background-image: -webkit-gradient(linear, left top, left bottom, from(var(--btn-danger-bg)), to(var(--btn-danger-bg-dark)));\n  background-image: linear-gradient(to bottom, var(--btn-danger-bg), var(--btn-danger-bg-dark));\n}\n.dropdown-container .btn-container .dropdown-state--danger:hover:not([disabled]) {\n  background-color: var(--btn-danger-bg-dark);\n  background-image: none;\n}\n.dropdown-container .btn-container .dropdown-state.disabled {\n  cursor: not-allowed;\n  opacity: 0.4;\n}\n.dropdown-container .btn-container .dropdown-state:focus {\n  border: 1px solid var(--color-azure-800);\n  -webkit-box-shadow: 0 0 0 1px var(--color-azure-800);\n  box-shadow: 0 0 0 1px var(--color-azure-800);\n  outline: none;\n}\n.dropdown-container .btn-container .dropdown-state:active {\n  -webkit-box-shadow: var(--active-box-shadow);\n  box-shadow: var(--active-box-shadow);\n}\n.dropdown-container .btn-container .down-arrow {\n  padding-left: 10px;\n}\n.dropdown-container .dropdown-menu {\n  background-color: #fff;\n  min-width: var(--dropdown-min-width);\n  max-width: var(--dropdown-max-width);\n  min-height: var(--dropdown-min-height);\n  max-height: var(--dropdown-max-height);\n  padding: 10px;\n  border-radius: 5px;\n  position: absolute;\n  top: 17px;\n  -webkit-box-shadow: rgba(0, 0, 0, 0.176) 0 6px 12px 0;\n  box-shadow: rgba(0, 0, 0, 0.176) 0 6px 12px 0;\n  -webkit-box-sizing: border-box;\n  box-sizing: border-box;\n  overflow-y: scroll;\n  list-style: none;\n  border: 1px solid rgba(0, 0, 0, 0.15);\n  -webkit-animation: dropdown-fade 0.15s;\n          animation: dropdown-fade 0.15s;\n  display: none;\n  z-index: 999;\n}\n.dropdown-container .dropdown-menu--open {\n  display: block;\n}\n.dropdown-container .dropdown-menu--search {\n  border-bottom-left-radius: 0;\n  border-bottom-right-radius: 0;\n}\n.dropdown-container .dropdown-menu .search-list {\n  overflow-y: scroll;\n  max-height: calc(var(--dropdown-max-height) - 105px);\n}\n.dropdown-container .dropdown-menu .dropdown-item {\n  padding: 5px 25px 8px 10px;\n  color: var(--btn-secondary-color);\n  font-size: var(--dropdown-font-size);\n  font-weight: var(--font-weight-500);\n  cursor: pointer;\n  background: transparent;\n  border-radius: 5px;\n}\n.dropdown-container .dropdown-menu .dropdown-item:hover {\n  background: var(--bg-color-hover);\n}\n.dropdown-container .dropdown-menu .searchable-item {\n  margin-left: 5px;\n  display: -ms-flexbox;\n  display: flex;\n}\n.dropdown-container .dropdown-menu .searchable-item .dropdown-item {\n  cursor: default;\n  padding-top: 0;\n  padding-bottom: 10px;\n  padding-left: 0;\n}\n.dropdown-container .dropdown-menu .searchable-item .dropdown-item:hover {\n  background-color: transparent;\n}\n.dropdown-container .dropdown-menu .search-btn-grp {\n  margin-top: 5px;\n  display: -ms-flexbox;\n  display: flex;\n}\n.dropdown-container .dropdown-menu .search-btn-grp fw-button {\n  margin-right: 5px;\n  cursor: pointer;\n}"; }
};

exports.fw_dropdown_button = DropdownButton;
