// @freshworks/crayons/dist/cjs/fw-radio.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-19272a91.js');

const Radio = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        /**
         * Sets the state to selected. If the attribute’s value is undefined, the value is set to false.
         */
        this.checked = false;
        /**
         * Disables the component on the interface. If the attribute’s value is undefined, the value is set to false.
         */
        this.disabled = false;
        /**
         * Label displayed on the interface, for the component.
         */
        this.label = '';
        /**
         * Identifier corresponding to the component, that is saved when the form data is saved.
         */
        this.value = '';
        /**
         * Name of the component, saved as part of form data.
         */
        this.name = '';
        this.fwSelect = index.createEvent(this, "fwSelect", 7);
        this.fwDeselect = index.createEvent(this, "fwDeselect", 7);
        this.fwFocus = index.createEvent(this, "fwFocus", 7);
        this.fwBlur = index.createEvent(this, "fwBlur", 7);
    }
    componentDidLoad() {
        this.radio.checked = this.checked;
        this.radio.disabled = this.disabled;
    }
    checkChanged(isChecked) {
        if (!this.disabled) {
            if (isChecked) {
                this.fwSelect.emit({
                    value: this.value,
                    checked: true,
                });
            }
            else {
                this.fwDeselect.emit();
            }
        }
        this.radio.checked = isChecked;
    }
    disabledChanged(isDisabled) {
        this.radio.disabled = isDisabled;
    }
    onFocus() {
        this.fwFocus.emit();
    }
    onBlur() {
        this.fwBlur.emit();
    }
    toggle() {
        if (!this.disabled) {
            this.checked = !this.checked;
        }
    }
    render() {
        return (index.h(index.Host, { class: "radio-container", onClick: () => this.toggle(), role: "radio", tabIndex: "0", "aria-disabled": this.disabled ? 'true' : 'false', "aria-checked": `${this.checked}`, onFocus: () => this.onFocus(), onBlur: () => this.onBlur() }, index.h("input", { type: "radio", ref: el => this.radio = el }), index.h("label", null, index.h("span", { class: "text" }, index.h("slot", null)), index.h("br", null), this.label !== ''
            ? index.h("span", { class: "label-field" }, this.label)
            : '')));
    }
    static get watchers() { return {
        "checked": ["checkChanged"],
        "disabled": ["disabledChanged"]
    }; }
    static get style() { return "/* Need to check with designer */\n/* Need to check with designer */\n:host {\n  font-family: -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, Oxygen, Ubuntu, Cantarell, \"Open Sans\", \"Helvetica Neue\", sans-serif;\n}\n\n:host(.radio-container) {\n  display: inline;\n  position: relative;\n}\n\n/* Focus event occurs on the root element */\n:host(:focus) {\n  outline: none;\n}\n:host(:focus) input[type=radio] + label::before {\n  border: 1px solid transparent;\n  -webkit-box-shadow: 0 0 0 2px #2c5cc5;\n  box-shadow: 0 0 0 2px #2c5cc5;\n  border-color: #081824;\n}\n:host(:focus) input[type=radio][disabled] + label::before {\n  -webkit-box-shadow: none;\n  box-shadow: none;\n  border: 1px solid #dadfe3;\n}\n\n.label-field {\n  font-size: 12px;\n  color: #475867;\n  letter-spacing: 0;\n  line-height: 20px;\n  position: relative;\n  padding-left: 28px;\n  font-weight: 400;\n}\n\n/* Hover event occurs on the root element */\n:host(:hover) input[type=radio] + label::before {\n  -webkit-box-shadow: 0 0 0 5px #ebeff3;\n  box-shadow: 0 0 0 5px #ebeff3;\n  border-color: #081824;\n}\n:host(:hover) input[type=radio][disabled] + label {\n  cursor: not-allowed;\n}\n:host(:hover) input[type=radio][disabled] + label::before {\n  -webkit-box-shadow: none;\n  box-shadow: none;\n  border: 1px solid #dadfe3;\n}\n\ninput[type=radio] {\n  display: none;\n}\ninput[type=radio] + label {\n  -webkit-user-select: none;\n     -moz-user-select: none;\n      -ms-user-select: none;\n          user-select: none;\n  cursor: pointer;\n  max-width: auto;\n  margin-bottom: 0;\n  vertical-align: middle;\n  font-size: 12px;\n  color: #12344d;\n  font-weight: 500;\n}\ninput[type=radio] + label .text {\n  padding-left: 27px;\n}\ninput[type=radio] + label::before, input[type=radio] + label::after {\n  content: \"\";\n  display: block;\n  position: absolute;\n  top: 0;\n  -webkit-box-sizing: border-box;\n  box-sizing: border-box;\n}\ninput[type=radio] + label::before {\n  left: 0;\n  border: 1px solid #cfd7df;\n  border-radius: 50px;\n  top: 4px;\n  width: 14px;\n  height: 14px;\n  background: #fff;\n  -webkit-transition: all 0.3s ease;\n  transition: all 0.3s ease;\n}\ninput[type=radio] + label::after {\n  left: 3px;\n  border-radius: 100%;\n  width: 8px;\n  height: 8px;\n  opacity: 0;\n  -webkit-transition: opacity 0.2s ease-in-out;\n  transition: opacity 0.2s ease-in-out;\n  -webkit-box-sizing: border-box;\n  box-sizing: border-box;\n}\ninput[type=radio]:checked + label::before {\n  background: #fff;\n  border-color: #2c5cc5;\n  -webkit-box-shadow: none;\n  box-shadow: none;\n}\ninput[type=radio]:checked + label::after {\n  border-radius: 50%;\n  background-color: #2c5cc5;\n  opacity: 1;\n  top: 7px;\n}\ninput[type=radio]:checked:focus + label::before {\n  border-color: #3868d3;\n  -webkit-box-shadow: 0 0 4px 1px rgba(44, 92, 197, 0.6);\n  box-shadow: 0 0 4px 1px rgba(44, 92, 197, 0.6);\n}\ninput[type=radio]:checked:focus + label::after {\n  background-color: #3868d3;\n}\ninput[type=radio][disabled] + label {\n  color: #92a2b1;\n}\ninput[type=radio][disabled] + label .label-field {\n  color: #92a2b1;\n}\ninput[type=radio][disabled] + label::before {\n  border-color: #dadfe3;\n  background-color: #ebeff3;\n}\ninput[type=radio][disabled] + label::after {\n  border-color: #ebeff3;\n  background-color: #dadfe3;\n}\ninput[type=radio][disabled]:checked + label {\n  color: #92a2b1;\n}"; }
};

exports.fw_radio = Radio;
