// @freshworks/crayons/dist/cjs/fw-radio-group.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-19272a91.js');
const utils = require('./utils-a88d9e0c.js');

const RadioGroup = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        /**
         * If true, a radio group can be saved without selecting any option. If an option is selected, the selection can be cleared. If the attribute’s value is undefined, the value is set to false.
         */
        this.allowEmpty = false;
        /**
         * Name of the component, saved as part of form data.
         */
        this.name = '';
        this.onSelect = (ev) => {
            const selectedRadio = ev.target;
            if (selectedRadio) {
                this.value = selectedRadio.value;
            }
        };
        this.onDeselect = async (ev) => {
            const selectedRadio = ev.target;
            if (this.allowEmpty && selectedRadio.value === this.value) {
                this.value = undefined;
            }
            await this.updateRadios();
        };
        this.fwChange = index.createEvent(this, "fwChange", 7);
    }
    async valueChanged(value) {
        await this.updateRadios();
        this.fwChange.emit({ value });
    }
    async connectedCallback() {
        const el = this.host;
        if (this.value === undefined) {
            const radio = utils.findCheckedOption(el, 'fw-radio');
            if (radio !== undefined) {
                await radio.componentOnReady();
                if (this.value === undefined) {
                    this.value = radio.value;
                }
            }
        }
        this.mutationO = utils.watchForOptions(el, 'fw-radio', async (newOption) => {
            if (newOption !== undefined) {
                newOption.componentOnReady().then(() => {
                    this.value = newOption.value;
                }).catch();
            }
            else {
                await this.updateRadios();
            }
        });
        await this.updateRadios();
    }
    disconnectedCallback() {
        if (this.mutationO) {
            this.mutationO.disconnect();
            this.mutationO = undefined;
        }
    }
    async updateRadios() {
        /**
         * Make sure we get all radios first
         * so values are up to date prior
         * to caching the radio group value
         */
        const radios = await this.getRadios();
        const { value } = this;
        let hasChecked = false;
        // Walk the DOM in reverse order, since the last selected one wins!
        for (const radio of radios) {
            if (!hasChecked && radio.value === value) {
                // correct value for this radio
                // but this radio isn't checked yet
                // and we haven't found a checked yet
                hasChecked = true;
                radio.checked = true;
            }
            else {
                // this radio doesn't have the correct value
                // or the radio group has been already checked
                radio.checked = false;
            }
        }
        // Reset value if
        if (!hasChecked) {
            this.value = undefined;
        }
    }
    getRadios() {
        return Promise.all(Array
            .from(this.host.querySelectorAll('fw-radio'))
            .map(r => r.componentOnReady()));
    }
    render() {
        const { host, name, value } = this;
        utils.renderHiddenField(host, name, value);
        return (index.h(index.Host, { role: "radiogroup", onFwSelect: this.onSelect, onFwDeselect: this.onDeselect }));
    }
    get host() { return index.getElement(this); }
    static get watchers() { return {
        "value": ["valueChanged"]
    }; }
};

exports.fw_radio_group = RadioGroup;
