// @freshworks/crayons/dist/cjs/fw-select.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-19272a91.js');
const utils = require('./utils-a88d9e0c.js');

const Select = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        /**
         * If the dropdown is shown or not
         */
        this.isExpanded = false;
        this.options = [];
        this.filteredOptions = [];
        this.hasFocus = false;
        this.didInit = false;
        /**
         * Label displayed on the interface, for the component.
         */
        this.label = '';
        /**
         * Name of the component, saved as part of form data.
         */
        this.name = '';
        /**
         * Type of option accepted as the input value. If a user tries to enter an option other than the specified type, the list is not populated.
         */
        this.type = 'text';
        /**
         * Theme based on which the list box is styled.
         */
        this.state = 'normal';
        /**
         * Descriptive or instructional text displayed below the list box.
         */
        this.stateText = '';
        /**
         * If true, the user cannot modify the default value selected. If the attribute's value is undefined, the value is set to true.
         */
        this.readonly = false;
        /**
         * If true, the select component is auto focused on the page
         */
        this.autofocus = false;
        /**
         * Specifies the select field as a mandatory field and displays an asterisk next to the label. If the attribute’s value is undefined, the value is set to false.
         */
        this.required = false;
        /**
         * If true, the user must select a value. The default value is not displayed.
         */
        this.forceSelect = true;
        /**
         * Disables the component on the interface. If the attribute’s value is undefined, the value is set to false.
         */
        this.disabled = false;
        /**
         * Enables selection of multiple options. If the attribute’s value is undefined, the value is set to false.
         */
        this.multiple = false;
        /**
         * Works with `multiple` enabled. Configures the maximum number of options that can be selected with a multi-select component.
         */
        this.max = Number.MAX_VALUE;
        this.changeEmittable = () => !this.disabled;
        this.closeDropdown = () => {
            this.selectList.style.display = 'none';
            this.overlay.style.display = 'none';
            this.isExpanded = false;
        };
        this.innerOnFocus = (e) => {
            if (this.changeEmittable()) {
                this.hasFocus = true;
                this.fwFocus.emit(e);
            }
        };
        this.innerOnClick = () => {
            if (this.changeEmittable()) {
                this.filteredOptions = this.options;
                this.selectList.style.display = 'block';
                this.selectList.style.width = String(this.select.clientWidth) + 'px';
                this.isExpanded = true;
                this.overlay.style.display = 'block';
            }
        };
        this.innerOnBlur = (e) => {
            if (this.changeEmittable()) {
                this.closeDropdown();
                this.hasFocus = false;
                this.fwBlur.emit(e);
            }
        };
        this.fwChange = index.createEvent(this, "fwChange", 7);
        this.fwFocus = index.createEvent(this, "fwFocus", 7);
        this.fwBlur = index.createEvent(this, "fwBlur", 7);
    }
    keyChanged(newValue, oldValue) {
        if (JSON.stringify(newValue) !== JSON.stringify(oldValue)) {
            this.options = this.options.map(option => {
                option.selected = Array.isArray(this.value)
                    ? this.value.includes(option.value)
                    : this.value === option.value;
                return option;
            });
            if (this.didInit) {
                this.fwChange.emit({ value: this.value });
            }
        }
    }
    fwSelectedHandler(selectedItem) {
        this.options = this.options.map(option => {
            if (selectedItem.detail.value === option.value) {
                option.selected = selectedItem.detail.selected;
            }
            else if (!this.multiple) {
                option.selected = false;
            }
            return option;
        });
        this.selectInput.value = '';
        this.renderInput();
        this.closeDropdown();
        selectedItem.stopPropagation();
    }
    optionsChangedHandler() {
        this.renderInput();
    }
    fwCloseHandler(ev) {
        this.options = this.options.map(option => {
            if (option.value === ev.detail.value) {
                option.selected = false;
            }
            return option;
        });
    }
    onKeyDonw(ev) {
        switch (ev.key) {
            case 'ArrowDown':
                this.innerOnClick();
                break;
            case 'Escape':
                this.innerOnBlur(ev);
                break;
        }
    }
    onInput() {
        const value = this.selectInput.value.toLowerCase();
        this.filteredOptions = value !== ''
            ? this.options.filter(option => option.text.toLowerCase().startsWith(value))
            : this.options;
        this.renderInput();
    }
    renderTags() {
        if (this.multiple) {
            return this.options
                .filter(option => option.selected)
                .map(option => index.h("fw-tag", { text: option.text, disabled: option.disabled, value: option.value }));
        }
    }
    renderInput() {
        const selectedOptions = this.options.filter(option => option.selected);
        if (selectedOptions.length > 0) {
            this.value = this.multiple ? selectedOptions.map(option => option.value) : selectedOptions[0].value || '';
            if (this.selectInput) {
                this.selectInput.value = this.multiple ? this.selectInput.value : selectedOptions[0].text || '';
            }
        }
        else if (selectedOptions.length === 0) {
            this.value = undefined;
        }
    }
    renderDropdown() {
        return this.filteredOptions.map(option => {
            var _a;
            return (index.h("fw-select-option", { value: option.value, selected: option.selected, disabled: option.disabled || (((_a = this.value) === null || _a === void 0 ? void 0 : _a.length) >= this.max), html: option.isHtml, htmlContent: option.htmlContent }, option.text));
        });
    }
    componentWillLoad() {
        const selectOptions = Array.from(this.host.querySelectorAll('fw-select-option'));
        const options = selectOptions.map(option => {
            return {
                isHtml: option.html,
                text: option.html ? option.optionText : option.textContent,
                value: option.value,
                selected: option.value === this.value || option.selected,
                disabled: option.disabled || this.disabled,
                htmlContent: option.html ? option.innerHTML : '',
            };
        });
        this.options = options;
        this.filteredOptions = this.options;
        this.host.innerHTML = '';
    }
    componentDidLoad() {
        this.renderInput();
        this.didInit = true;
    }
    async getSelectedItem() {
        return this.options.filter(option => option.selected);
    }
    async setSelectedValues(values) {
        if (this.multiple) {
            this.options.forEach(option => {
                option.selected = values.includes(option.value);
            });
            this.renderInput();
        }
    }
    render() {
        const { host, name, value } = this;
        utils.renderHiddenField(host, name, value);
        return (index.h(index.Host, { "aria-disabled": this.disabled, class: {
                'has-focus': this.hasFocus,
            } }, this.label !== '' ? index.h("label", { class: { 'required': this.required } }, " ", this.label, " ") : '', index.h("div", { class: "select-container" }, index.h("div", { class: {
                'input-container': true,
                [this.state]: true,
                'select-disabled': this.disabled,
            }, ref: select => this.select = select, onClick: () => this.innerOnClick() }, index.h("div", { class: "input-container-inner" }, this.renderTags(), index.h("input", { ref: selectInput => this.selectInput = selectInput, class: {
                'multiple-select': this.multiple,
            }, autoComplete: "off", autoFocus: this.autofocus, disabled: this.disabled, name: this.name, placeholder: this.value ? '' : (this.placeholder || ''), readOnly: this.readonly, required: this.required, type: this.type, value: "", onInput: () => this.onInput(), onFocus: e => this.innerOnFocus(e), onBlur: e => this.innerOnBlur(e) }), index.h("span", { class: { 'dropdown-status-icon': true, 'expanded': this.isExpanded } }))), index.h("ul", { tabindex: "0", class: "dropdown", ref: ul => this.selectList = ul }, this.renderDropdown()), this.stateText !== '' ?
            index.h("span", { class: "help-block" }, this.stateText) : ''), index.h("div", { class: "overlay", ref: overlay => this.overlay = overlay, onClick: () => this.closeDropdown() })));
    }
    get host() { return index.getElement(this); }
    static get watchers() { return {
        "value": ["keyChanged"],
        "options": ["optionsChangedHandler"]
    }; }
    static get style() { return ":host {\n  --color-milk: #fff;\n  --color-elephant-900: #12344d;\n  --color-elephant-800: #264966;\n  --color-elephant-700: #345c7c;\n  --color-elephant-600: #447093;\n  --color-smoke-700: #475867;\n  --color-smoke-300: #92a2b1;\n  --color-smoke-100: #cfd7df;\n  --color-smoke-50: #ebeff3;\n  --color-smoke-25: #f3f5f7;\n  --color-jungle-800: #00795b;\n  --color-jungle-500: #00a886;\n  --color-jungle-100: #b4e5da;\n  --color-jungle-50: #e0f5f1;\n  --color-azure-800: #2c5cc5;\n  --color-azure-100: #bbdcfe;\n  --color-azure-50: #e5f2fd;\n  --color-persimmon-900: #c82124;\n  --color-persimmon-800: #d72d30;\n  --color-persimmon-100: #ffd0d6;\n  --color-persimmon-50: #ffecf0;\n  --color-casablanca-700: #e86f25;\n  --color-casablanca-300: #f8ab59;\n  --color-casablanca-100: #fedcb3;\n  --color-casablanca-50: #fef1e1;\n  --border-color: var(--color-smoke-100);\n  --border-success-color: var(--color-jungle-100);\n  --border-info-color: var(--color-azure-100);\n  --border-danger-color: var(--color-persimmon-100);\n  --border-warning-color: var(--color-casablanca-100);\n  --bg-dark: var(--color-elephant-900);\n  --bg-success: var(--color-jungle-50);\n  --bg-info: var(--color-azure-50);\n  --bg-danger: var(--color-persimmon-50);\n  --bg-warning: var(--color-casablanca-50);\n  --radius: 4px;\n  --radius-small: 2px;\n}\n\n:host {\n  --font-stack: -apple-system, blinkmacsystemfont, \"Segoe UI\", roboto, oxygen, ubuntu, cantarell, \"Open Sans\", \"Helvetica Neue\", sans-serif;\n  --font-weight-400: 400;\n  --font-weight-300: 300;\n  --font-weight-500: 500;\n  --font-weight-600: 600;\n  --font-weight-700: 700;\n  --font-size-10: 10px;\n  --font-size-12: 12px;\n  --font-size-14: 14px;\n  --font-size-16: 16px;\n  --font-size-18: 18px;\n  --font-size-20: 20px;\n  --font-size-24: 24px;\n  --text-default: var(--color-elephant-900);\n  --text-secondary: var(--color-smoke-700);\n  --text-disabled: var(--color-smoke-300);\n  --text-success: var(--color-jungle-800);\n  --text-info: var(--color-azure-800);\n  --text-danger: var(--color-persimmon-800);\n  --text-warning: var(--color-casablanca-700);\n  --text-link: var(--color-azure-800);\n  --icon-primary: var(--color-smoke-700);\n  --icon-primary-hover: var(--color-smoke-100);\n}\n\n:host {\n  font-family: var(--font-stack);\n  -webkit-font-smoothing: antialiased;\n  -moz-osx-font-smoothing: grayscale;\n  -webkit-box-sizing: border-box;\n  box-sizing: border-box;\n}\n\n/**\n  \@prop --min-height: Minimum height of the select component\n  \@prop --max-height: Maximum height of the select component\n*/\n:host {\n  --label-font: var(--font-stack);\n  --input-bg: var(--color-milk);\n  --help-color: var(--color-smoke-300);\n  --error-color: var(--color-persimmon-800);\n  --input-color: var(--color-elephant-900);\n  --input-disabled-bg: var(--color-smoke-25);\n  --input-hover-color: var(--color-smoke-700);\n  --input-focus-color: var(--color-azure-800);\n  --input-border: var(--color-smoke-100);\n  --warning-color: var(--color-casablanca-300);\n  --min-height: 10px;\n  --max-height: 200px;\n}\n\nlabel {\n  font-size: var(--font-size-12);\n  font-weight: var(--font-weight-500);\n  color: var(--color-elephant-900);\n  margin-bottom: 0;\n  padding-bottom: 4px;\n  padding-left: 2px;\n  display: block;\n}\nlabel.required::after {\n  content: \"*\";\n  position: relative;\n  display: inline-block;\n  top: 2px;\n  font-size: var(--font-size-14);\n  color: var(--error-color);\n  padding-left: 2px;\n}\n\n.input-container-inner {\n  display: -ms-flexbox;\n  display: flex;\n  width: 100%;\n  -ms-flex-wrap: wrap;\n  flex-wrap: wrap;\n  position: relative;\n}\n.input-container-inner input {\n  -ms-flex-positive: 1;\n  flex-grow: 1;\n  border: none;\n  margin-top: 3px;\n  font-size: var(--font-size-12);\n  font-weight: var(--font-weight-500);\n  letter-spacing: 0;\n  line-height: 1.4;\n  background-color: var(--input-bg);\n  min-width: 20px;\n  -webkit-box-sizing: border-box;\n  box-sizing: border-box;\n  overflow: hidden;\n}\n.input-container-inner input:focus {\n  border: none;\n  outline: none;\n}\n.input-container-inner input.multi-select {\n  width: auto;\n}\n.input-container-inner input:disabled {\n  background-color: var(--input-disabled-bg) !important;\n  border-color: transparent;\n  cursor: not-allowed;\n}\n\n.input-container {\n  width: calc(100% - 10px);\n  border: 0;\n  border: 1px solid var(--input-border);\n  margin: 5px 0 0;\n  border-radius: 4px;\n  padding: 4px 0 4px 10px;\n  background-color: var(--input-bg);\n  -webkit-box-shadow: none;\n  box-shadow: none;\n  min-height: 24px;\n  display: inline-block;\n}\n.input-container:hover {\n  border: 1px var(--input-hover-color) solid;\n  -webkit-transition: 0.2s linear;\n  transition: 0.2s linear;\n}\n.input-container.error {\n  border-color: var(--error-color);\n}\n.input-container.error span.dropdown-status-icon {\n  border-color: var(--error-color);\n}\n.input-container.error ~ span.help-block {\n  color: var(--error-color);\n}\n.input-container.warning {\n  border-color: var(--warning-color);\n}\n.input-container.warning span.dropdown-status-icon {\n  border-color: var(--warning-color);\n}\n.input-container.warning ~ span.help-block {\n  color: var(--warning-color);\n}\n.input-container.select-disabled {\n  color: var(--color-smoke-50);\n  background-color: var(--input-disabled-bg) !important;\n  border-style: solid;\n  cursor: not-allowed;\n}\n.input-container.select-disabled:hover {\n  border: 1px solid var(--input-border);\n}\n\n:host(.has-focus) .input-container {\n  outline: none;\n  background: var(--input-bg);\n  border: 1px solid transparent;\n  -webkit-box-shadow: 0 0 0 2px var(--input-focus-color);\n  box-shadow: 0 0 0 2px var(--input-focus-color);\n}\n:host(.has-focus) .input-container.error {\n  border-color: var(--error-color);\n  -webkit-box-shadow: 0 0 0 1px var(--error-color);\n  box-shadow: 0 0 0 1px var(--error-color);\n}\n:host(.has-focus) .input-container.error span.dropdown-status-icon {\n  border-color: var(--error-color);\n}\n:host(.has-focus) .input-container.error ~ span.help-block {\n  color: var(--error-color);\n}\n:host(.has-focus) .input-container.warning {\n  border-color: var(--warning-color);\n  -webkit-box-shadow: 0 0 0 1px var(--warning-color);\n  box-shadow: 0 0 0 1px var(--warning-color);\n}\n:host(.has-focus) .input-container.warning span.dropdown-status-icon {\n  border-color: var(--warning-color);\n}\n:host(.has-focus) .input-container.warning ~ span.help-block {\n  color: var(--warning-color);\n}\n\n.select-container {\n  margin-bottom: 16px;\n  width: inherit;\n  height: inherit;\n  position: relative;\n}\n.select-container .dropdown {\n  z-index: 99;\n}\n.select-container ul {\n  min-height: var(--min-height);\n  max-height: var(--max-height);\n  overflow-y: scroll;\n  overflow-x: hidden;\n  padding: 8px;\n  list-style: none;\n  margin: 0;\n  border-radius: 4px;\n  margin-top: 2px;\n  display: none;\n  position: absolute;\n  z-index: 1;\n  background: var(--input-bg);\n  width: 100%;\n  -webkit-box-sizing: border-box;\n  box-sizing: border-box;\n  outline: none;\n  -webkit-box-shadow: 0 2px 18px 0 rgba(18, 52, 77, 0.16), 0 2px 4px 0 rgba(18, 52, 77, 0.06);\n  box-shadow: 0 2px 18px 0 rgba(18, 52, 77, 0.16), 0 2px 4px 0 rgba(18, 52, 77, 0.06);\n  -webkit-user-select: none;\n     -moz-user-select: none;\n      -ms-user-select: none;\n          user-select: none;\n  -webkit-animation: dropdown-fade 0.15s;\n          animation: dropdown-fade 0.15s;\n  will-change: auto;\n}\n.select-container .dropdown-status-icon {\n  position: absolute;\n  display: inline-block;\n  right: 12px;\n  content: \"\";\n  top: 45%;\n  width: 5px;\n  height: 5px;\n  background-color: transparent;\n  border-style: inherit;\n  border-top: 1px solid #12344d;\n  border-right: 1px solid #12344d;\n  border-left: transparent;\n  -webkit-transform: rotate(135deg);\n  transform: rotate(135deg);\n  -webkit-transition: all 0.15s;\n  transition: all 0.15s;\n}\n.select-container .help-block {\n  font-size: var(--font-size-12);\n  margin-top: 3px;\n  color: var(--help-color);\n  position: inherit;\n  margin-bottom: 0;\n  display: block;\n  padding-left: 2px;\n}\n.select-container .dropdown-status-icon.expanded {\n  -webkit-transform: rotate(315deg);\n  transform: rotate(315deg);\n}\n\n.overlay {\n  width: 100vw;\n  height: 100vh;\n  display: none;\n  position: fixed;\n  top: 0;\n  left: 0;\n  z-index: 95;\n  background-color: transparent;\n}\n\n::-moz-placeholder {\n  color: var(--color-smoke-300);\n  opacity: 1;\n}\n\n:-ms-input-placeholder {\n  color: var(--color-smoke-300);\n  opacity: 1;\n}\n\n::-webkit-input-placeholder {\n  color: var(--color-smoke-300);\n  opacity: 1;\n}\n\n::-ms-input-placeholder {\n  color: var(--color-smoke-300);\n  opacity: 1;\n}\n\n::placeholder {\n  color: var(--color-smoke-300);\n  opacity: 1;\n}\n\n:-ms-input-placeholder {\n  color: var(--color-smoke-300);\n}\n\n::-ms-input-placeholder {\n  color: var(--color-smoke-300);\n}\n\n\@-webkit-keyframes dropdown-fade {\n  0% {\n    opacity: 0;\n    -webkit-transform: translateY(-10%);\n    transform: translateY(-10%);\n  }\n  50% {\n    opacity: 0.7;\n    -webkit-transform: translateY(-5%);\n    transform: translateY(-5%);\n  }\n  100% {\n    opacity: 1;\n    -webkit-transform: translateY(0%);\n    transform: translateY(0%);\n  }\n}\n\n\@keyframes dropdown-fade {\n  0% {\n    opacity: 0;\n    -webkit-transform: translateY(-10%);\n    transform: translateY(-10%);\n  }\n  50% {\n    opacity: 0.7;\n    -webkit-transform: translateY(-5%);\n    transform: translateY(-5%);\n  }\n  100% {\n    opacity: 1;\n    -webkit-transform: translateY(0%);\n    transform: translateY(0%);\n  }\n}"; }
};

exports.fw_select = Select;
