// @freshworks/crayons/dist/cjs/fw-timepicker.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-19272a91.js');
const moment_min = require('./moment.min-0023acf8.js');
const utils = require('./utils-a88d9e0c.js');

const Timepicker = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        /**
         * State for all the time value\s
         */
        this.timeValues = [];
        /**
         * Format in which time values are populated in the list box. If the value is hh:mm p, the time values are in the 12-hour format. If the value is hh:mm, the time values are in the 24-hr format.
         */
        this.format = 'hh:mm A';
        /**
         * Set true to disable the element
         */
        this.disabled = false;
        /**
         * Represent the intervals and can be a number or array of numbers representing the minute values in an hour
         */
        this.isMeridianFormat = this.format === 'hh:mm A';
        /**
         * Name of the component, saved as part of form data.
         */
        this.name = '';
        /**
         * Time interval between the values displayed in the list, specified in minutes.
         */
        this.interval = 30;
        /**
         * Lower time-limit for the values displayed in the list. If this attribute’s value is in the hh:mm format, it is assumed to be hh:mm AM.
         */
        this.minTime = this.isMeridianFormat ? '12:00 AM' : '00:00';
        /**
         * Upper time-limit for the values displayed in the list. If this attribute’s value is in the hh:mm format, it is assumed to be hh:mm AM.
         */
        this.maxTime = this.isMeridianFormat ? '11:30 PM' : '23:30';
        /**
         * Boolean representing whethere it is default end time
         */
        this.isDefaultEndTime = ['11:30 PM', '23:30'].includes(this.maxTime);
        this.getTimeOptionsMeta = nonMeridianFormat => {
            const preferredFormat = this.format;
            const timeIntervalArgs = {
                interval: this.interval,
                startTime: moment_min.moment(this.minTime, preferredFormat).format(nonMeridianFormat),
                endTime: moment_min.moment(this.maxTime, preferredFormat).format(nonMeridianFormat),
            };
            return timeIntervalArgs;
        };
        this.setTimeValues = () => {
            const meridianFormat = 'hh:mm A';
            const nonMeridianFormat = 'HH:mm';
            const { interval, startTime, endTime } = this.getTimeOptionsMeta(nonMeridianFormat);
            let currentTimeInMs = moment_min.moment(startTime, nonMeridianFormat).valueOf();
            const endTimeInMs = moment_min.moment(endTime, nonMeridianFormat).valueOf();
            while (currentTimeInMs <= endTimeInMs) {
                this.timeValues.push({
                    meridianFormat: moment_min.moment(currentTimeInMs).format(meridianFormat),
                    nonMeridianFormat: moment_min.moment(currentTimeInMs).format(nonMeridianFormat),
                });
                currentTimeInMs = moment_min.moment(currentTimeInMs)
                    .add(interval, 'minutes')
                    .valueOf();
            }
        };
    }
    currentTimeLabel(time) {
        return this.isMeridianFormat ? time.meridianFormat : time.nonMeridianFormat;
    }
    currentTimeValue(time) {
        return time.nonMeridianFormat;
    }
    setTimeValue(e) {
        const { value } = e.detail;
        this.value = value;
    }
    setEndTime() {
        if (this.isDefaultEndTime) {
            this.maxTime = this.isMeridianFormat ?
                `11:59 PM` : `23:59`;
        }
    }
    componentWillLoad() {
        if (this.interval !== 30) {
            this.setEndTime();
        }
        this.setTimeValues();
    }
    render() {
        const { host, name, value } = this;
        utils.renderHiddenField(host, name, value);
        return (index.h("fw-select", { disabled: this.disabled, value: this.value, onFwChange: e => this.setTimeValue(e) }, this.timeValues.map(time => index.h("fw-select-option", { value: this.currentTimeValue(time) }, this.currentTimeLabel(time)))));
    }
    get host() { return index.getElement(this); }
};

exports.fw_timepicker = Timepicker;
