// @freshworks/crayons/dist/crayons/fw-textarea.entry.js
import { r as registerInstance, c as createEvent, h, H as Host, g as getElement } from './index-e608b2e3.js';
import { r as renderHiddenField } from './utils-926a70fd.js';

const Textarea = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.hasFocus = false;
        /**
         * Label displayed on the interface, for the component.
         */
        this.label = '';
        /**
         * Default value displayed in the input box.
         */
        this.value = '';
        /**
         * Name of the component, saved as part of form data.
         */
        this.name = '';
        /**
         * Theme based on which the input box is styled.
         */
        this.state = 'normal';
        /**
         * Type of text wrapping used by the input box. If the value is hard, the text in the textarea is wrapped (contains line breaks) when the form data is saved. If the value is soft, the text in the textarea is saved as a single line, when the form data is saved.
         */
        this.wrap = 'soft';
        /**
         * Descriptive or instructional text displayed below the input box.
         */
        this.stateText = '';
        /**
         * If true, the textarea is autofocused
         */
        this.autofocus = false;
        /**
         * If true, the user cannot enter a value in the input box. If the attribute’s value is undefined, the value is set to false.
         */
        this.readonly = false;
        /**
         * Specifies the input box as a mandatory field and displays an asterisk next to the label. If the attribute’s value is undefined, the value is set to false.
         */
        this.required = false;
        /**
         * Disables the text area on the interface. If the attribute’s value is undefined, the value is set to false.
         */
        this.disabled = false;
        this.onInput = (ev) => {
            const input = ev.target;
            if (input) {
                this.value = input.value || '';
            }
            this.fwInput.emit(ev);
        };
        this.onFocus = () => {
            this.hasFocus = true;
            this.fwFocus.emit();
        };
        this.onBlur = () => {
            this.hasFocus = false;
            this.fwBlur.emit();
        };
        this.fwChange = createEvent(this, "fwChange", 7);
        this.fwFocus = createEvent(this, "fwFocus", 7);
        this.fwBlur = createEvent(this, "fwBlur", 7);
        this.fwInput = createEvent(this, "fwInput", 7);
    }
    watchHandler(newValue) {
        this.fwChange.emit({ value: newValue });
    }
    getValue() {
        return this.value || '';
    }
    hasValue() {
        return this.getValue().length > 0;
    }
    /**
     * Sets focus on a specific `fw-textarea`. Use this method instead of the global `input.focus()`.
     */
    async setFocus() {
        if (this.nativeInput) {
            this.nativeInput.focus();
        }
    }
    render() {
        const { host, name, value } = this;
        renderHiddenField(host, name, value);
        return (h(Host, { "aria-disabled": this.disabled, class: {
                'has-value': this.hasValue(),
                'has-focus': this.hasFocus,
            } }, h("div", { class: "textarea-container" }, this.label !== '' ? h("label", { class: {
                'required': this.required,
            } }, this.label) : '', h("div", { class: {
                'textarea-container-inner': true,
                [this.state]: true,
            } }, h("textarea", { class: {
                'responsive': this.cols === undefined,
            }, ref: input => this.nativeInput = input, disabled: this.disabled, name: this.name, placeholder: this.placeholder || '', minLength: this.minlength, maxLength: this.maxlength, readOnly: this.readonly, autoFocus: this.autofocus, required: this.required, value: this.value, onInput: e => this.onInput(e), onBlur: this.onBlur, onFocus: this.onFocus, rows: this.rows, cols: this.cols, wrap: this.wrap })), this.stateText !== '' ?
            h("span", { class: "help-block" }, this.stateText) : '')));
    }
    get host() { return getElement(this); }
    static get watchers() { return {
        "value": ["watchHandler"]
    }; }
    static get style() { return ":host {\n  --color-milk: #fff;\n  --color-elephant-900: #12344d;\n  --color-elephant-800: #264966;\n  --color-elephant-700: #345c7c;\n  --color-elephant-600: #447093;\n  --color-smoke-700: #475867;\n  --color-smoke-300: #92a2b1;\n  --color-smoke-100: #cfd7df;\n  --color-smoke-50: #ebeff3;\n  --color-smoke-25: #f3f5f7;\n  --color-jungle-800: #00795b;\n  --color-jungle-500: #00a886;\n  --color-jungle-100: #b4e5da;\n  --color-jungle-50: #e0f5f1;\n  --color-azure-800: #2c5cc5;\n  --color-azure-100: #bbdcfe;\n  --color-azure-50: #e5f2fd;\n  --color-persimmon-900: #c82124;\n  --color-persimmon-800: #d72d30;\n  --color-persimmon-100: #ffd0d6;\n  --color-persimmon-50: #ffecf0;\n  --color-casablanca-700: #e86f25;\n  --color-casablanca-300: #f8ab59;\n  --color-casablanca-100: #fedcb3;\n  --color-casablanca-50: #fef1e1;\n  --border-color: var(--color-smoke-100);\n  --border-success-color: var(--color-jungle-100);\n  --border-info-color: var(--color-azure-100);\n  --border-danger-color: var(--color-persimmon-100);\n  --border-warning-color: var(--color-casablanca-100);\n  --bg-dark: var(--color-elephant-900);\n  --bg-success: var(--color-jungle-50);\n  --bg-info: var(--color-azure-50);\n  --bg-danger: var(--color-persimmon-50);\n  --bg-warning: var(--color-casablanca-50);\n  --radius: 4px;\n  --radius-small: 2px;\n}\n\n:host {\n  --font-stack: -apple-system, blinkmacsystemfont, \"Segoe UI\", roboto, oxygen, ubuntu, cantarell, \"Open Sans\", \"Helvetica Neue\", sans-serif;\n  --font-weight-400: 400;\n  --font-weight-300: 300;\n  --font-weight-500: 500;\n  --font-weight-600: 600;\n  --font-weight-700: 700;\n  --font-size-10: 10px;\n  --font-size-12: 12px;\n  --font-size-14: 14px;\n  --font-size-16: 16px;\n  --font-size-18: 18px;\n  --font-size-20: 20px;\n  --font-size-24: 24px;\n  --text-default: var(--color-elephant-900);\n  --text-secondary: var(--color-smoke-700);\n  --text-disabled: var(--color-smoke-300);\n  --text-success: var(--color-jungle-800);\n  --text-info: var(--color-azure-800);\n  --text-danger: var(--color-persimmon-800);\n  --text-warning: var(--color-casablanca-700);\n  --text-link: var(--color-azure-800);\n  --icon-primary: var(--color-smoke-700);\n  --icon-primary-hover: var(--color-smoke-100);\n}\n\n:host {\n  font-family: var(--font-stack);\n  -webkit-font-smoothing: antialiased;\n  -moz-osx-font-smoothing: grayscale;\n  -webkit-box-sizing: border-box;\n  box-sizing: border-box;\n}\n\n/* Need to check with designer */\n/* Need to check with designer */\n/**\n  \@prop --input-color: Color of the textarea input\n**/\n:host {\n  --label-font: var(--font-stack);\n  --input-bg: var(--color-milk);\n  --help-color: var(--color-smoke-300);\n  --error-color: var(--color-persimmon-800);\n  --input-color: var(--color-elephant-900);\n  --input-disabled-bg: var(--color-smoke-25);\n  --input-hover-color: var(--color-smoke-700);\n  --input-focus-color: var(--color-azure-800);\n  --input-border: var(--color-smoke-100);\n  --warning-color: var(--color-casablanca-300);\n}\n\n.textarea-container {\n  margin-bottom: 16px;\n}\n\nlabel {\n  font-size: var(--font-size-12);\n  font-weight: var(--font-weight-500);\n  color: var(--color-elephant-900);\n  margin-bottom: 0;\n  padding-bottom: 4px;\n  padding-left: 2px;\n  display: block;\n}\nlabel.required::after {\n  content: \"*\";\n  position: relative;\n  display: inline-block;\n  top: 2px;\n  font-size: var(--font-size-14);\n  color: var(--error-color);\n  padding-left: 2px;\n}\n\n.textarea-container-inner {\n  display: block;\n  width: 100%;\n  position: relative;\n}\n.textarea-container-inner textarea {\n  border: 0;\n  border: 1px solid var(--input-border);\n  margin: 5px 0 0;\n  border-radius: 4px;\n  padding: 4px 12px 5px;\n  background-color: var(--input-bg);\n  -webkit-box-shadow: none;\n  box-shadow: none;\n  min-height: 24px;\n  font-size: var(--font-size-12);\n  font-weight: var(--font-weight-500);\n  letter-spacing: 0;\n  line-height: 20px;\n  font-family: inherit;\n  cursor: text;\n  display: inline-block;\n  color: var(--input-color);\n}\n.textarea-container-inner textarea:hover {\n  border: 1px var(--input-hover-color) solid;\n  -webkit-transition: 0.2s linear;\n  transition: 0.2s linear;\n}\n.textarea-container-inner textarea:focus {\n  outline: none;\n  background: var(--input-bg);\n  border: 1px solid transparent;\n  -webkit-box-shadow: 0 0 0 2px var(--input-focus-color);\n  box-shadow: 0 0 0 2px var(--input-focus-color);\n}\n.textarea-container-inner textarea[disabled] {\n  color: var(--color-smoke-300);\n  background-color: var(--input-disabled-bg);\n  border-style: solid;\n}\n.textarea-container-inner textarea[disabled]:hover {\n  border: 1px solid var(--input-border);\n  cursor: not-allowed;\n}\n.textarea-container-inner + .help-block {\n  font-size: var(--font-size-12);\n  margin-top: 3px;\n  color: var(--help-color);\n  position: inherit;\n  margin-bottom: 0;\n  display: block;\n  padding-left: 2px;\n}\n.textarea-container-inner.error > textarea {\n  border-color: var(--error-color);\n}\n.textarea-container-inner.error > textarea:focus {\n  -webkit-box-shadow: none;\n  box-shadow: none;\n  border-color: var(--error-color);\n}\n.textarea-container-inner.error > textarea:hover {\n  border-color: var(--error-color);\n}\n.textarea-container-inner.error + .help-block {\n  color: var(--error-color);\n}\n.textarea-container-inner.warning > textarea {\n  border-color: var(--warning-color);\n}\n.textarea-container-inner.warning > textarea:focus {\n  -webkit-box-shadow: none;\n  box-shadow: none;\n  border-color: var(--warning-color);\n}\n.textarea-container-inner.warning > textarea:hover {\n  border-color: var(--warning-color);\n}\n.textarea-container-inner.warning + .help-block {\n  color: var(--warning-color);\n}\n\n::-moz-placeholder {\n  color: var(--color-smoke-300);\n  opacity: 1;\n}\n\n:-ms-input-placeholder {\n  color: var(--color-smoke-300);\n  opacity: 1;\n}\n\n::-webkit-input-placeholder {\n  color: var(--color-smoke-300);\n  opacity: 1;\n}\n\n::-ms-input-placeholder {\n  color: var(--color-smoke-300);\n  opacity: 1;\n}\n\n::placeholder {\n  color: var(--color-smoke-300);\n  opacity: 1;\n}\n\n:-ms-input-placeholder {\n  color: var(--color-smoke-300);\n}\n\n::-ms-input-placeholder {\n  color: var(--color-smoke-300);\n}\n\n.responsive {\n  width: 100%;\n  -webkit-box-sizing: border-box;\n  box-sizing: border-box;\n}"; }
};

export { Textarea as fw_textarea };
