// @freshworks/crayons/dist/crayons/fw-toast.entry.js
import { r as registerInstance, c as createEvent, d as getAssetPath, h, H as Host } from './index-e608b2e3.js';

const Toast = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        /**
         * visibility state of toast
         */
        this.isOpen = false;
        /**
         * Icon svg html content
         */
        this.svgHTML = '';
        /**
         * To indicate toast close timeout
         */
        this.isTimedOut = false;
        /**
         * To add close animation class to toast
         */
        this.fadeOut = false;
        /**
         * State icon size
         */
        this.iconSize = 14;
        /**
         * Type of the toast - success,failure, warning, inprogress
         */
        this.type = 'warning';
        /**
         * Time duration of the toast visibility
         */
        this.timeout = 4000;
        /**
         *  The Content of the action link
         */
        this.actionLinkText = '';
        /**
         *  won't close automatically
         */
        this.sticky = false;
        /**
         *  position of the toast notification in screen
         */
        this.position = 'top-center';
        this.fwLinkClick = createEvent(this, "fwLinkClick", 7);
    }
    async getSVGHTML(iconName) {
        const response = await fetch(getAssetPath(`toast-assets/icons/${iconName}.svg`));
        const data = await response.text();
        return data;
    }
    setSVGState(iconName) {
        if (this.type === 'inprogress') {
            return;
        }
        this.getSVGHTML(iconName).then(res => {
            this.svgHTML = res;
        }).catch();
    }
    componentWillLoad() {
        this.setSVGState(this.type);
    }
    async trigger(configs) {
        if (this.isOpen || this.timerId) {
            await this.closeToast();
        }
        Object.keys(configs).forEach(key => {
            this[key] = configs[key];
        });
        this.fadeOut = false;
        this.isOpen = true;
        this.isTimedOut = false;
        this.timerId = setTimeout(async () => {
            if (!this.sticky) {
                if (!this.pauseOnHover || this.pauseOnHover && !this.isMouseHovered) {
                    await this.closeToast();
                }
                this.isTimedOut = true;
            }
        }, this.timeout);
    }
    async mouseHover(value = false) {
        this.isMouseHovered = value;
        if (this.isTimedOut && !this.isMouseHovered) {
            await this.closeToast();
        }
    }
    closingAnimation() {
        this.fadeOut = true;
        return new Promise(resolve => setTimeout(() => { this.isOpen = false; resolve(); }, 500));
    }
    async closeToast() {
        if (this.timerId) {
            clearTimeout(this.timerId);
        }
        await this.closingAnimation();
    }
    render() {
        return (h(Host, { class: `toast ${this.position} ${this.type} ${this.isOpen ? 'is-open' : ''} ${this.fadeOut ? 'fade-out' : ''}`, "aria-hidden": this.isOpen ? 'false' : 'true', onmouseover: () => this.mouseHover(true), onmouseout: () => this.mouseHover(false) }, h("div", null, this.type === 'inprogress' ? (h("fw-spinner", { class: "icon" })) : (h("fw-icon", { class: "icon", size: this.iconSize, name: this.type })), h("span", { class: "content" }, this.content), h("fw-icon", { size: 7, color: "#000", name: "cross", class: "cross", onClick: () => this.closeToast() }), this.actionLinkText.length > 0 ? (h("div", { class: "action-link", onClick: () => this.fwLinkClick.emit() }, this.actionLinkText)) : '')));
    }
    static get assetsDirs() { return ["toast-assets"]; }
    static get watchers() { return {
        "type": ["setSVGState"]
    }; }
    static get style() { return ":host {\n  --color-milk: #fff;\n  --color-elephant-900: #12344d;\n  --color-elephant-800: #264966;\n  --color-elephant-700: #345c7c;\n  --color-elephant-600: #447093;\n  --color-smoke-700: #475867;\n  --color-smoke-300: #92a2b1;\n  --color-smoke-100: #cfd7df;\n  --color-smoke-50: #ebeff3;\n  --color-smoke-25: #f3f5f7;\n  --color-jungle-800: #00795b;\n  --color-jungle-500: #00a886;\n  --color-jungle-100: #b4e5da;\n  --color-jungle-50: #e0f5f1;\n  --color-azure-800: #2c5cc5;\n  --color-azure-100: #bbdcfe;\n  --color-azure-50: #e5f2fd;\n  --color-persimmon-900: #c82124;\n  --color-persimmon-800: #d72d30;\n  --color-persimmon-100: #ffd0d6;\n  --color-persimmon-50: #ffecf0;\n  --color-casablanca-700: #e86f25;\n  --color-casablanca-300: #f8ab59;\n  --color-casablanca-100: #fedcb3;\n  --color-casablanca-50: #fef1e1;\n  --border-color: var(--color-smoke-100);\n  --border-success-color: var(--color-jungle-100);\n  --border-info-color: var(--color-azure-100);\n  --border-danger-color: var(--color-persimmon-100);\n  --border-warning-color: var(--color-casablanca-100);\n  --bg-dark: var(--color-elephant-900);\n  --bg-success: var(--color-jungle-50);\n  --bg-info: var(--color-azure-50);\n  --bg-danger: var(--color-persimmon-50);\n  --bg-warning: var(--color-casablanca-50);\n  --radius: 4px;\n  --radius-small: 2px;\n}\n\n:host {\n  --font-stack: -apple-system, blinkmacsystemfont, \"Segoe UI\", roboto, oxygen, ubuntu, cantarell, \"Open Sans\", \"Helvetica Neue\", sans-serif;\n  --font-weight-400: 400;\n  --font-weight-300: 300;\n  --font-weight-500: 500;\n  --font-weight-600: 600;\n  --font-weight-700: 700;\n  --font-size-10: 10px;\n  --font-size-12: 12px;\n  --font-size-14: 14px;\n  --font-size-16: 16px;\n  --font-size-18: 18px;\n  --font-size-20: 20px;\n  --font-size-24: 24px;\n  --text-default: var(--color-elephant-900);\n  --text-secondary: var(--color-smoke-700);\n  --text-disabled: var(--color-smoke-300);\n  --text-success: var(--color-jungle-800);\n  --text-info: var(--color-azure-800);\n  --text-danger: var(--color-persimmon-800);\n  --text-warning: var(--color-casablanca-700);\n  --text-link: var(--color-azure-800);\n  --icon-primary: var(--color-smoke-700);\n  --icon-primary-hover: var(--color-smoke-100);\n}\n\n:host {\n  font-family: var(--font-stack);\n  -webkit-font-smoothing: antialiased;\n  -moz-osx-font-smoothing: grayscale;\n  -webkit-box-sizing: border-box;\n  box-sizing: border-box;\n}\n\n:host {\n  display: none;\n}\n\n:host(.is-open) {\n  display: block;\n  -webkit-animation-duration: 0.5s;\n          animation-duration: 0.5s;\n  -webkit-animation-name: fadeIn;\n          animation-name: fadeIn;\n  z-index: 999;\n}\n\n:host(.toast) {\n  position: fixed;\n  top: 10px;\n  width: 400px;\n  padding: 10px;\n  border: 1px solid var(--color-smoke-50);\n  border-radius: 4px;\n  background-color: var(--color-milk);\n}\n\n:host(.toast.success) {\n  border-top: 5px solid var(--color-jungle-500);\n}\n:host(.toast.error) {\n  border-top: 5px solid var(--color-persimmon-800);\n}\n:host(.toast.warning) {\n  border-top: 5px solid var(--color-casablanca-300);\n}\n:host(.toast.inprogress) {\n  border-top: 5px solid var(--color-azure-800);\n}\n:host(.toast.top-center) {\n  left: calc(50% - 200px);\n}\n:host(.toast.top-left) {\n  left: 10px;\n}\n:host(.toast.top-right) {\n  right: 10px;\n}\n\n.content {\n  color: var(--color-elephant-900);\n  line-height: 20px;\n  width: 83%;\n  float: left;\n  margin: 2px auto 8px;\n  font-size: var(--font-size-14);\n  font-weight: var(--font-weight-500);\n}\n\n.icon {\n  margin: 5px 16px 0 5px;\n  vertical-align: text-top;\n  float: left;\n}\n\n.action-link {\n  color: var(--color-azure-800);\n  line-height: 14px;\n  margin-left: 38px;\n  cursor: pointer;\n  float: left;\n  font-size: var(--font-size-12);\n  font-weight: var(--font-weight-600);\n}\n\n.cross {\n  float: right;\n  cursor: pointer;\n  margin-right: 6px;\n}\n\n\@-webkit-keyframes fadeOut {\n  100% {\n    top: -600px;\n  }\n}\n\n\@keyframes fadeOut {\n  100% {\n    top: -600px;\n  }\n}\n\@-webkit-keyframes fadeIn {\n  0% {\n    top: -600px;\n  }\n  100% {\n    top: 10px;\n  }\n}\n\@keyframes fadeIn {\n  0% {\n    top: -600px;\n  }\n  100% {\n    top: 10px;\n  }\n}\n:host(.fade-out) {\n  -webkit-animation-duration: 1s;\n          animation-duration: 1s;\n  -webkit-animation-name: fadeOut;\n          animation-name: fadeOut;\n}"; }
};

export { Toast as fw_toast };
