// @freshworks/crayons/dist/collection/components/select/select.js
import { Host, h } from "@stencil/core";
import { renderHiddenField } from '../../utils/utils';
export class Select {
    constructor() {
        /**
         * If the dropdown is shown or not
         */
        this.isExpanded = false;
        this.options = [];
        this.filteredOptions = [];
        this.hasFocus = false;
        this.didInit = false;
        /**
         * Label displayed on the interface, for the component.
         */
        this.label = '';
        /**
         * Name of the component, saved as part of form data.
         */
        this.name = '';
        /**
         * Type of option accepted as the input value. If a user tries to enter an option other than the specified type, the list is not populated.
         */
        this.type = 'text';
        /**
         * Theme based on which the list box is styled.
         */
        this.state = 'normal';
        /**
         * Descriptive or instructional text displayed below the list box.
         */
        this.stateText = '';
        /**
         * If true, the user cannot modify the default value selected. If the attribute's value is undefined, the value is set to true.
         */
        this.readonly = false;
        /**
         * If true, the select component is auto focused on the page
         */
        this.autofocus = false;
        /**
         * Specifies the select field as a mandatory field and displays an asterisk next to the label. If the attribute’s value is undefined, the value is set to false.
         */
        this.required = false;
        /**
         * If true, the user must select a value. The default value is not displayed.
         */
        this.forceSelect = true;
        /**
         * Disables the component on the interface. If the attribute’s value is undefined, the value is set to false.
         */
        this.disabled = false;
        /**
         * Enables selection of multiple options. If the attribute’s value is undefined, the value is set to false.
         */
        this.multiple = false;
        /**
         * Works with `multiple` enabled. Configures the maximum number of options that can be selected with a multi-select component.
         */
        this.max = Number.MAX_VALUE;
        this.changeEmittable = () => !this.disabled;
        this.closeDropdown = () => {
            this.selectList.style.display = 'none';
            this.overlay.style.display = 'none';
            this.isExpanded = false;
        };
        this.innerOnFocus = (e) => {
            if (this.changeEmittable()) {
                this.hasFocus = true;
                this.fwFocus.emit(e);
            }
        };
        this.innerOnClick = () => {
            if (this.changeEmittable()) {
                this.filteredOptions = this.options;
                this.selectList.style.display = 'block';
                this.selectList.style.width = String(this.select.clientWidth) + 'px';
                this.isExpanded = true;
                this.overlay.style.display = 'block';
            }
        };
        this.innerOnBlur = (e) => {
            if (this.changeEmittable()) {
                this.closeDropdown();
                this.hasFocus = false;
                this.fwBlur.emit(e);
            }
        };
    }
    keyChanged(newValue, oldValue) {
        if (JSON.stringify(newValue) !== JSON.stringify(oldValue)) {
            this.options = this.options.map(option => {
                option.selected = Array.isArray(this.value)
                    ? this.value.includes(option.value)
                    : this.value === option.value;
                return option;
            });
            if (this.didInit) {
                this.fwChange.emit({ value: this.value });
            }
        }
    }
    fwSelectedHandler(selectedItem) {
        this.options = this.options.map(option => {
            if (selectedItem.detail.value === option.value) {
                option.selected = selectedItem.detail.selected;
            }
            else if (!this.multiple) {
                option.selected = false;
            }
            return option;
        });
        this.selectInput.value = '';
        this.renderInput();
        this.closeDropdown();
        selectedItem.stopPropagation();
    }
    optionsChangedHandler() {
        this.renderInput();
    }
    fwCloseHandler(ev) {
        this.options = this.options.map(option => {
            if (option.value === ev.detail.value) {
                option.selected = false;
            }
            return option;
        });
    }
    onKeyDonw(ev) {
        switch (ev.key) {
            case 'ArrowDown':
                this.innerOnClick();
                break;
            case 'Escape':
                this.innerOnBlur(ev);
                break;
        }
    }
    onInput() {
        const value = this.selectInput.value.toLowerCase();
        this.filteredOptions = value !== ''
            ? this.options.filter(option => option.text.toLowerCase().startsWith(value))
            : this.options;
        this.renderInput();
    }
    renderTags() {
        if (this.multiple) {
            return this.options
                .filter(option => option.selected)
                .map(option => h("fw-tag", { text: option.text, disabled: option.disabled, value: option.value }));
        }
    }
    renderInput() {
        const selectedOptions = this.options.filter(option => option.selected);
        if (selectedOptions.length > 0) {
            this.value = this.multiple ? selectedOptions.map(option => option.value) : selectedOptions[0].value || '';
            if (this.selectInput) {
                this.selectInput.value = this.multiple ? this.selectInput.value : selectedOptions[0].text || '';
            }
        }
        else if (selectedOptions.length === 0) {
            this.value = undefined;
        }
    }
    renderDropdown() {
        return this.filteredOptions.map(option => {
            var _a;
            return (h("fw-select-option", { value: option.value, selected: option.selected, disabled: option.disabled || (((_a = this.value) === null || _a === void 0 ? void 0 : _a.length) >= this.max), html: option.isHtml, htmlContent: option.htmlContent }, option.text));
        });
    }
    componentWillLoad() {
        const selectOptions = Array.from(this.host.querySelectorAll('fw-select-option'));
        const options = selectOptions.map(option => {
            return {
                isHtml: option.html,
                text: option.html ? option.optionText : option.textContent,
                value: option.value,
                selected: option.value === this.value || option.selected,
                disabled: option.disabled || this.disabled,
                htmlContent: option.html ? option.innerHTML : '',
            };
        });
        this.options = options;
        this.filteredOptions = this.options;
        this.host.innerHTML = '';
    }
    componentDidLoad() {
        this.renderInput();
        this.didInit = true;
    }
    async getSelectedItem() {
        return this.options.filter(option => option.selected);
    }
    async setSelectedValues(values) {
        if (this.multiple) {
            this.options.forEach(option => {
                option.selected = values.includes(option.value);
            });
            this.renderInput();
        }
    }
    render() {
        const { host, name, value } = this;
        renderHiddenField(host, name, value);
        return (h(Host, { "aria-disabled": this.disabled, class: {
                'has-focus': this.hasFocus,
            } },
            this.label !== '' ? h("label", { class: { 'required': this.required } },
                " ",
                this.label,
                " ") : '',
            h("div", { class: "select-container" },
                h("div", { class: {
                        'input-container': true,
                        [this.state]: true,
                        'select-disabled': this.disabled,
                    }, ref: select => this.select = select, onClick: () => this.innerOnClick() },
                    h("div", { class: "input-container-inner" },
                        this.renderTags(),
                        h("input", { ref: selectInput => this.selectInput = selectInput, class: {
                                'multiple-select': this.multiple,
                            }, autoComplete: "off", autoFocus: this.autofocus, disabled: this.disabled, name: this.name, placeholder: this.value ? '' : (this.placeholder || ''), readOnly: this.readonly, required: this.required, type: this.type, value: "", onInput: () => this.onInput(), onFocus: e => this.innerOnFocus(e), onBlur: e => this.innerOnBlur(e) }),
                        h("span", { class: { 'dropdown-status-icon': true, 'expanded': this.isExpanded } }))),
                h("ul", { tabindex: "0", class: "dropdown", ref: ul => this.selectList = ul }, this.renderDropdown()),
                this.stateText !== '' ?
                    h("span", { class: "help-block" }, this.stateText) : ''),
            h("div", { class: "overlay", ref: overlay => this.overlay = overlay, onClick: () => this.closeDropdown() })));
    }
    static get is() { return "fw-select"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() { return {
        "$": ["select.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["select.css"]
    }; }
    static get properties() { return {
        "label": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Label displayed on the interface, for the component."
            },
            "attribute": "label",
            "reflect": false,
            "defaultValue": "''"
        },
        "value": {
            "type": "any",
            "mutable": true,
            "complexType": {
                "original": "any",
                "resolved": "any",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Value of the option that is displayed as the default selection, in the list box. Must be a valid value corresponding to the fw-select-option components used in Select."
            },
            "attribute": "value",
            "reflect": false
        },
        "name": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Name of the component, saved as part of form data."
            },
            "attribute": "name",
            "reflect": false,
            "defaultValue": "''"
        },
        "type": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "'text' | 'number'",
                "resolved": "\"number\" | \"text\"",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Type of option accepted as the input value. If a user tries to enter an option other than the specified type, the list is not populated."
            },
            "attribute": "type",
            "reflect": false,
            "defaultValue": "'text'"
        },
        "placeholder": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string | null",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "Text displayed in the list box before an option is selected."
            },
            "attribute": "placeholder",
            "reflect": false
        },
        "state": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "'normal' | 'warning' | 'error'",
                "resolved": "\"error\" | \"normal\" | \"warning\"",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Theme based on which the list box is styled."
            },
            "attribute": "state",
            "reflect": false,
            "defaultValue": "'normal'"
        },
        "stateText": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Descriptive or instructional text displayed below the list box."
            },
            "attribute": "state-text",
            "reflect": false,
            "defaultValue": "''"
        },
        "readonly": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "If true, the user cannot modify the default value selected. If the attribute's value is undefined, the value is set to true."
            },
            "attribute": "readonly",
            "reflect": false,
            "defaultValue": "false"
        },
        "autofocus": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "If true, the select component is auto focused on the page"
            },
            "attribute": "autofocus",
            "reflect": false,
            "defaultValue": "false"
        },
        "required": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Specifies the select field as a mandatory field and displays an asterisk next to the label. If the attribute\u2019s value is undefined, the value is set to false."
            },
            "attribute": "required",
            "reflect": false,
            "defaultValue": "false"
        },
        "forceSelect": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "If true, the user must select a value. The default value is not displayed."
            },
            "attribute": "force-select",
            "reflect": false,
            "defaultValue": "true"
        },
        "disabled": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Disables the component on the interface. If the attribute\u2019s value is undefined, the value is set to false."
            },
            "attribute": "disabled",
            "reflect": false,
            "defaultValue": "false"
        },
        "multiple": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Enables selection of multiple options. If the attribute\u2019s value is undefined, the value is set to false."
            },
            "attribute": "multiple",
            "reflect": false,
            "defaultValue": "false"
        },
        "max": {
            "type": "number",
            "mutable": false,
            "complexType": {
                "original": "number",
                "resolved": "number",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Works with `multiple` enabled. Configures the maximum number of options that can be selected with a multi-select component."
            },
            "attribute": "max",
            "reflect": false,
            "defaultValue": "Number.MAX_VALUE"
        }
    }; }
    static get states() { return {
        "isExpanded": {},
        "options": {},
        "filteredOptions": {},
        "hasFocus": {},
        "didInit": {}
    }; }
    static get events() { return [{
            "method": "fwChange",
            "name": "fwChange",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": "Triggered when a value is selected or deselected from the list box options."
            },
            "complexType": {
                "original": "any",
                "resolved": "any",
                "references": {}
            }
        }, {
            "method": "fwFocus",
            "name": "fwFocus",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": "Triggered when the list box comes into focus."
            },
            "complexType": {
                "original": "any",
                "resolved": "any",
                "references": {}
            }
        }, {
            "method": "fwBlur",
            "name": "fwBlur",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": "Triggered when the list box loses focus."
            },
            "complexType": {
                "original": "any",
                "resolved": "any",
                "references": {}
            }
        }]; }
    static get methods() { return {
        "getSelectedItem": {
            "complexType": {
                "signature": "() => Promise<any>",
                "parameters": [],
                "references": {
                    "Promise": {
                        "location": "global"
                    }
                },
                "return": "Promise<any>"
            },
            "docs": {
                "text": "",
                "tags": []
            }
        },
        "setSelectedValues": {
            "complexType": {
                "signature": "(values: string[]) => Promise<any>",
                "parameters": [{
                        "tags": [],
                        "text": ""
                    }],
                "references": {
                    "Promise": {
                        "location": "global"
                    }
                },
                "return": "Promise<any>"
            },
            "docs": {
                "text": "",
                "tags": []
            }
        }
    }; }
    static get elementRef() { return "host"; }
    static get watchers() { return [{
            "propName": "value",
            "methodName": "keyChanged"
        }, {
            "propName": "options",
            "methodName": "optionsChangedHandler"
        }]; }
    static get listeners() { return [{
            "name": "fwSelected",
            "method": "fwSelectedHandler",
            "target": undefined,
            "capture": false,
            "passive": false
        }, {
            "name": "fwClosed",
            "method": "fwCloseHandler",
            "target": undefined,
            "capture": false,
            "passive": false
        }, {
            "name": "keydown",
            "method": "onKeyDonw",
            "target": undefined,
            "capture": false,
            "passive": false
        }]; }
}
