// @freshworks/crayons/dist/collection/components/datepicker/datepicker.js
import { h } from "@stencil/core";
import moment from 'moment-mini';
import { renderHiddenField } from '../../utils/utils';
const weekDay = ['S', 'M', 'T', 'W', 'T', 'F', 'S'];
const monthArr = [
    'Jan',
    'Feb',
    'Mar',
    'Apr',
    'May',
    'Jun',
    'Jul',
    'Aug',
    'Sep',
    'Oct',
    'Nov',
    'Dec',
];
export class Datepicker {
    constructor() {
        /**
         *   Type of date selection enabled for the calendar. If the value is range, a user can select a date range in the calendar.
         */
        this.mode = 'single date';
        /**
         *   Format in which the date values selected in the calendar are populated in the input box and saved when the form data is saved.
         */
        this.dateFormat = 'DD-MM-YYYY';
        /**
         * Name of the component, saved as part of form data.
         */
        this.name = '';
        this.getSupportedYears = () => {
            const yearsArr = [];
            let year = 1970;
            while (year < 2050) {
                yearsArr.push(year.toString());
                year++;
            }
            return yearsArr;
        };
        this.getDayDetails = args => {
            const date = args.index - args.firstDay;
            const day = args.index % 7;
            let prevMonth = args.month - 1;
            let prevYear = Number(args.year);
            if (prevMonth < 0) {
                prevMonth = 11;
                prevYear--;
            }
            const prevMonthNumberOfDays = moment([prevYear, prevMonth]).daysInMonth() || 0;
            const _date = (date < 0
                ? prevMonthNumberOfDays + date
                : date % args.numberOfDays) + 1;
            const month = this._getValidDateInMonth(date, args);
            const timestamp = moment([args.year, args.month, _date]).valueOf();
            return { date: _date, day, month, timestamp };
        };
        this.getMonthDetails = (year, month) => {
            const firstDay = (new Date(year, month)).getDay();
            const numberOfDays = moment([year, month]).daysInMonth() || 0;
            const monthArray = [];
            const rows = 6;
            let currentDay;
            let index = 0;
            const cols = 7;
            for (let row = 0; row < rows; row++) {
                for (let col = 0; col < cols; col++) {
                    currentDay = this.getDayDetails({ index, numberOfDays, firstDay, year, month });
                    monthArray.push(currentDay);
                    index++;
                }
            }
            return monthArray;
        };
        this.setMonth = offset => {
            let year = Number(this.year);
            let month = this.month + offset;
            if (month === -1) {
                month = 11;
                year--;
            }
            else if (month === 12) {
                month = 0;
                year++;
            }
            this.year = year.toString();
            this.month = month;
            this.toMonth = this.month === 11 ? 0 : this.month + 1;
            this.toYear = this.toMonth === 0 ? this.yearCalculation(this.year, 1) : this.year;
            this.monthDetails = this.getMonthDetails(year, month);
            this.nextMonthDetails = this.month === 11
                ? this.getMonthDetails(this.yearCalculation(this.year, 1), 0)
                : this.getMonthDetails(this.year, this.month + 1);
        };
        this.isCurrentDay = day => {
            return day.timestamp === this.todayTimestamp;
        };
        this.isSelectedDay = ({ timestamp }) => {
            return timestamp === this.selectedDay || timestamp === this.startDate || timestamp === this.endDate;
        };
        this.isInRange = ({ timestamp }) => {
            const { endDate } = this;
            const { startDate } = this;
            return startDate && endDate && ((timestamp >= startDate && timestamp <= endDate));
        };
        this.onDateClick = ({ timestamp }) => {
            if (this.showSingleDatePicker()) {
                this.selectedDay = timestamp;
            }
            else if (this.showDateRangePicker()) {
                this.handleRangeSelection(timestamp);
                this.dateHovered = '';
            }
        };
    }
    displayDatePicker() {
        this.showDatePicker = true;
    }
    handleButtonClick(e) {
        const isUpdateRange = e.composedPath()[0].classList.value.includes('update-range-value');
        const isUpdateDate = e.composedPath()[0].classList.value.includes('update-date-value');
        if (isUpdateRange) {
            this.startDateFormatted = moment(this.startDate).format(this.dateFormat);
            this.endDateFormatted = moment(this.endDate).format(this.dateFormat);
            if (this.startDate && this.endDate) {
                this.value = this.startDateFormatted + ' To ' + this.endDateFormatted;
            }
            this.fromDate = this.startDateFormatted;
            this.toDate = this.endDateFormatted;
            this.fwChange.emit({ fromDate: this.startDateFormatted, toDate: this.endDateFormatted });
        }
        else if (isUpdateDate) {
            this.value = this.selectedDay ? moment(this.selectedDay).format(this.dateFormat) : '';
            this.fwChange.emit(this.value);
        }
        this.showDatePicker = false;
    }
    /* Listener to handle Month Year dropdown changes
    */
    handleMonthYearDropDownSelection(e) {
        if (e.path[0].tagName !== 'FW-DATEPICKER') {
            e.stopPropagation();
        }
        const newValue = e.detail && e.detail.value;
        if (!newValue) {
            return;
        }
        if (this.mode === 'range') {
            this.handleDateRangeDropDownUpdate(e, newValue);
            this.nextMonthDetails = this.getMonthDetails(this.toYear, this.toMonth);
        }
        else {
            this.handleSingleDateDropDownUpdate(e, newValue);
        }
        this.monthDetails = this.getMonthDetails(this.year, this.month);
    }
    handleSingleDateDropDownUpdate(e, newValue) {
        const isMonthUpdate = e.composedPath()[0].classList.value.includes('single-month-selector');
        const isYearUpdate = e.composedPath()[0].classList.value.includes('single-year-selector');
        if (isMonthUpdate) {
            this.month = monthArr.indexOf(newValue);
        }
        else if (isYearUpdate) {
            this.year = newValue;
        }
    }
    handleDateRangeDropDownUpdate(e, newValue) {
        const isFromMonthUpdate = e.composedPath()[0].classList.value.includes('from-month-selector');
        const isFromYearUpdate = e.composedPath()[0].classList.value.includes('from-year-selector');
        const isToMonthUpdate = e.composedPath()[0].classList.value.includes('to-month-selector');
        const isToYearUpdate = e.composedPath()[0].classList.value.includes('to-year-selector');
        if (isFromMonthUpdate) {
            this.month = monthArr.indexOf(newValue);
            if (this.month === 11) {
                this.toMonth = 0;
                this.toYear = this.yearCalculation(this.year, 1);
            }
            else {
                this.toMonth = this.month + 1;
            }
        }
        else if (isFromYearUpdate) {
            this.year = newValue;
            this.toYear = this.month === 11 ? this.yearCalculation(this.year, 1) : this.year;
        }
        else if (isToMonthUpdate) {
            this.toMonth = monthArr.indexOf(newValue);
            if (this.toMonth === 0) {
                this.month = 11;
                this.year = this.yearCalculation(this.toYear, -1);
            }
            else {
                this.month = this.toMonth - 1;
            }
        }
        else if (isToYearUpdate) {
            this.toYear = newValue;
            this.year = this.toMonth === 0 ? this.yearCalculation(this.toYear, -1) : this.toYear;
        }
    }
    yearCalculation(year, offset) {
        const resultYear = Number(year) + offset;
        return resultYear.toString();
    }
    componentWillLoad() {
        this.year = moment().year().toString();
        this.month = moment().month();
        this.toMonth = this.month === 11 ? 0 : this.month + 1;
        this.toYear = this.toMonth === 0 ? this.yearCalculation(this.year, 1) : this.year;
        this.monthDetails = this.getMonthDetails(this.year, this.month);
        this.todayTimestamp = moment().startOf('date').valueOf();
        this.setInitalValues();
    }
    setInitalValues() {
        this.nextMonthDetails = this.month === 11
            ? this.getMonthDetails(this.yearCalculation(this.year, 1), 0)
            : this.getMonthDetails(this.year, this.month + 1);
        this.placeholder = this.placeholder || (this.mode === 'range'
            ? 'Select Date Range'
            : 'Select Date');
        this.supportedYears = this.getSupportedYears();
        this.selectedDay = this.value !== undefined ? moment(this.value, this.dateFormat).valueOf() : undefined;
        this.startDate = this.fromDate !== undefined ? moment(this.fromDate, this.dateFormat).valueOf() : undefined;
        this.endDate = this.toDate !== undefined ? moment(this.toDate, this.dateFormat).valueOf() : undefined;
        if (this.mode === 'range' && this.startDate && this.endDate) {
            const formattedFromDate = moment(this.startDate).format(this.dateFormat);
            const formattedToDate = moment(this.endDate).format(this.dateFormat);
            this.value = `${formattedFromDate} To ${formattedToDate}`;
        }
    }
    _getValidDateInMonth(date, args) {
        if (this.minDate !== undefined && this.maxDate !== undefined) {
            if (date < 0) {
                return -1;
            }
            const dateFormat = this.dateFormat || 'DD-MM-YYYY';
            const minDate = moment(this.minDate, dateFormat);
            const maxDate = moment(this.maxDate, dateFormat);
            const argDate = moment([args.year, args.month, date + 1]);
            const isValid = minDate.valueOf() <= argDate.valueOf() && argDate.valueOf() <= maxDate.valueOf();
            return !isValid
                ? -1
                : date >= args.numberOfDays
                    ? 1
                    : 0;
        }
        return date < 0
            ? -1
            : date >= args.numberOfDays
                ? 1
                : 0;
    }
    handleDateHover(day) {
        if (this.startDate && !this.endDate) {
            if (this.startDate > day.timestamp) {
                this.endDate = this.startDate;
                this.startDate = undefined;
            }
            this.dateHovered = day.timestamp;
        }
        else if (!this.startDate && this.endDate) {
            if (this.endDate < day.timestamp) {
                this.startDate = this.endDate;
                this.endDate = undefined;
            }
            this.dateHovered = day.timestamp;
        }
    }
    isHoverInRange({ timestamp }) {
        const { startDate, endDate, dateHovered } = this;
        const startDateCondtion = startDate && dateHovered && (timestamp <= dateHovered && timestamp >= startDate);
        const endDateCondition = endDate && dateHovered && (timestamp >= dateHovered && timestamp <= endDate);
        return startDateCondtion || endDateCondition;
    }
    handleRangeSelection(timestamp) {
        if (this.startDate && this.endDate) {
            this.endDate = undefined;
            this.startDate = timestamp;
        }
        else if (this.startDate && !this.endDate) {
            if (timestamp > this.startDate) {
                this.endDate = timestamp;
            }
            else if (timestamp < this.startDate) {
                this.endDate = this.startDate;
                this.startDate = timestamp;
            }
        }
        else if (!this.startDate && this.endDate) {
            this.startDate = timestamp;
        }
        else if (!this.startDate && !this.endDate) {
            this.startDate = timestamp;
        }
    }
    getCellStyle(day) {
        let cellStyle = 'c-day-container';
        if (day.month !== 0) {
            cellStyle += ' disabled';
        }
        if (this.isCurrentDay(day)) {
            cellStyle += ' highlight';
        }
        if (this.isSelectedDay(day) || day.timestamp === this.dateHovered) {
            cellStyle += ' highlight-blue';
        }
        if (this.isInRange(day) || this.isHoverInRange(day)) {
            cellStyle += ' highlight-range';
        }
        if (day.timestamp === this.startDate) {
            cellStyle += ' start-day';
        }
        if (day.timestamp === this.endDate) {
            cellStyle += ' end-day';
        }
        if (this.startDate && this.dateHovered < this.startDate && day.timestamp === this.dateHovered) {
            cellStyle += ' start-day';
        }
        else if (this.endDate && this.dateHovered < this.endDate && day.timestamp === this.dateHovered) {
            cellStyle += ' start-day';
        }
        else if (this.startDate && this.dateHovered > this.startDate && day.timestamp === this.dateHovered) {
            cellStyle += ' end-day';
        }
        return cellStyle;
    }
    renderCalendar(monthDetails) {
        const days = monthDetails.map((day, index) => {
            return (h("div", { class: this.getCellStyle(day), key: index },
                h("div", { class: "cdc-day" },
                    h("span", { onClick: () => this.onDateClick(day), onMouseOver: () => this.handleDateHover(day) }, day.date))));
        });
        return (h("div", { class: "c-container" },
            h("div", { class: "cc-head" }, weekDay.map((day, index) => h("div", { key: index, class: "cch-name" }, day))),
            h("div", { class: "cc-body" }, days)));
    }
    showSingleDatePicker() {
        return this.showDatePicker && this.mode !== 'range';
    }
    showDateRangePicker() {
        return this.showDatePicker && this.mode === 'range';
    }
    render() {
        const { host, name, value } = this;
        renderHiddenField(host, name, value);
        return (h("div", null,
            h("div", { onClick: () => this.showDatePicker = false, class: this.showDatePicker ? 'overlay-show' : 'overlay-hide' }),
            h("fw-input", { value: this.value, class: "date-input", placeholder: this.placeholder, readonly: true }),
            this.showSingleDatePicker()
                ? (h("div", { class: "datepicker" },
                    h("div", { class: "mdp-container" },
                        h("div", { class: "mdpc-head" },
                            h("div", { class: "mdpch-button" },
                                h("div", { class: "mdpchb-inner", onClick: () => this.setMonth(-1) },
                                    h("span", { class: "mdpchbi-left-arrow" }))),
                            h("div", { class: "mdpch-container" },
                                h("span", { class: "mdpchc-month" },
                                    h("fw-select", { class: "single-month-selector", readonly: true, value: monthArr[this.month] }, monthArr.map((month, i) => h("fw-select-option", { value: month, key: i, selected: month === monthArr[this.month] }, month)))),
                                h("span", { class: "mdpchc-year" },
                                    h("fw-select", { class: "single-year-selector", readonly: true, value: this.year }, this.supportedYears.map((year, i) => h("fw-select-option", { value: year, key: i, selected: year === this.year }, year))))),
                            h("div", { class: "mdpch-button-right" },
                                h("div", { class: "mdpchb-inner", onClick: () => this.setMonth(1) },
                                    h("span", { class: "mdpchbi-right-arrow" })))),
                        h("div", { class: "mdpc-body" }, this.renderCalendar(this.monthDetails))),
                    h("div", { class: "mdpc-footer" },
                        h("fw-button", { color: "primary", class: "update-date-value" }, "Update"),
                        h("fw-button", { color: "secondary", class: "close-date-picker" }, "Cancel"))))
                : '',
            this.showDateRangePicker()
                ? (h("div", { class: "daterangepicker" },
                    h("div", { class: "mdp-range-container" },
                        h("div", { class: "mdpc-head" },
                            h("div", { class: "mdpch-button" },
                                h("div", { class: "mdpchb-inner", onClick: () => this.setMonth(-1) },
                                    h("span", { class: "mdpchbi-left-arrow" }))),
                            h("div", { class: "mdpch-container" },
                                h("span", { class: "mdpchc-month" },
                                    h("fw-select", { class: "from-month-selector", readonly: true, value: monthArr[this.month] }, monthArr.map((month, i) => h("fw-select-option", { value: month, key: i, selected: month === monthArr[this.month] }, month)))),
                                h("span", { class: "mdpchc-year" },
                                    h("fw-select", { class: "from-year-selector", readonly: true, value: this.year }, this.supportedYears.map((year, i) => h("fw-select-option", { value: year, key: i, selected: year === this.year }, year))))),
                            h("div", { class: "mdpch-button-right" },
                                h("div", { class: "mdpchb-inner", onClick: () => this.setMonth(1) },
                                    h("span", { class: "mdpchbi-right-arrow" }))),
                            h("div", { class: "mdpch-container-right" },
                                h("span", { class: "mdpchc-year" },
                                    h("fw-select", { class: "to-year-selector", readonly: true, value: this.toYear }, this.supportedYears.map((year, i) => h("fw-select-option", { value: year, key: i, selected: year === this.toYear }, year)))),
                                h("span", { class: "mdpchc-month" },
                                    h("fw-select", { class: "to-month-selector", readonly: true, value: monthArr[this.toMonth] }, monthArr.map((month, i) => h("fw-select-option", { value: month, key: i, selected: month === monthArr[this.toMonth] }, month)))))),
                        h("div", { class: "body-container" },
                            h("div", { class: "mdpc-body" }, this.renderCalendar(this.monthDetails)),
                            h("div", { class: "mdpc-body mdpc-body-right" }, this.renderCalendar(this.nextMonthDetails)))),
                    h("div", { class: "mdpc-range-footer" },
                        h("fw-button", { color: "primary", class: "update-range-value" }, "Update"),
                        h("fw-button", { color: "secondary", class: "close-date-picker" }, "Cancel"))))
                : ''));
    }
    static get is() { return "fw-datepicker"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() { return {
        "$": ["datepicker.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["datepicker.css"]
    }; }
    static get properties() { return {
        "mode": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "'single date' | 'range'",
                "resolved": "\"range\" | \"single date\"",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Type of date selection enabled for the calendar. If the value is range, a user can select a date range in the calendar."
            },
            "attribute": "mode",
            "reflect": false,
            "defaultValue": "'single date'"
        },
        "minDate": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Earliest date a user can select in the calendar, if mode is range."
            },
            "attribute": "min-date",
            "reflect": false
        },
        "maxDate": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Latest date a user can select in the calendar, if mode is range."
            },
            "attribute": "max-date",
            "reflect": false
        },
        "fromDate": {
            "type": "string",
            "mutable": true,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Starting date of the date range that is preselected in the calendar, if mode is range. Must be a date later than the min-date value."
            },
            "attribute": "from-date",
            "reflect": false
        },
        "toDate": {
            "type": "string",
            "mutable": true,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Ending date of the date range that is preselected in the calendar, if mode is range. Must be a date earlier than the max-date value."
            },
            "attribute": "to-date",
            "reflect": false
        },
        "dateFormat": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Format in which the date values selected in the calendar are populated in the input box and saved when the form data is saved."
            },
            "attribute": "date-format",
            "reflect": false,
            "defaultValue": "'DD-MM-YYYY'"
        },
        "value": {
            "type": "string",
            "mutable": true,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Date that is preselected in the calendar, if mode is single date or undefined."
            },
            "attribute": "value",
            "reflect": false
        },
        "name": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Name of the component, saved as part of form data."
            },
            "attribute": "name",
            "reflect": false,
            "defaultValue": "''"
        },
        "placeholder": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Text displayed in the input box before a user selects a date or date range."
            },
            "attribute": "placeholder",
            "reflect": false
        }
    }; }
    static get states() { return {
        "showDatePicker": {},
        "year": {},
        "toYear": {},
        "monthDetails": {},
        "nextMonthDetails": {},
        "month": {},
        "todayTimestamp": {},
        "selectedDay": {},
        "startDate": {},
        "endDate": {},
        "startDateFormatted": {},
        "endDateFormatted": {},
        "dateHovered": {},
        "supportedYears": {},
        "toMonth": {}
    }; }
    static get events() { return [{
            "method": "fwChange",
            "name": "fwChange",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": "Triggered when the update button clicked"
            },
            "complexType": {
                "original": "any",
                "resolved": "any",
                "references": {}
            }
        }]; }
    static get elementRef() { return "host"; }
    static get listeners() { return [{
            "name": "fwFocus",
            "method": "displayDatePicker",
            "target": undefined,
            "capture": false,
            "passive": false
        }, {
            "name": "fwClick",
            "method": "handleButtonClick",
            "target": undefined,
            "capture": false,
            "passive": false
        }, {
            "name": "fwChange",
            "method": "handleMonthYearDropDownSelection",
            "target": undefined,
            "capture": false,
            "passive": false
        }]; }
}
