// @freshworks/crayons/dist/collection/components/modal/modal.js
import { h } from "@stencil/core";
export class Modal {
    constructor() {
        /**
         * The icon to be displayed with the title
         */
        this.icon = '';
        /**
         * The text for the success button
         */
        this.successText = 'OK';
        /**
         * The text for the cancel button
         */
        this.cancelText = 'Cancel';
        /**
         * Size of the modal
         */
        this.size = 'standard';
        /**
         * Enable custom footer
         */
        this.customFooter = false;
        /**
         * Hides the footer
         */
        this.hideFooter = false;
        /**
         * Toggle the visibility of the modal
         */
        this.visible = false;
    }
    visibilityChange(isVisible) {
        if (!isVisible) {
            this.fwClosed.emit();
        }
    }
    closeModal() {
        this.visible = false;
    }
    renderIcon() {
        return h("fw-icon", { class: "icon", name: this.icon, size: 16 });
    }
    renderFooter() {
        return this.customFooter
            ? h("slot", { name: "footer" })
            : (h("span", null,
                h("fw-button", { color: "secondary", onClick: () => this.closeModal() }, this.cancelText),
                h("fw-button", { color: "primary", onClick: () => this.fwAction.emit() }, this.successText)));
    }
    render() {
        return (h("div", { class: { 'modal-container': true, 'visible': this.visible } },
            h("div", { class: { 'modal': true, [this.size]: true } },
                h("div", { class: "modal-header-container" },
                    h("div", { class: "modal-header" },
                        this.icon !== '' ? this.renderIcon() : '',
                        h("div", null,
                            this.titleText,
                            h("div", { class: "description" }, this.description))),
                    h("button", { class: "close-btn", onClick: () => this.closeModal() },
                        h("fw-icon", { name: "cross-big" }))),
                h("div", { class: "content" },
                    h("slot", null)),
                this.hideFooter
                    ? ''
                    : (h("div", { class: "modal-footer-container" },
                        h("div", { class: "modal-footer" }, this.renderFooter()))))));
    }
    static get is() { return "fw-modal"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() { return {
        "$": ["modal.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["modal.css"]
    }; }
    static get properties() { return {
        "titleText": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "The title text to be displayed on the modal"
            },
            "attribute": "title-text",
            "reflect": false
        },
        "description": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "The title text to be displayed on the modal"
            },
            "attribute": "description",
            "reflect": false
        },
        "icon": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "The icon to be displayed with the title"
            },
            "attribute": "icon",
            "reflect": false,
            "defaultValue": "''"
        },
        "successText": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "The text for the success button"
            },
            "attribute": "success-text",
            "reflect": false,
            "defaultValue": "'OK'"
        },
        "cancelText": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "The text for the cancel button"
            },
            "attribute": "cancel-text",
            "reflect": false,
            "defaultValue": "'Cancel'"
        },
        "size": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "'standard' | 'small' | 'large'",
                "resolved": "\"large\" | \"small\" | \"standard\"",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Size of the modal"
            },
            "attribute": "size",
            "reflect": false,
            "defaultValue": "'standard'"
        },
        "customFooter": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Enable custom footer"
            },
            "attribute": "custom-footer",
            "reflect": false,
            "defaultValue": "false"
        },
        "hideFooter": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Hides the footer"
            },
            "attribute": "hide-footer",
            "reflect": false,
            "defaultValue": "false"
        },
        "visible": {
            "type": "boolean",
            "mutable": true,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Toggle the visibility of the modal"
            },
            "attribute": "visible",
            "reflect": true,
            "defaultValue": "false"
        }
    }; }
    static get events() { return [{
            "method": "fwAction",
            "name": "fwAction",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": "Triggered when the default action button is clicked."
            },
            "complexType": {
                "original": "void",
                "resolved": "void",
                "references": {}
            }
        }, {
            "method": "fwClosed",
            "name": "fwClosed",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": "Triggered when modal is closed."
            },
            "complexType": {
                "original": "void",
                "resolved": "void",
                "references": {}
            }
        }]; }
    static get watchers() { return [{
            "propName": "visible",
            "methodName": "visibilityChange"
        }]; }
}
