// @freshworks/crayons/dist/collection/components/radio/radio.js
import { Host, h } from "@stencil/core";
export class Radio {
    constructor() {
        /**
         * Sets the state to selected. If the attribute’s value is undefined, the value is set to false.
         */
        this.checked = false;
        /**
         * Disables the component on the interface. If the attribute’s value is undefined, the value is set to false.
         */
        this.disabled = false;
        /**
         * Label displayed on the interface, for the component.
         */
        this.label = '';
        /**
         * Identifier corresponding to the component, that is saved when the form data is saved.
         */
        this.value = '';
        /**
         * Name of the component, saved as part of form data.
         */
        this.name = '';
    }
    componentDidLoad() {
        this.radio.checked = this.checked;
        this.radio.disabled = this.disabled;
    }
    checkChanged(isChecked) {
        if (!this.disabled) {
            if (isChecked) {
                this.fwSelect.emit({
                    value: this.value,
                    checked: true,
                });
            }
            else {
                this.fwDeselect.emit();
            }
        }
        this.radio.checked = isChecked;
    }
    disabledChanged(isDisabled) {
        this.radio.disabled = isDisabled;
    }
    onFocus() {
        this.fwFocus.emit();
    }
    onBlur() {
        this.fwBlur.emit();
    }
    toggle() {
        if (!this.disabled) {
            this.checked = !this.checked;
        }
    }
    render() {
        return (h(Host, { class: "radio-container", onClick: () => this.toggle(), role: "radio", tabIndex: "0", "aria-disabled": this.disabled ? 'true' : 'false', "aria-checked": `${this.checked}`, onFocus: () => this.onFocus(), onBlur: () => this.onBlur() },
            h("input", { type: "radio", ref: el => this.radio = el }),
            h("label", null,
                h("span", { class: "text" },
                    h("slot", null)),
                h("br", null),
                this.label !== ''
                    ? h("span", { class: "label-field" }, this.label)
                    : '')));
    }
    static get is() { return "fw-radio"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() { return {
        "$": ["radio.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["radio.css"]
    }; }
    static get properties() { return {
        "checked": {
            "type": "boolean",
            "mutable": true,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Sets the state to selected. If the attribute\u2019s value is undefined, the value is set to false."
            },
            "attribute": "checked",
            "reflect": true,
            "defaultValue": "false"
        },
        "disabled": {
            "type": "boolean",
            "mutable": true,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Disables the component on the interface. If the attribute\u2019s value is undefined, the value is set to false."
            },
            "attribute": "disabled",
            "reflect": true,
            "defaultValue": "false"
        },
        "label": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Label displayed on the interface, for the component."
            },
            "attribute": "label",
            "reflect": false,
            "defaultValue": "''"
        },
        "value": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Identifier corresponding to the component, that is saved when the form data is saved."
            },
            "attribute": "value",
            "reflect": false,
            "defaultValue": "''"
        },
        "name": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Name of the component, saved as part of form data."
            },
            "attribute": "name",
            "reflect": false,
            "defaultValue": "''"
        }
    }; }
    static get events() { return [{
            "method": "fwSelect",
            "name": "fwSelect",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": "Triggered when the radio button in focus is selected."
            },
            "complexType": {
                "original": "any",
                "resolved": "any",
                "references": {}
            }
        }, {
            "method": "fwDeselect",
            "name": "fwDeselect",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": "Triggered when the radio button in focus is cleared."
            },
            "complexType": {
                "original": "any",
                "resolved": "any",
                "references": {}
            }
        }, {
            "method": "fwFocus",
            "name": "fwFocus",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": "Triggered when the radio button comes into focus."
            },
            "complexType": {
                "original": "void",
                "resolved": "void",
                "references": {}
            }
        }, {
            "method": "fwBlur",
            "name": "fwBlur",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": "Triggered when the radio button loses focus."
            },
            "complexType": {
                "original": "void",
                "resolved": "void",
                "references": {}
            }
        }]; }
    static get watchers() { return [{
            "propName": "checked",
            "methodName": "checkChanged"
        }, {
            "propName": "disabled",
            "methodName": "disabledChanged"
        }]; }
}
