// @freshworks/crayons/dist/collection/components/toast/toast.js
import { Host, getAssetPath, h } from "@stencil/core";
export class Toast {
    constructor() {
        /**
         * visibility state of toast
         */
        this.isOpen = false;
        /**
         * Icon svg html content
         */
        this.svgHTML = '';
        /**
         * To indicate toast close timeout
         */
        this.isTimedOut = false;
        /**
         * To add close animation class to toast
         */
        this.fadeOut = false;
        /**
         * State icon size
         */
        this.iconSize = 14;
        /**
         * Type of the toast - success,failure, warning, inprogress
         */
        this.type = 'warning';
        /**
         * Time duration of the toast visibility
         */
        this.timeout = 4000;
        /**
         *  The Content of the action link
         */
        this.actionLinkText = '';
        /**
         *  won't close automatically
         */
        this.sticky = false;
        /**
         *  position of the toast notification in screen
         */
        this.position = 'top-center';
    }
    async getSVGHTML(iconName) {
        const response = await fetch(getAssetPath(`toast-assets/icons/${iconName}.svg`));
        const data = await response.text();
        return data;
    }
    setSVGState(iconName) {
        if (this.type === 'inprogress') {
            return;
        }
        this.getSVGHTML(iconName).then(res => {
            this.svgHTML = res;
        }).catch();
    }
    componentWillLoad() {
        this.setSVGState(this.type);
    }
    async trigger(configs) {
        if (this.isOpen || this.timerId) {
            await this.closeToast();
        }
        Object.keys(configs).forEach(key => {
            this[key] = configs[key];
        });
        this.fadeOut = false;
        this.isOpen = true;
        this.isTimedOut = false;
        this.timerId = setTimeout(async () => {
            if (!this.sticky) {
                if (!this.pauseOnHover || this.pauseOnHover && !this.isMouseHovered) {
                    await this.closeToast();
                }
                this.isTimedOut = true;
            }
        }, this.timeout);
    }
    async mouseHover(value = false) {
        this.isMouseHovered = value;
        if (this.isTimedOut && !this.isMouseHovered) {
            await this.closeToast();
        }
    }
    closingAnimation() {
        this.fadeOut = true;
        return new Promise(resolve => setTimeout(() => { this.isOpen = false; resolve(); }, 500));
    }
    async closeToast() {
        if (this.timerId) {
            clearTimeout(this.timerId);
        }
        await this.closingAnimation();
    }
    render() {
        return (h(Host, { class: `toast ${this.position} ${this.type} ${this.isOpen ? 'is-open' : ''} ${this.fadeOut ? 'fade-out' : ''}`, "aria-hidden": this.isOpen ? 'false' : 'true', onmouseover: () => this.mouseHover(true), onmouseout: () => this.mouseHover(false) },
            h("div", null,
                this.type === 'inprogress' ? (h("fw-spinner", { class: "icon" })) : (h("fw-icon", { class: "icon", size: this.iconSize, name: this.type })),
                h("span", { class: "content" }, this.content),
                h("fw-icon", { size: 7, color: "#000", name: "cross", class: "cross", onClick: () => this.closeToast() }),
                this.actionLinkText.length > 0 ? (h("div", { class: "action-link", onClick: () => this.fwLinkClick.emit() }, this.actionLinkText)) : '')));
    }
    static get is() { return "fw-toast"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() { return {
        "$": ["toast.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["toast.css"]
    }; }
    static get assetsDirs() { return ["toast-assets"]; }
    static get properties() { return {
        "type": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "'success' | 'error' | 'warning' | 'inprogress'",
                "resolved": "\"error\" | \"inprogress\" | \"success\" | \"warning\"",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Type of the toast - success,failure, warning, inprogress"
            },
            "attribute": "type",
            "reflect": false,
            "defaultValue": "'warning'"
        },
        "timeout": {
            "type": "number",
            "mutable": false,
            "complexType": {
                "original": "number",
                "resolved": "number",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Time duration of the toast visibility"
            },
            "attribute": "timeout",
            "reflect": false,
            "defaultValue": "4000"
        },
        "content": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "The content to be diaplyed in toast"
            },
            "attribute": "content",
            "reflect": false
        },
        "actionLinkText": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "The Content of the action link"
            },
            "attribute": "action-link-text",
            "reflect": false,
            "defaultValue": "''"
        },
        "sticky": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "won't close automatically"
            },
            "attribute": "sticky",
            "reflect": false,
            "defaultValue": "false"
        },
        "pauseOnHover": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Pause the toast from hiding on mouse hover"
            },
            "attribute": "pause-on-hover",
            "reflect": false
        },
        "position": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "'top-center' | 'top-left' | 'top-right'",
                "resolved": "\"top-center\" | \"top-left\" | \"top-right\"",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "position of the toast notification in screen"
            },
            "attribute": "position",
            "reflect": false,
            "defaultValue": "'top-center'"
        }
    }; }
    static get states() { return {
        "isOpen": {},
        "isMouseHovered": {},
        "svgHTML": {},
        "isTimedOut": {},
        "timerId": {},
        "fadeOut": {},
        "iconSize": {}
    }; }
    static get events() { return [{
            "method": "fwLinkClick",
            "name": "fwLinkClick",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": "Triggered when the action link clicked."
            },
            "complexType": {
                "original": "any",
                "resolved": "any",
                "references": {}
            }
        }]; }
    static get methods() { return {
        "trigger": {
            "complexType": {
                "signature": "(configs: object) => Promise<void>",
                "parameters": [{
                        "tags": [],
                        "text": ""
                    }],
                "references": {
                    "Promise": {
                        "location": "global"
                    }
                },
                "return": "Promise<void>"
            },
            "docs": {
                "text": "",
                "tags": []
            }
        }
    }; }
    static get watchers() { return [{
            "propName": "type",
            "methodName": "setSVGState"
        }]; }
}
