// @freshworks/crayons/dist/esm-es5/fw-modal.entry.js
import { r as registerInstance, c as createEvent, h } from './index-107459b7.js';
var Modal = /** @class */ (function () {
    function Modal(hostRef) {
        registerInstance(this, hostRef);
        /**
         * The icon to be displayed with the title
         */
        this.icon = '';
        /**
         * The text for the success button
         */
        this.successText = 'OK';
        /**
         * The text for the cancel button
         */
        this.cancelText = 'Cancel';
        /**
         * Size of the modal
         */
        this.size = 'standard';
        /**
         * Enable custom footer
         */
        this.customFooter = false;
        /**
         * Hides the footer
         */
        this.hideFooter = false;
        /**
         * Toggle the visibility of the modal
         */
        this.visible = false;
        this.fwAction = createEvent(this, "fwAction", 7);
        this.fwClosed = createEvent(this, "fwClosed", 7);
    }
    Modal.prototype.visibilityChange = function (isVisible) {
        if (!isVisible) {
            this.fwClosed.emit();
        }
    };
    Modal.prototype.closeModal = function () {
        this.visible = false;
    };
    Modal.prototype.renderIcon = function () {
        return h("fw-icon", { class: "icon", name: this.icon, size: 16 });
    };
    Modal.prototype.renderFooter = function () {
        var _this = this;
        return this.customFooter
            ? h("slot", { name: "footer" })
            : (h("span", null, h("fw-button", { color: "secondary", onClick: function () { return _this.closeModal(); } }, this.cancelText), h("fw-button", { color: "primary", onClick: function () { return _this.fwAction.emit(); } }, this.successText)));
    };
    Modal.prototype.render = function () {
        var _a;
        var _this = this;
        return (h("div", { class: { 'modal-container': true, 'visible': this.visible } }, h("div", { class: (_a = { 'modal': true }, _a[this.size] = true, _a) }, h("div", { class: "modal-header-container" }, h("div", { class: "modal-header" }, this.icon !== '' ? this.renderIcon() : '', h("div", null, this.titleText, h("div", { class: "description" }, this.description))), h("button", { class: "close-btn", onClick: function () { return _this.closeModal(); } }, h("fw-icon", { name: "cross-big" }))), h("div", { class: "content" }, h("slot", null)), this.hideFooter
            ? ''
            : (h("div", { class: "modal-footer-container" }, h("div", { class: "modal-footer" }, this.renderFooter()))))));
    };
    Object.defineProperty(Modal, "watchers", {
        get: function () {
            return {
                "visible": ["visibilityChange"]
            };
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Modal, "style", {
        get: function () { return ":host{--color-milk:#fff;--color-elephant-900:#12344d;--color-elephant-800:#264966;--color-elephant-700:#345c7c;--color-elephant-600:#447093;--color-smoke-700:#475867;--color-smoke-300:#92a2b1;--color-smoke-100:#cfd7df;--color-smoke-50:#ebeff3;--color-smoke-25:#f3f5f7;--color-jungle-800:#00795b;--color-jungle-500:#00a886;--color-jungle-100:#b4e5da;--color-jungle-50:#e0f5f1;--color-azure-800:#2c5cc5;--color-azure-100:#bbdcfe;--color-azure-50:#e5f2fd;--color-persimmon-900:#c82124;--color-persimmon-800:#d72d30;--color-persimmon-100:#ffd0d6;--color-persimmon-50:#ffecf0;--color-casablanca-700:#e86f25;--color-casablanca-300:#f8ab59;--color-casablanca-100:#fedcb3;--color-casablanca-50:#fef1e1;--border-color:var(--color-smoke-100);--border-success-color:var(--color-jungle-100);--border-info-color:var(--color-azure-100);--border-danger-color:var(--color-persimmon-100);--border-warning-color:var(--color-casablanca-100);--bg-dark:var(--color-elephant-900);--bg-success:var(--color-jungle-50);--bg-info:var(--color-azure-50);--bg-danger:var(--color-persimmon-50);--bg-warning:var(--color-casablanca-50);--radius:4px;--radius-small:2px}:host{--font-stack:-apple-system, blinkmacsystemfont, \"Segoe UI\", roboto, oxygen, ubuntu, cantarell, \"Open Sans\", \"Helvetica Neue\", sans-serif;--font-weight-400:400;--font-weight-300:300;--font-weight-500:500;--font-weight-600:600;--font-weight-700:700;--font-size-10:10px;--font-size-12:12px;--font-size-14:14px;--font-size-16:16px;--font-size-18:18px;--font-size-20:20px;--font-size-24:24px;--text-default:var(--color-elephant-900);--text-secondary:var(--color-smoke-700);--text-disabled:var(--color-smoke-300);--text-success:var(--color-jungle-800);--text-info:var(--color-azure-800);--text-danger:var(--color-persimmon-800);--text-warning:var(--color-casablanca-700);--text-link:var(--color-azure-800);--icon-primary:var(--color-smoke-700);--icon-primary-hover:var(--color-smoke-100)}:host{font-family:var(--font-stack);-webkit-font-smoothing:antialiased;-moz-osx-font-smoothing:grayscale;-webkit-box-sizing:border-box;box-sizing:border-box}.modal-container{height:100vh;width:100vw;position:fixed;top:0;right:0;bottom:0;left:0;display:none;z-index:990;background-color:rgba(18, 52, 77, 0.5);-ms-flex-align:center;align-items:center;-ms-flex-pack:center;justify-content:center;-webkit-transition:all 0.3s linear;transition:all 0.3s linear}.modal{position:relative;max-height:70vh;background:#fff;border-radius:4px;display:-ms-flexbox;display:flex;z-index:999;-ms-flex-direction:column;flex-direction:column;-webkit-animation:\"modal-entry\" 0.5s 1;animation:\"modal-entry\" 0.5s 1}.modal .modal-header-container{padding:32px 32px 12px;position:relative;background:#fff;border-radius:4px 4px 0 0;-webkit-transition:all 0.1s linear;transition:all 0.1s linear;display:-ms-flexbox;display:flex;-ms-flex-pack:justify;justify-content:space-between;-ms-flex:none;flex:none}.modal .modal-header-container .modal-header{font-size:18px;font-weight:600;margin:0;line-height:1.1;display:-ms-flexbox;display:flex}.modal .modal-header-container .modal-header .description{font-size:12px;font-weight:300;color:#475867;line-height:1.6}.modal .modal-header-container .icon{margin-right:8px}.modal .modal-header-container .close-btn{background-color:transparent;background-image:none;border:1px solid transparent;color:#264966;padding:4px 6px;min-width:16px;height:24px;position:absolute;top:8px;right:8px;-webkit-transition:all 0.3s;transition:all 0.3s}.modal .modal-header-container .close-btn:hover{background-color:#ebeff3;border-radius:4px;border-color:#ebeff3;cursor:pointer}.modal .content{padding:0 32px 32px;height:100%;width:auto;overflow:scroll}.modal .modal-footer-container{padding:12px 16px;-webkit-box-sizing:border-box;box-sizing:border-box;height:56px;background-color:#f3f5f7;border-radius:0 0 4px 4px;display:-ms-flexbox;display:flex;-ms-flex-align:center;align-items:center;-ms-flex-pack:end;justify-content:flex-end;-ms-flex:none;flex:none}.modal .modal-footer-container fw-button{margin-left:12px}.standard{width:512px}.small{width:424px}.large{width:800px}.visible{display:-ms-flexbox;display:flex}\@-webkit-keyframes modal-entry{0%{-webkit-transform:translateY(-10px);transform:translateY(-10px)}100%{-webkit-transform:translateY(0);transform:translateY(0)}}\@keyframes modal-entry{0%{-webkit-transform:translateY(-10px);transform:translateY(-10px)}100%{-webkit-transform:translateY(0);transform:translateY(0)}}"; },
        enumerable: true,
        configurable: true
    });
    return Modal;
}());
export { Modal as fw_modal };
