// @freshworks/crayons/dist/esm-es5/fw-tabs.entry.js
import { r as registerInstance, c as createEvent, h, g as getElement } from './index-107459b7.js';
var Tabs = /** @class */ (function () {
    function Tabs(hostRef) {
        registerInstance(this, hostRef);
        /**
         * The index of the activated Tab(Starts from 0)
         */
        this.activeTabIndex = 0;
        /**
         * Active class for tab container
         */
        this.activeChildClass = '';
        this.fwChange = createEvent(this, "fwChange", 7);
    }
    Tabs.prototype.changeIndex = function (newValue, oldValue) {
        var _a;
        if (((_a = this.tabs[newValue]) === null || _a === void 0 ? void 0 : _a.disabled) || newValue >= this.tabs.length) {
            this.activeTabIndex = 0;
        }
        else if (newValue !== oldValue) {
            this.displayTab(newValue);
        }
    };
    Tabs.prototype.init = function () {
        var _this = this;
        this.tabs = Array.from(this.el.querySelectorAll('fw-tab')).filter(function (node) { return node.parentNode === _this.el; });
        this.displayTab(this.activeTabIndex);
    };
    Tabs.prototype.displayTab = function (index) {
        var _a;
        this.fwChange.emit({ tabIndex: this.activeTabIndex });
        this.activeTabIndex = index;
        this.tabs = (_a = this.tabs) === null || _a === void 0 ? void 0 : _a.map(function (tab, i) {
            tab.style.display = index === i ? 'block' : 'none';
            return tab;
        });
    };
    Tabs.prototype.componentWillLoad = function () {
        this.init();
    };
    Tabs.prototype.connectedCallback = function () {
        var _this = this;
        this.mutationO = new MutationObserver(function () {
            _this.init();
        });
        this.mutationO.observe(this.el, {
            childList: true,
        });
    };
    Tabs.prototype.disconnectedCallback = function () {
        if (this.mutationO) {
            this.mutationO.disconnect();
            this.mutationO = undefined;
        }
    };
    Tabs.prototype.render = function () {
        var _this = this;
        return (h("div", { class: "tabs" }, h("ul", { role: "tablist", class: "tabs__items" }, this.tabs.map(function (tab, index) { return h("li", { onClick: function () { return tab.disabled ? '' : _this.activeTabIndex = index; }, class: "tabs__item" }, h("div", { id: '#tab-' + index, class: 'tabs__item__nav ' + (index === _this.activeTabIndex ? 'active' : '') + (tab.disabled ? 'disabled' : '') }, h("span", { class: "tab-title--tab-icon" }, tab.tabHeaderHtml
            ? h("span", { innerHTML: tab.tabHeaderHtml })
            : h("span", { class: "tab-title" }, tab.tabHeader)))); })), h("div", { class: "tabs__content" }, h("slot", null))));
    };
    Object.defineProperty(Tabs.prototype, "el", {
        get: function () { return getElement(this); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Tabs, "watchers", {
        get: function () {
            return {
                "activeTabIndex": ["changeIndex"]
            };
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Tabs, "style", {
        get: function () { return ":host{--color-milk:#fff;--color-elephant-900:#12344d;--color-elephant-800:#264966;--color-elephant-700:#345c7c;--color-elephant-600:#447093;--color-smoke-700:#475867;--color-smoke-300:#92a2b1;--color-smoke-100:#cfd7df;--color-smoke-50:#ebeff3;--color-smoke-25:#f3f5f7;--color-jungle-800:#00795b;--color-jungle-500:#00a886;--color-jungle-100:#b4e5da;--color-jungle-50:#e0f5f1;--color-azure-800:#2c5cc5;--color-azure-100:#bbdcfe;--color-azure-50:#e5f2fd;--color-persimmon-900:#c82124;--color-persimmon-800:#d72d30;--color-persimmon-100:#ffd0d6;--color-persimmon-50:#ffecf0;--color-casablanca-700:#e86f25;--color-casablanca-300:#f8ab59;--color-casablanca-100:#fedcb3;--color-casablanca-50:#fef1e1;--border-color:var(--color-smoke-100);--border-success-color:var(--color-jungle-100);--border-info-color:var(--color-azure-100);--border-danger-color:var(--color-persimmon-100);--border-warning-color:var(--color-casablanca-100);--bg-dark:var(--color-elephant-900);--bg-success:var(--color-jungle-50);--bg-info:var(--color-azure-50);--bg-danger:var(--color-persimmon-50);--bg-warning:var(--color-casablanca-50);--radius:4px;--radius-small:2px}:host{--font-stack:-apple-system, blinkmacsystemfont, \"Segoe UI\", roboto, oxygen, ubuntu, cantarell, \"Open Sans\", \"Helvetica Neue\", sans-serif;--font-weight-400:400;--font-weight-300:300;--font-weight-500:500;--font-weight-600:600;--font-weight-700:700;--font-size-10:10px;--font-size-12:12px;--font-size-14:14px;--font-size-16:16px;--font-size-18:18px;--font-size-20:20px;--font-size-24:24px;--text-default:var(--color-elephant-900);--text-secondary:var(--color-smoke-700);--text-disabled:var(--color-smoke-300);--text-success:var(--color-jungle-800);--text-info:var(--color-azure-800);--text-danger:var(--color-persimmon-800);--text-warning:var(--color-casablanca-700);--text-link:var(--color-azure-800);--icon-primary:var(--color-smoke-700);--icon-primary-hover:var(--color-smoke-100)}:host{font-family:var(--font-stack);-webkit-font-smoothing:antialiased;-moz-osx-font-smoothing:grayscale;-webkit-box-sizing:border-box;box-sizing:border-box}:host{--tab-header-font-size:var(--font-size-12);--tabs-color-primary:var(--color-azure-800)}.tabs__items{height:40px;list-style:none;width:100%;margin:0;padding:0}.tabs__item{float:left;position:relative}.tabs__item a{text-decoration:none;color:var(--color-smoke-300)}.tabs__item__nav{margin:0 12px;line-height:40px;color:var(--color-smoke-300);position:relative;overflow:hidden;display:block;padding:0 6px;cursor:pointer;font-size:var(--tab-header-font-size);font-weight:var(--font-weight-600)}.tabs__item__nav:hover::after{opacity:0.3;background:var(--color-smoke-300)}.tabs__item__nav.active{color:var(--tabs-color-primary);font-weight:var(--font-weight-600)}.tabs__item__nav.active .app-icon{fill:var(--tabs-color-primary)}.tabs__item__nav.active::after{opacity:1;background:var(--tabs-color-primary)}.tabs__item__nav.disabled{cursor:not-allowed;opacity:0.7}.tabs__item__nav::after{position:absolute;bottom:0;left:0;width:100%;height:3px;background:var(--tabs-color-primary);content:\"\";-webkit-transition:height transform 0.5s;transition:height transform 0.5s;opacity:0;border-radius:3px 3px 0 0}.tabs__content{clear:both;float:none}.tabs__content__pane{display:none;padding-top:5px}.tabs__content__pane.active{display:block}.tabs__content__pane--fade{opacity:0;-webkit-transition:opacity 0.15s linear;transition:opacity 0.15s linear}.tabs__content__pane--fade.in{opacity:1}"; },
        enumerable: true,
        configurable: true
    });
    return Tabs;
}());
export { Tabs as fw_tabs };
