// @freshworks/crayons/dist/esm/fw-input.entry.js
import { r as registerInstance, c as createEvent, h, H as Host, g as getElement } from './index-5957b9c5.js';
import { r as renderHiddenField } from './utils-926a70fd.js';

const Input = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.hasFocus = false;
        /**
         * Label displayed on the interface, for the component.
         */
        this.label = '';
        /**
         * Default value displayed in the input box.
         */
        this.value = '';
        /**
         * Type of value accepted as the input value. If a user enters a value other than the specified type, the input box is not populated.
         */
        this.type = 'text';
        /**
         * Specifies whether the browser can display suggestions to autocomplete the text value.
         */
        this.autocomplete = 'off';
        /**
         * Specifies whether the browser can auto focus the input field
         */
        this.autofocus = false;
        /**
         * Displays a right-justified clear icon in the text box. Clicking the icon clears the input text. If the attribute’s value is undefined, the value is set to false. For a read-only input box, the clear icon is not displayed unless a default value is specified for the input box.
         */
        this.clearInput = false;
        /**
         * Name of the component, saved as part of form data.
         */
        this.name = '';
        /**
         * Theme based on which the text box is styled.
         */
        this.state = 'normal';
        /**
         * Descriptive or instructional text displayed below the text box.
         */
        this.stateText = '';
        /**
         * If true, the user cannot enter a value in the input box. If the attribute’s value is undefined, the value is set to false.
         */
        this.readonly = false;
        /**
         * Specifies the input box as a mandatory field and displays an asterisk next to the label. If the attribute’s value is undefined, the value is set to false.
         */
        this.required = false;
        /**
         * Disables the component on the interface. If the attribute’s value is undefined, the value is set to false.
         */
        this.disabled = false;
        /**
         * Identifier of the icon that is displayed in the left side of the text box. The attribute’s value must be a valid svg file in the repo of icons (assets/icons).
         */
        this.iconLeft = undefined;
        /**
         * Identifier of the icon that is displayed in the right side of the text box. The attribute’s value must be a valid svg file in the repo of icons (assets/icons).
         */
        this.iconRight = undefined;
        this.onInput = (ev) => {
            const input = ev.target;
            if (input) {
                this.value = input.value || '';
            }
            this.fwInput.emit(ev);
        };
        this.onFocus = () => {
            this.hasFocus = true;
            this.fwFocus.emit();
        };
        this.onBlur = () => {
            this.hasFocus = false;
            this.fwBlur.emit();
        };
        this.clearTextInput = (ev) => {
            if (!this.readonly && !this.disabled && ev) {
                this.value = '';
                if (this.nativeInput) {
                    this.nativeInput.value = '';
                }
                this.fwInputClear.emit();
            }
        };
        this.fwChange = createEvent(this, "fwChange", 7);
        this.fwFocus = createEvent(this, "fwFocus", 7);
        this.fwBlur = createEvent(this, "fwBlur", 7);
        this.fwInput = createEvent(this, "fwInput", 7);
        this.fwInputClear = createEvent(this, "fwInputClear", 7);
    }
    watchHandler(newValue) {
        this.fwChange.emit({ value: newValue });
    }
    showClearButton() {
        return this.clearInput && !this.readonly && !this.disabled && this.value.length > 0;
    }
    getValue() {
        return this.value || '';
    }
    hasValue() {
        return this.getValue().length > 0;
    }
    /**
     * Sets focus on a specific `fw-input`. Use this method instead of the global `input.focus()`.
     */
    async setFocus() {
        if (this.nativeInput) {
            this.nativeInput.focus();
        }
    }
    render() {
        const { host, name, value } = this;
        renderHiddenField(host, name, value);
        return (h(Host, { "aria-disabled": this.disabled, class: {
                'has-value': this.hasValue(),
                'has-focus': this.hasFocus,
            } }, h("div", { class: "input-container" }, this.label !== '' ? h("label", { class: {
                'required': this.required,
            } }, this.label) : '', h("div", { class: {
                'input-container-inner': true,
                [this.state]: true,
                'left-icon': this.iconLeft !== undefined,
                'right-icon': this.iconRight !== undefined,
            } }, h("input", { ref: input => this.nativeInput = input, autoComplete: this.autocomplete, autoFocus: this.autofocus, disabled: this.disabled, name: this.name, placeholder: this.placeholder || '', minLength: this.minlength, maxLength: this.maxlength, readOnly: this.readonly, required: this.required, type: this.type, value: this.value, onInput: e => this.onInput(e), onBlur: this.onBlur, onFocus: this.onFocus }), this.iconLeft !== undefined ? h("fw-icon", { class: "icon left", name: this.iconLeft }) : '', this.iconRight !== undefined ? h("fw-icon", { class: "icon right", name: this.iconRight }) : '', this.showClearButton() ?
            h("div", { class: "clear-button", onClick: e => this.clearTextInput(e) }, h("svg", { xmlns: "http://www.w3.org/2000/svg", width: "32", height: "32", viewBox: "0 0 32 32", class: "clear-button-img" }, h("path", { d: "M17.992 16l8.796-8.796a1.409 1.409 0 0 0-1.992-1.992L16 14.008 7.204 5.212a1.409 1.409 0 0 0-1.992 1.992L14.008 16l-8.796 8.796a1.409 1.409 0 0 0 1.992 1.992L16 17.992l8.796 8.796a1.409 1.409 0 0 0 1.992-1.992L17.992 16z" }))) : ''), this.stateText !== '' ?
            h("span", { class: "help-block" }, this.stateText) : '')));
    }
    get host() { return getElement(this); }
    static get watchers() { return {
        "value": ["watchHandler"]
    }; }
    static get style() { return "/* Need to check with designer */\n/* Need to check with designer */\n:host {\n  font-family: -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, Oxygen, Ubuntu, Cantarell, \"Open Sans\", \"Helvetica Neue\", sans-serif;\n}\n\n*,\n::after,\n::before {\n  -webkit-box-sizing: border-box;\n  box-sizing: border-box;\n}\n\n.input-container {\n  margin-bottom: 16px;\n  width: inherit;\n  height: inherit;\n}\n\nlabel {\n  font-size: 12px;\n  color: #475867;\n  font-weight: 500;\n  margin-bottom: 0;\n  padding-bottom: 4px;\n  padding-left: 2px;\n  display: block;\n}\nlabel.required::after {\n  content: \"*\";\n  position: relative;\n  display: inline-block;\n  top: 2px;\n  font-size: 14px;\n  color: #d72d30;\n  padding-left: 2px;\n  font-weight: 700;\n}\n\n.input-container-inner {\n  display: block;\n  width: 100%;\n  position: relative;\n}\n.input-container-inner input {\n  width: 100%;\n  border: 0;\n  border: 1px solid #cfd7df;\n  margin: 5px 0 0;\n  border-radius: 4px;\n  padding: 4px 12px 5px;\n  resize: none;\n  background-color: #fff;\n  -webkit-box-shadow: none;\n  box-shadow: none;\n  min-height: 24px;\n  font-size: 12px;\n  font-weight: 500;\n  letter-spacing: 0;\n  line-height: 20px;\n  color: #183247;\n  -webkit-box-sizing: border-box;\n  box-sizing: border-box;\n  cursor: text;\n  display: inline-block;\n  font-family: inherit;\n}\n.input-container-inner input:hover {\n  border: 1px #475867 solid;\n  -webkit-transition: 0.2s linear;\n  transition: 0.2s linear;\n}\n.input-container-inner input:focus {\n  outline: none;\n  background: #fff;\n  background-color: #fff;\n  border: 1px solid transparent;\n  -webkit-box-shadow: 0 0 0 2px #2c5cc5;\n  box-shadow: 0 0 0 2px #2c5cc5;\n}\n.input-container-inner input[disabled] {\n  color: #92a2b1;\n  background-color: #f5f7f9;\n  border-style: solid;\n  pointer-events: none;\n}\n.input-container-inner input[disabled]:hover {\n  border: 1px solid #cfd7df;\n}\n.input-container-inner .clear-button {\n  top: 13px;\n  position: absolute;\n  right: 10px;\n  display: inline-block;\n}\n.input-container-inner .clear-button:hover {\n  cursor: pointer;\n  pointer-events: initial;\n}\n.input-container-inner .clear-button .clear-button-img {\n  width: 15px;\n  height: 15px;\n}\n.input-container-inner + .help-block {\n  font-family: -apple-system, blinkmacsystemfont, \"Segoe UI\", \"Roboto\", \"Helvetica Neue\", arial, sans-serif;\n  font-size: 11px;\n  margin-top: 3px;\n  color: #acb6be;\n  position: inherit;\n  margin-bottom: 0;\n  display: block;\n  padding-left: 2px;\n  -webkit-font-smoothing: antialiased;\n  -moz-osx-font-smoothing: grayscale;\n}\n.input-container-inner.error > input {\n  border-color: #d72d30;\n}\n.input-container-inner.error > input:focus {\n  -webkit-box-shadow: none;\n  box-shadow: none;\n  border-color: #d72d30;\n}\n.input-container-inner.error > input:hover {\n  border-color: #d72d30;\n}\n.input-container-inner.error + .help-block {\n  color: #d72d30;\n}\n.input-container-inner.warning > input {\n  border-color: #f8ab59;\n}\n.input-container-inner.warning > input:focus {\n  -webkit-box-shadow: none;\n  box-shadow: none;\n  border-color: #f8ab59;\n}\n.input-container-inner.warning > input:hover {\n  border-color: #f8ab59;\n}\n.input-container-inner.warning + .help-block {\n  color: #f8ab59;\n}\n\n::-webkit-input-placeholder {\n  color: #acb6be;\n}\n\n::-moz-placeholder {\n  color: #acb6be;\n}\n\n:-ms-input-placeholder {\n  color: #acb6be;\n}\n\n:-moz-placeholder {\n  color: #acb6be;\n}\n\n.icon {\n  position: absolute;\n  top: 8px;\n  left: 3px;\n  width: 25px;\n  height: 25px;\n  border-radius: 3px;\n  padding: 3px 0 0 6px;\n  display: block;\n  -ms-flex-pack: center;\n  -ms-flex-align: center;\n  align-items: center;\n  background-color: #ebeff3;\n}\n.icon.left {\n  left: 3px;\n}\n.icon.left::after {\n  left: 28px;\n  content: \"\";\n  width: 1px;\n  display: block;\n  height: 20px;\n  background-color: #ebeff3;\n  position: absolute;\n  top: 3px;\n}\n.icon.right {\n  left: auto;\n  right: 3px;\n}\n.icon.right::before {\n  right: 28px;\n  content: \"\";\n  width: 1px;\n  display: block;\n  height: 20px;\n  background-color: #ebeff3;\n  position: absolute;\n  top: 3px;\n}\n\n.left-icon input {\n  padding-left: 38px;\n}\n\n.right-icon input {\n  padding-right: 38px;\n}\n\ninput::-webkit-outer-spin-button,\ninput::-webkit-inner-spin-button {\n  -webkit-appearance: none;\n  margin: 0;\n}\n\ninput[type=number] {\n  -moz-appearance: textfield;\n}"; }
};

export { Input as fw_input };
