// @freshworks/crayons/dist/collection/components/select-option/select-option.js
import { h } from "@stencil/core";
export class SelectOption {
    constructor() {
        /**
         * Sets the state of the option to selected. The selected option is highlighted and a check mark is displayed next to it. If the attribute’s value is undefined, the value is set to false.
         */
        this.selected = false;
        /**
         * Sets the state of the option to disabled. The selected option is disabled and greyed out. If the attribute’s value is undefined, the value is set to false.
         */
        this.disabled = false;
        /**
         * States that the option is an HTML value. If the attribute's value is undefined, the value is set to true.
         */
        this.html = false;
    }
    onOptionSelected() {
        if (this.disabled) {
            return;
        }
        this.selected = !this.selected;
        const { value, selected } = this;
        this.fwSelected.emit({ value, selected });
    }
    render() {
        return (h("li", { ref: el => this.nativeLi = el, class: { 'select-option': true, 'selected': this.selected, 'disabled': this.disabled }, onMouseDown: () => this.onOptionSelected() }, this.html
            ? '' :
            h("slot", null)));
    }
    componentDidLoad() {
        if (this.html) {
            this.nativeLi.innerHTML = this.htmlContent;
        }
    }
    static get is() { return "fw-select-option"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() { return {
        "$": ["select-option.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["select-option.css"]
    }; }
    static get properties() { return {
        "value": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Value corresponding to the option, that is saved  when the form data is saved."
            },
            "attribute": "value",
            "reflect": true
        },
        "selected": {
            "type": "boolean",
            "mutable": true,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Sets the state of the option to selected. The selected option is highlighted and a check mark is displayed next to it. If the attribute\u2019s value is undefined, the value is set to false."
            },
            "attribute": "selected",
            "reflect": true,
            "defaultValue": "false"
        },
        "disabled": {
            "type": "boolean",
            "mutable": true,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Sets the state of the option to disabled. The selected option is disabled and greyed out. If the attribute\u2019s value is undefined, the value is set to false."
            },
            "attribute": "disabled",
            "reflect": true,
            "defaultValue": "false"
        },
        "html": {
            "type": "boolean",
            "mutable": true,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "States that the option is an HTML value. If the attribute's value is undefined, the value is set to true."
            },
            "attribute": "html",
            "reflect": true,
            "defaultValue": "false"
        },
        "optionText": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Alternate text displayed on the interface, in place of the actual HTML content."
            },
            "attribute": "option-text",
            "reflect": true
        },
        "htmlContent": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "HTML content that is displayed as the option."
            },
            "attribute": "html-content",
            "reflect": false
        }
    }; }
    static get events() { return [{
            "method": "fwSelected",
            "name": "fwSelected",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": "Triggered when an option is selected."
            },
            "complexType": {
                "original": "any",
                "resolved": "any",
                "references": {}
            }
        }]; }
}
