// @freshworks/crayons/dist/collection/components/button/button.js
import { Host, h } from "@stencil/core";
export class Button {
    constructor() {
        /**
         *  Button type based on which actions are performed when the button is clicked.
         */
        this.type = 'button';
        /**
         * Identifier of  the theme based on which the button is styled.
         */
        this.color = 'primary';
        /**
         * Disables the button on the interface. If the attribute’s value is undefined, the value is set to false.
         */
        this.disabled = false;
        /**
         * Sets the button to a full-width block. If the attribute’s value is undefined, the value is set to false.
         */
        this.expand = false;
        /**
         * Size of the button.
         */
        this.size = 'normal';
        /**
         *  Accepts the id of the fw-modal component to open it on click
         */
        this.modalTriggerId = '';
    }
    onFocus() {
        this.fwFocus.emit();
    }
    onBlur() {
        this.fwBlur.emit();
    }
    handleClick(event) {
        if (this.modalTriggerId !== '') {
            const modal = document.getElementById(this.modalTriggerId);
            modal.visible = true;
        }
        else if (this.type === 'submit') {
            const form = this.host.closest('form');
            if (form) {
                event.preventDefault();
                const fakeSubmit = document.createElement('button');
                fakeSubmit.type = 'submit';
                fakeSubmit.style.display = 'none';
                form.appendChild(fakeSubmit);
                fakeSubmit.click();
                fakeSubmit.remove();
            }
        }
        this.fwClick.emit();
    }
    render() {
        return (h(Host, { onClick: (e) => this.disabled ? undefined : this.handleClick(e), onFocus: () => this.onFocus(), onBlur: () => this.onBlur() },
            h("button", { type: this.type, class: `
            fw-btn fw-btn--${this.color.toLowerCase()}
            fw-btn--${this.size.toLowerCase()}
            ${this.expand ? 'fw-btn--block' : ''}
            `, disabled: this.disabled },
                h("slot", null))));
    }
    static get is() { return "fw-button"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() { return {
        "$": ["button.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["button.css"]
    }; }
    static get properties() { return {
        "type": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "'button' | 'reset' | 'submit'",
                "resolved": "\"button\" | \"reset\" | \"submit\"",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Button type based on which actions are performed when the button is clicked."
            },
            "attribute": "type",
            "reflect": false,
            "defaultValue": "'button'"
        },
        "color": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "'primary' | 'secondary' | 'danger' | 'link' | 'text'",
                "resolved": "\"danger\" | \"link\" | \"primary\" | \"secondary\" | \"text\"",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Identifier of  the theme based on which the button is styled."
            },
            "attribute": "color",
            "reflect": false,
            "defaultValue": "'primary'"
        },
        "disabled": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Disables the button on the interface. If the attribute\u2019s value is undefined, the value is set to false."
            },
            "attribute": "disabled",
            "reflect": true,
            "defaultValue": "false"
        },
        "expand": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Sets the button to a full-width block. If the attribute\u2019s value is undefined, the value is set to false."
            },
            "attribute": "expand",
            "reflect": false,
            "defaultValue": "false"
        },
        "size": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "'normal' | 'mini' | 'small'",
                "resolved": "\"mini\" | \"normal\" | \"small\"",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Size of the button."
            },
            "attribute": "size",
            "reflect": false,
            "defaultValue": "'normal'"
        },
        "modalTriggerId": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Accepts the id of the fw-modal component to open it on click"
            },
            "attribute": "modal-trigger-id",
            "reflect": false,
            "defaultValue": "''"
        }
    }; }
    static get events() { return [{
            "method": "fwClick",
            "name": "fwClick",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": "Triggered when the button is clicked."
            },
            "complexType": {
                "original": "void",
                "resolved": "void",
                "references": {}
            }
        }, {
            "method": "fwFocus",
            "name": "fwFocus",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": "Triggered when the button comes into focus."
            },
            "complexType": {
                "original": "void",
                "resolved": "void",
                "references": {}
            }
        }, {
            "method": "fwBlur",
            "name": "fwBlur",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": "Triggered when the button loses focus."
            },
            "complexType": {
                "original": "void",
                "resolved": "void",
                "references": {}
            }
        }]; }
    static get elementRef() { return "host"; }
}
