// @freshworks/crayons/dist/collection/components/checkbox/checkbox.js
import { Host, h } from "@stencil/core";
import { renderHiddenField } from '../../utils/utils';
export class Checkbox {
    constructor() {
        /**
         * Sets the state of the check box to selected. If the attribute’s value is undefined, the value is set to false.
         */
        this.checked = false;
        /**
         * Disables the check box on the interface. If the attribute’s value is undefined, the value is set to false.
         */
        this.disabled = false;
        /**
         * Label displayed on the interface, for the check box.
         */
        this.label = '';
        /**
         * Name of the component, saved as part of form data.
         */
        this.name = '';
        /**
         * Identifier corresponding to the component, that is saved when the form data is saved.
         */
        this.value = '';
    }
    componentDidLoad() {
        this.checkbox.checked = this.checked;
        this.checkbox.disabled = this.disabled;
    }
    checkChanged(isChecked) {
        if (!this.disabled) {
            this.fwChange.emit({
                value: this.value,
                checked: isChecked,
            });
        }
        this.checkbox.checked = isChecked;
    }
    disabledChanged(isDisabled) {
        this.checkbox.disabled = isDisabled;
    }
    onFocus() {
        this.fwFocus.emit();
    }
    onBlur() {
        this.fwBlur.emit();
    }
    toggle() {
        if (!this.disabled) {
            this.checked = !this.checked;
        }
    }
    render() {
        const { host, name, value } = this;
        if (this.checked) {
            renderHiddenField(host, name, value);
        }
        return (h(Host, { class: "checkbox-container", onClick: () => this.toggle(), role: "checkbox", tabIndex: "0", "aria-disabled": this.disabled ? 'true' : 'false', "aria-checked": `${this.checked}`, onFocus: () => this.onFocus(), onBlur: () => this.onBlur() },
            h("input", { type: "checkbox", ref: el => this.checkbox = el }),
            h("label", null,
                h("span", { class: "text" },
                    h("slot", null)),
                h("br", null),
                this.label !== ''
                    ? h("span", { class: "label-field" }, this.label)
                    : '')));
    }
    static get is() { return "fw-checkbox"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() { return {
        "$": ["checkbox.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["checkbox.css"]
    }; }
    static get properties() { return {
        "checked": {
            "type": "boolean",
            "mutable": true,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Sets the state of the check box to selected. If the attribute\u2019s value is undefined, the value is set to false."
            },
            "attribute": "checked",
            "reflect": true,
            "defaultValue": "false"
        },
        "disabled": {
            "type": "boolean",
            "mutable": true,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Disables the check box on the interface. If the attribute\u2019s value is undefined, the value is set to false."
            },
            "attribute": "disabled",
            "reflect": true,
            "defaultValue": "false"
        },
        "label": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Label displayed on the interface, for the check box."
            },
            "attribute": "label",
            "reflect": false,
            "defaultValue": "''"
        },
        "name": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Name of the component, saved as part of form data."
            },
            "attribute": "name",
            "reflect": false,
            "defaultValue": "''"
        },
        "value": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Identifier corresponding to the component, that is saved when the form data is saved."
            },
            "attribute": "value",
            "reflect": false,
            "defaultValue": "''"
        }
    }; }
    static get events() { return [{
            "method": "fwChange",
            "name": "fwChange",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": "Triggered when the check box\u2019s value is modified."
            },
            "complexType": {
                "original": "any",
                "resolved": "any",
                "references": {}
            }
        }, {
            "method": "fwFocus",
            "name": "fwFocus",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": "Triggered when the check box comes into focus."
            },
            "complexType": {
                "original": "void",
                "resolved": "void",
                "references": {}
            }
        }, {
            "method": "fwBlur",
            "name": "fwBlur",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": "Triggered when the check box loses focus."
            },
            "complexType": {
                "original": "void",
                "resolved": "void",
                "references": {}
            }
        }]; }
    static get elementRef() { return "host"; }
    static get watchers() { return [{
            "propName": "checked",
            "methodName": "checkChanged"
        }, {
            "propName": "disabled",
            "methodName": "disabledChanged"
        }]; }
}
