// @freshworks/crayons/dist/collection/components/radio-group/radio-group.js
import { Host, h } from "@stencil/core";
import { findCheckedOption, renderHiddenField, watchForOptions } from '../../utils/utils';
export class RadioGroup {
    constructor() {
        /**
         * If true, a radio group can be saved without selecting any option. If an option is selected, the selection can be cleared. If the attribute’s value is undefined, the value is set to false.
         */
        this.allowEmpty = false;
        /**
         * Name of the component, saved as part of form data.
         */
        this.name = '';
        this.onSelect = (ev) => {
            const selectedRadio = ev.target;
            if (selectedRadio) {
                this.value = selectedRadio.value;
            }
        };
        this.onDeselect = async (ev) => {
            const selectedRadio = ev.target;
            if (this.allowEmpty && selectedRadio.value === this.value) {
                this.value = undefined;
            }
            await this.updateRadios();
        };
    }
    async valueChanged(value) {
        await this.updateRadios();
        this.fwChange.emit({ value });
    }
    async connectedCallback() {
        const el = this.host;
        if (this.value === undefined) {
            const radio = findCheckedOption(el, 'fw-radio');
            if (radio !== undefined) {
                await radio.componentOnReady();
                if (this.value === undefined) {
                    this.value = radio.value;
                }
            }
        }
        this.mutationO = watchForOptions(el, 'fw-radio', async (newOption) => {
            if (newOption !== undefined) {
                newOption.componentOnReady().then(() => {
                    this.value = newOption.value;
                }).catch();
            }
            else {
                await this.updateRadios();
            }
        });
        await this.updateRadios();
    }
    disconnectedCallback() {
        if (this.mutationO) {
            this.mutationO.disconnect();
            this.mutationO = undefined;
        }
    }
    async updateRadios() {
        /**
         * Make sure we get all radios first
         * so values are up to date prior
         * to caching the radio group value
         */
        const radios = await this.getRadios();
        const { value } = this;
        let hasChecked = false;
        // Walk the DOM in reverse order, since the last selected one wins!
        for (const radio of radios) {
            if (!hasChecked && radio.value === value) {
                // correct value for this radio
                // but this radio isn't checked yet
                // and we haven't found a checked yet
                hasChecked = true;
                radio.checked = true;
            }
            else {
                // this radio doesn't have the correct value
                // or the radio group has been already checked
                radio.checked = false;
            }
        }
        // Reset value if
        if (!hasChecked) {
            this.value = undefined;
        }
    }
    getRadios() {
        return Promise.all(Array
            .from(this.host.querySelectorAll('fw-radio'))
            .map(r => r.componentOnReady()));
    }
    render() {
        const { host, name, value } = this;
        renderHiddenField(host, name, value);
        return (h(Host, { role: "radiogroup", onFwSelect: this.onSelect, onFwDeselect: this.onDeselect }));
    }
    static get is() { return "fw-radio-group"; }
    static get properties() { return {
        "allowEmpty": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "If true, a radio group can be saved without selecting any option. If an option is selected, the selection can be cleared. If the attribute\u2019s value is undefined, the value is set to false."
            },
            "attribute": "allow-empty",
            "reflect": false,
            "defaultValue": "false"
        },
        "name": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Name of the component, saved as part of form data."
            },
            "attribute": "name",
            "reflect": false,
            "defaultValue": "''"
        },
        "value": {
            "type": "any",
            "mutable": true,
            "complexType": {
                "original": "any | null",
                "resolved": "any",
                "references": {}
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "Default option that is selected when the radio group is displayed on the interface. Must be a valid value corresponding to the fw-radio components used in the Radio Group."
            },
            "attribute": "value",
            "reflect": false
        }
    }; }
    static get events() { return [{
            "method": "fwChange",
            "name": "fwChange",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": "Triggered when an option in the Radio Group is selected or deselected."
            },
            "complexType": {
                "original": "any",
                "resolved": "any",
                "references": {}
            }
        }]; }
    static get elementRef() { return "host"; }
    static get watchers() { return [{
            "propName": "value",
            "methodName": "valueChanged"
        }]; }
}
